<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Catalog &mdash; stdpopsim 0.1.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorials" href="tutorial.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> stdpopsim
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Catalog</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_AedAeg">Aedes aegypti</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_AedAeg_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_AnaPla">Anas platyrhynchos</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_AnaPla_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_AnaPla_models">Demographic Models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_anapla_models_mallardblackduck_2l19">North American Mallard/Black Duck split</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_AnoCar">Anolis carolinensis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_AnoCar_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_AnoGam">Anopheles gambiae</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_AnoGam_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_AnoGam_models">Demographic Models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_anogam_models_gabonag1000g_1a17">Stairwayplot estimates of N(t) for Gabon sample</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_ApiMel">Apis mellifera</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_ApiMel_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_AraTha">Arabidopsis thaliana</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_AraTha_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_AraTha_genetic_maps">Genetic Maps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_aratha_genetic_maps_salomeaveraged_tair10">SalomeAveraged_TAIR10</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_AraTha_models">Demographic Models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_aratha_models_southmiddleatlas_1d17">South Middle Atlas piecewise constant size</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_aratha_models_african2epoch_1h18">South Middle Atlas African two epoch model</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_aratha_models_african3epoch_1h18">South Middle Atlas African three epoch model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_BosTau">Bos taurus</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_BosTau_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_BosTau_models">Demographic Models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_bostau_models_holsteinfriesian_1m13">Piecewise size changes in Holstein-Friesian cattle.</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_CaeEle">Caenorhabditis elegans</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_CaeEle_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_CaeEle_genetic_maps">Genetic Maps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_caeele_genetic_maps_rockmanriail_ce11">RockmanRIAIL_ce11</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_CanFam">Canis familiaris</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_CanFam_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_CanFam_genetic_maps">Genetic Maps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_canfam_genetic_maps_campbell2016_canfam3_1">Campbell2016_CanFam3_1</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_ChlRei">Chlamydomonas reinhardtii</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_ChlRei_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_DroMel">Drosophila melanogaster</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_DroMel_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_DroMel_genetic_maps">Genetic Maps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_dromel_genetic_maps_comeroncrossover_dm6">ComeronCrossover_dm6</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_dromel_genetic_maps_comeroncrossoverv2_dm6">ComeronCrossoverV2_dm6</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_DroMel_models">Demographic Models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_dromel_models_african3epoch_1s16">Three epoch African population</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_dromel_models_outofafrica_2l06">Three epoch model for African and European populations</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_DroMel_annotations">Annotations</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_dromel_annotations_flybase_bdgp6.32.51_exons">FlyBase_BDGP6.32.51_exons</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_dromel_annotations_flybase_bdgp6.32.51_cds">FlyBase_BDGP6.32.51_CDS</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_DroMel_dfe">Distribution of Fitness Effects (DFEs)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_dromel_dfes_gamma_h17">Gamma_H17</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_dromel_dfes_lognormalpluspositive_r16">LognormalPlusPositive_R16</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_DroSec">Drosophila sechellia</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_DroSec_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_EscCol">Escherichia coli</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_EscCol_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_GasAcu">Gasterosteus aculeatus</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_GasAcu_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_HelAnn">Helianthus annuus</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_HelAnn_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_HelMel">Heliconius melpomene</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_HelMel_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_HomSap">Homo sapiens</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_HomSap_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_HomSap_genetic_maps">Genetic Maps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_hapmapii_grch37">HapMapII_GRCh37</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_decodesexaveraged_grch36">DeCodeSexAveraged_GRCh36</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoacb_grch38">PyrhoACB_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoasw_grch38">PyrhoASW_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhobeb_grch38">PyrhoBEB_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhocdx_grch38">PyrhoCDX_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoceu_grch38">PyrhoCEU_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhochb_grch38">PyrhoCHB_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhochs_grch38">PyrhoCHS_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoclm_grch38">PyrhoCLM_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoesn_grch38">PyrhoESN_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhofin_grch38">PyrhoFIN_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhogbr_grch38">PyrhoGBR_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhogih_grch38">PyrhoGIH_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhogwd_grch38">PyrhoGWD_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoibs_grch38">PyrhoIBS_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoitu_grch38">PyrhoITU_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhojpt_grch38">PyrhoJPT_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhokhv_grch38">PyrhoKHV_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrholwk_grch38">PyrhoLWK_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhomsl_grch38">PyrhoMSL_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhomxl_grch38">PyrhoMXL_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhopel_grch38">PyrhoPEL_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhopjl_grch38">PyrhoPJL_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhopur_grch38">PyrhoPUR_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhostu_grch38">PyrhoSTU_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhotsi_grch38">PyrhoTSI_GRCh38</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoyri_grch38">PyrhoYRI_GRCh38</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_HomSap_models">Demographic Models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_outofafricaextendedneandertaladmixturepulse_3i21">Three population out-of-Africa with an extended pulse of         Neandertal admixture into Europeans</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_outofafrica_3g09">Three population out-of-Africa</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_outofafrica_2t12">Two population out-of-Africa</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_africa_1t12">African population</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_americanadmixture_4b11">American admixture</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_outofafricaarchaicadmixture_5r19">Three population out-of-Africa with archaic admixture</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_zigzag_1s14">Periodic growth and decline.</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_ancienteurasia_9k19">Multi-population model of ancient Eurasia</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_papuansoutofafrica_10j19">Out-of-Africa with archaic admixture into Papuans</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_ashksub_7g19">Ashkenazi Jewish with substructure and European admixture</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_outofafrica_4j17">4 population out of Africa</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_africa_1b08">African-americans population</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_models_ancienteurope_4a21">Multi-population model of ancient Europe</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_HomSap_annotations">Annotations</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_annotations_ensembl_havana_104_exons">ensembl_havana_104_exons</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_annotations_ensembl_havana_104_cds">ensembl_havana_104_CDS</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_HomSap_dfe">Distribution of Fitness Effects (DFEs)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_dfes_gamma_k17">Gamma_K17</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_homsap_dfes_gamma_h17">Gamma_H17</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_PapAnu">Papio anubis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_PapAnu_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_PapAnu_models">Demographic Models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_papanu_models_singlepopsmcpp_1w22">SMC++ estimates of N(t) for Papio Anubis individuals</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_PonAbe">Pongo abelii</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_PonAbe_genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_PonAbe_genetic_maps">Genetic Maps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_ponabe_genetic_maps_naterpa_ponabe3">NaterPA_PonAbe3</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_ponabe_genetic_maps_naterpp_ponabe3">NaterPP_PonAbe3</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_PonAbe_models">Demographic Models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sec_catalog_ponabe_models_twospecies_2l11">Two population orangutan model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sec_catalog_StrAga">Streptococcus agalactiae</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sec_catalog_StrAga_genome">Genome</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#generic-models">Generic models</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli_arguments.html">Command-line options</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelogs.html">Changelogs</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">stdpopsim</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Catalog</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/catalog.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="catalog">
<span id="sec-catalog"></span><h1>Catalog<a class="headerlink" href="#catalog" title="Permalink to this headline"></a></h1>
<p>With <code class="docutils literal notranslate"><span class="pre">stdpopsim</span></code>, you can run simulations from a number of demographic models
that were implemented from published demographic histories. These models have been
rigorously checked and tested by multiple people, so you can rest easy knowing that
your simulations are reproducible and bug-free!</p>
<p>This catalog shows you all of the possible options that you can use to configure
your simulation.
It is organised around a number of choices that you’ll need to make about the
<a class="reference internal" href="api.html#stdpopsim.Species" title="stdpopsim.Species"><code class="xref py py-class docutils literal notranslate"><span class="pre">Species</span></code></a> you wish to simulate:</p>
<ol class="arabic simple">
<li><p>Which <strong>chromosome</strong>? (ie. which <a class="reference internal" href="api.html#stdpopsim.Genome" title="stdpopsim.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a> object?)</p></li>
<li><p>Which <strong>genetic map</strong>? (ie. which <code class="xref py py-class docutils literal notranslate"><span class="pre">GeneticMap</span></code> object?)</p></li>
<li><p>Which <strong>model of demographic history</strong>? (ie. which <a class="reference internal" href="api.html#stdpopsim.DemographicModel" title="stdpopsim.DemographicModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographicModel</span></code></a> object)</p></li>
</ol>
<p>For instance, suppose you are interested in simulating modern human samples of</p>
<ol class="arabic simple">
<li><p>chromosome 22, using</p></li>
<li><p>the HapMapII genetic map, under</p></li>
<li><p>a 3-population Out-of-Africa model.</p></li>
</ol>
<p>The following command simulates 2 samples from each of the three populations,
and saves the output to a file called <code class="docutils literal notranslate"><span class="pre">test.trees</span></code>:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>stdpopsim HomSap -c chr22 -o test.trees -g HapMapII_GRCh37 -d OutOfAfrica_3G09 <span class="m">2</span> <span class="m">2</span> <span class="m">2</span>
</pre></div>
</div>
<p>(To learn more about using <code class="docutils literal notranslate"><span class="pre">stdpopsim</span></code> via the command-line, read our
<a class="reference internal" href="tutorial.html#sec-cli-tute"><span class="std std-ref">tutorial</span></a> about it.)</p>
<p>Are there other well-known organisms, genetic maps or models that
you’d like to see in <code class="docutils literal notranslate"><span class="pre">stdpopsim</span></code>? Head to our <a class="reference internal" href="development.html#sec-development"><span class="std std-ref">Development</span></a>
page to learn about the process for adding new items to the catalog.
Then, if you feel ready, make an issue on our
<a class="reference external" href="https://github.com/popgensims/stdpopsim/issues">GitHub page</a>.</p>
<section id="sec_catalog_AedAeg">
<span id="sec_catalog_AedAeg"></span><h2>Aedes aegypti<a class="headerlink" href="#sec_catalog_AedAeg" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>AedAeg</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Aedes aegypti</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Yellow fever mosquito</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>0.06666666666666667 (<a class="reference external" href="https://doi.org/10.1186/s12915-017-0351-0">Crawford et al., 2017</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>1000000.0 (<a class="reference external" href="https://doi.org/10.1186/s12915-017-0351-0">Crawford et al., 2017</a>)</p>
</dd>
</dl>
<section id="sec_catalog_AedAeg_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_AedAeg_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>310827022</p></td>
<td><p>3.06e-09</p></td>
<td><p>3.5e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>474425716</p></td>
<td><p>2.49e-09</p></td>
<td><p>3.5e-09</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>409777670</p></td>
<td><p>2.91e-09</p></td>
<td><p>3.5e-09</p></td>
</tr>
<tr class="row-odd"><td><p>MT</p></td>
<td><p>16790</p></td>
<td><p>0</p></td>
<td><p>3.5e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_AnaPla">
<span id="sec_catalog_AnaPla"></span><h2>Anas platyrhynchos<a class="headerlink" href="#sec_catalog_AnaPla" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>AnaPla</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Anas platyrhynchos</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Mallard</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>4 (<a class="reference external" href="https://doi.org/10.1111/mec.15343">Lavretsky et al., 2020</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>156000 (<a class="reference external" href="https://doi.org/10.24272/j.issn.2095-8137.2020.133">Guo et al., 2021</a>)</p>
</dd>
</dl>
<section id="sec_catalog_AnaPla_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_AnaPla_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>208326429</p></td>
<td><p>1.52e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>162939446</p></td>
<td><p>1.39e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>119723720</p></td>
<td><p>9.38e-09</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>77626585</p></td>
<td><p>1.2e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>64988622</p></td>
<td><p>1.21e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>39543408</p></td>
<td><p>3.04e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>37812880</p></td>
<td><p>2.59e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>33348632</p></td>
<td><p>1.28e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>26742597</p></td>
<td><p>1.34e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>22933227</p></td>
<td><p>2.5e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>22193879</p></td>
<td><p>1.5e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>22338721</p></td>
<td><p>3.15e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p>21714986</p></td>
<td><p>1.41e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p>20320564</p></td>
<td><p>8.61e-09</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p>18227546</p></td>
<td><p>3.79e-09</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>16053328</p></td>
<td><p>3.86e-09</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p>15319648</p></td>
<td><p>3.13e-09</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>18</p></td>
<td><p>13333155</p></td>
<td><p>1.58e-09</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>19</p></td>
<td><p>12198306</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>20</p></td>
<td><p>12091001</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>21</p></td>
<td><p>8553409</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>22</p></td>
<td><p>16160689</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>23</p></td>
<td><p>7977799</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>24</p></td>
<td><p>7737077</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>25</p></td>
<td><p>7574731</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>26</p></td>
<td><p>6918023</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>27</p></td>
<td><p>6270716</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>28</p></td>
<td><p>5960150</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>29</p></td>
<td><p>1456683</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>30</p></td>
<td><p>1872559</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>Z</p></td>
<td><p>81233375</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>31</p></td>
<td><p>2637124</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>32</p></td>
<td><p>3473573</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>33</p></td>
<td><p>2151773</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>34</p></td>
<td><p>7214884</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>35</p></td>
<td><p>5548691</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>36</p></td>
<td><p>3997205</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>37</p></td>
<td><p>3148754</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>38</p></td>
<td><p>2836164</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-odd"><td><p>39</p></td>
<td><p>2018729</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
<tr class="row-even"><td><p>40</p></td>
<td><p>1354177</p></td>
<td><p>1.43e-08</p></td>
<td><p>4.83e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_AnaPla_models">
<h3>Demographic Models<a class="headerlink" href="#sec_catalog_AnaPla_models" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_anapla_models_mallardblackduck_2l19">MallardBlackDuck_2L19</a></p></td>
<td><p>North American Mallard/Black Duck split</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_anapla_models_mallardblackduck_2l19">
<span id="sec_catalog_anapla_models_mallardblackduck_2l19"></span><h4>North American Mallard/Black Duck split<a class="headerlink" href="#sec_catalog_anapla_models_mallardblackduck_2l19" title="Permalink to this headline"></a></h4>
<p>
        This is a model fit to contemporary samples of wild North American
        mallard and black duck, using the “split-migration” model of dadi.
        See Figure 6 of Lavretsky et al 2019.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>MallardBlackDuck_2L19</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>North American Mallard/Black Duck split</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>3</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>Mallard</p></td>
<td><p>None</p></td>
<td><p>Wild North American mallards</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>Black_duck</p></td>
<td><p>None</p></td>
<td><p>Wild black ducks</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>Ancestral</p></td>
<td><p>158076.25</p></td>
<td><p>Ancestral population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Lavretsky et al., 2019. <a class="reference external" href="https://doi.org/10.1111/mec.15343">https://doi.org/10.1111/mec.15343</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>819,535</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>1,570,000</p></td>
<td><p>Mallard pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>1,370,000</p></td>
<td><p>Black duck pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-6 per gen.)</p></td>
<td><p>0.84</p></td>
<td><p>Black-Mallard migration rate</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-6 per gen.)</p></td>
<td><p>1.10</p></td>
<td><p>Mallard-Black migration rate</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>158076</p></td>
<td><p>Mallard/Black split time</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>4</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate (subst/gen)</p></td>
<td><p>4.83e-9</p></td>
<td><p>Mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_anapla_models_mallardblackduck_2l19.png" src="_images/sec_catalog_anapla_models_mallardblackduck_2l19.png" />
</section>
</section>
<section id="sec_catalog_AnoCar">
<span id="sec_catalog_AnoCar"></span><h2>Anolis carolinensis<a class="headerlink" href="#sec_catalog_AnoCar" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>AnoCar</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Anolis carolinensis</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Anole lizard</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>1.5 (<a class="reference external" href="https://doi.org/10.1093/ilar.45.1.54">Lovern et al., 2004</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>3050000.0 (<a class="reference external" href="https://doi.org/10.1093/gbe/evz110">Pombi et al., 2019</a>)</p>
</dd>
</dl>
<section id="sec_catalog_AnoCar_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_AnoCar_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>263920458</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>199619895</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>204416410</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>156502444</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>150641573</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>80741955</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-even"><td><p>LGa</p></td>
<td><p>7025928</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-odd"><td><p>LGb</p></td>
<td><p>3271537</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-even"><td><p>LGc</p></td>
<td><p>9478905</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-odd"><td><p>LGd</p></td>
<td><p>1094478</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-even"><td><p>LGf</p></td>
<td><p>4257874</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-odd"><td><p>LGg</p></td>
<td><p>424765</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-even"><td><p>LGh</p></td>
<td><p>248369</p></td>
<td><p>1e-08</p></td>
<td><p>2.1e-10</p></td>
</tr>
<tr class="row-odd"><td><p>MT</p></td>
<td><p>17223</p></td>
<td><p>0</p></td>
<td><p>2.1e-10</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_AnoGam">
<span id="sec_catalog_AnoGam"></span><h2>Anopheles gambiae<a class="headerlink" href="#sec_catalog_AnoGam" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>AnoGam</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Anopheles gambiae</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Anopheles gambiae</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>0.09090909090909091 (<a class="reference external" href="https://doi.org/10.1038/nature24995">Ag1000G Consortium, 2017</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>1000000.0 (<a class="reference external" href="https://doi.org/10.1038/nature24995">Ag1000G Consortium, 2017</a>)</p>
</dd>
</dl>
<section id="sec_catalog_AnoGam_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_AnoGam_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2L</p></td>
<td><p>49364325</p></td>
<td><p>1.3e-08</p></td>
<td><p>3.5e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2R</p></td>
<td><p>61545105</p></td>
<td><p>1.3e-08</p></td>
<td><p>3.5e-09</p></td>
</tr>
<tr class="row-even"><td><p>3L</p></td>
<td><p>41963435</p></td>
<td><p>1.3e-08</p></td>
<td><p>3.5e-09</p></td>
</tr>
<tr class="row-odd"><td><p>3R</p></td>
<td><p>53200684</p></td>
<td><p>1.6e-08</p></td>
<td><p>3.5e-09</p></td>
</tr>
<tr class="row-even"><td><p>X</p></td>
<td><p>24393108</p></td>
<td><p>2.04e-08</p></td>
<td><p>3.5e-09</p></td>
</tr>
<tr class="row-odd"><td><p>Mt</p></td>
<td><p>15363</p></td>
<td><p>0</p></td>
<td><p>3.5e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_AnoGam_models">
<h3>Demographic Models<a class="headerlink" href="#sec_catalog_AnoGam_models" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_anogam_models_gabonag1000g_1a17">GabonAg1000G_1A17</a></p></td>
<td><p>Stairwayplot estimates of N(t) for Gabon sample</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_anogam_models_gabonag1000g_1a17">
<span id="sec_catalog_anogam_models_gabonag1000g_1a17"></span><h4>Stairwayplot estimates of N(t) for Gabon sample<a class="headerlink" href="#sec_catalog_anogam_models_gabonag1000g_1a17" title="Permalink to this headline"></a></h4>
<p>
        These estimates were done as part of the Ag1000G 2017 Consortium paper.
        Stairwayplot was run with the addition of a
        misorientation parameter using SFS information from each population.
        The model contains 110 distinct epochs, so only some summaries
        are reported in the population table.
        </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>GabonAg1000G_1A17</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Stairwayplot estimates of N(t) for Gabon sample</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>GAS</p></td>
<td><p>0</p></td>
<td><p>Gabon gambiae population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Ag1000G Consortium, 2017. <a class="reference external" href="https://doi.org/10.1038/nature24995">https://doi.org/10.1038/nature24995</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>4,069,863</p></td>
<td><p>Modern population size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>205,766</p></td>
<td><p>Mean coalescence time / 4</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>1/11</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate (subst/gen)</p></td>
<td><p>3.5e-9</p></td>
<td><p>Mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_anogam_models_gabonag1000g_1a17.png" src="_images/sec_catalog_anogam_models_gabonag1000g_1a17.png" />
</section>
</section>
<section id="sec_catalog_ApiMel">
<span id="sec_catalog_ApiMel"></span><h2>Apis mellifera<a class="headerlink" href="#sec_catalog_ApiMel" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>ApiMel</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Apis mellifera</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Apis mellifera (DH4)</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>2 (<a class="reference external" href="https://doi.org/10.1111/mec.14122">Nelson et al., 2017</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>200000.0 (<a class="reference external" href="https://doi.org/10.1038/ng.3077">Wallberg et al., 2014</a>)</p>
</dd>
</dl>
<section id="sec_catalog_ApiMel_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_ApiMel_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CM009931.2</p></td>
<td><p>27754200</p></td>
<td><p>2.39e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>CM009932.2</p></td>
<td><p>16089512</p></td>
<td><p>2.46e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-even"><td><p>CM009933.2</p></td>
<td><p>13619445</p></td>
<td><p>2.41e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>CM009934.2</p></td>
<td><p>13404451</p></td>
<td><p>2.76e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-even"><td><p>CM009935.2</p></td>
<td><p>13896941</p></td>
<td><p>2.14e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>CM009936.2</p></td>
<td><p>17789102</p></td>
<td><p>2.12e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-even"><td><p>CM009937.2</p></td>
<td><p>14198698</p></td>
<td><p>2.34e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>CM009938.2</p></td>
<td><p>12717210</p></td>
<td><p>2.09e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-even"><td><p>CM009939.2</p></td>
<td><p>12354651</p></td>
<td><p>2.46e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>CM009940.2</p></td>
<td><p>12360052</p></td>
<td><p>2.48e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-even"><td><p>CM009941.2</p></td>
<td><p>16352600</p></td>
<td><p>2.03e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>CM009942.2</p></td>
<td><p>11514234</p></td>
<td><p>2.12e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-even"><td><p>CM009943.2</p></td>
<td><p>11279722</p></td>
<td><p>2.34e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>CM009944.2</p></td>
<td><p>10670842</p></td>
<td><p>2.46e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-even"><td><p>CM009945.2</p></td>
<td><p>9534514</p></td>
<td><p>2.21e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>CM009946.2</p></td>
<td><p>7238532</p></td>
<td><p>2.28e-07</p></td>
<td><p>3.4e-09</p></td>
</tr>
<tr class="row-even"><td><p>CM009947.2</p></td>
<td><p>16343</p></td>
<td><p>0</p></td>
<td><p>3.4e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_AraTha">
<span id="sec_catalog_AraTha"></span><h2>Arabidopsis thaliana<a class="headerlink" href="#sec_catalog_AraTha" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>AraTha</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Arabidopsis thaliana</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>A. thaliana</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>1.0 (<a class="reference external" href="https://doi.org/10.1890/0012-9658(2002)083[1006:GTINSO]2.0.CO;2">Donohue, 2002</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>10000 (<a class="reference external" href="https://doi.org/10.1016/j.cell.2016.05.063">1001GenomesConsortium, 2016</a>)</p>
</dd>
</dl>
<section id="sec_catalog_AraTha_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_AraTha_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>30427671</p></td>
<td><p>8.06452e-10</p></td>
<td><p>7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>19698289</p></td>
<td><p>8.06452e-10</p></td>
<td><p>7e-09</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>23459830</p></td>
<td><p>8.06452e-10</p></td>
<td><p>7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>18585056</p></td>
<td><p>8.06452e-10</p></td>
<td><p>7e-09</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>26975502</p></td>
<td><p>8.06452e-10</p></td>
<td><p>7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>Mt</p></td>
<td><p>366924</p></td>
<td><p>0</p></td>
<td><p>7e-09</p></td>
</tr>
<tr class="row-even"><td><p>Pt</p></td>
<td><p>154478</p></td>
<td><p>0</p></td>
<td><p>7e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_AraTha_genetic_maps">
<h3>Genetic Maps<a class="headerlink" href="#sec_catalog_AraTha_genetic_maps" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_aratha_genetic_maps_salomeaveraged_tair10">SalomeAveraged_TAIR10</a></p></td>
<td><p>2012</p></td>
<td><p>Crossover frequency map averaged over 17 populations</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_aratha_genetic_maps_salomeaveraged_tair10">
<span id="sec_catalog_aratha_genetic_maps_salomeaveraged_tair10"></span><h4>SalomeAveraged_TAIR10<a class="headerlink" href="#sec_catalog_aratha_genetic_maps_salomeaveraged_tair10" title="Permalink to this headline"></a></h4>
<p>
        This map is based on the study of crossover frequencies in over 7000
        plants in 17 F2 populations derived from crosses between 18 A. thaliana
        accessions. Salomé et al provide genetic maps for each of these
        populations. To get a single map for each chromosome, the Haldane map
        function distances were converted to recombination rates (cM/Mb) for
        each cross and then averaged across the 17 populations using loess.
        The map was constructed on genome version TAIR7, and lifted to TAIR10.
        </p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Salomé et al., 2012. <a class="reference external" href="https://doi.org/10.1038/hdy.2011.95">https://doi.org/10.1038/hdy.2011.95</a></p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_AraTha_models">
<h3>Demographic Models<a class="headerlink" href="#sec_catalog_AraTha_models" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_aratha_models_southmiddleatlas_1d17">SouthMiddleAtlas_1D17</a></p></td>
<td><p>South Middle Atlas piecewise constant size</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_aratha_models_african2epoch_1h18">African2Epoch_1H18</a></p></td>
<td><p>South Middle Atlas African two epoch model</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_aratha_models_african3epoch_1h18">African3Epoch_1H18</a></p></td>
<td><p>South Middle Atlas African three epoch model</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_aratha_models_southmiddleatlas_1d17">
<span id="sec_catalog_aratha_models_southmiddleatlas_1d17"></span><h4>South Middle Atlas piecewise constant size<a class="headerlink" href="#sec_catalog_aratha_models_southmiddleatlas_1d17" title="Permalink to this headline"></a></h4>
<p>
            This model comes from MSMC using two randomly sampled homozygous
            individuals (Khe32 and Ifr4) from the South Middle Atlas region
            from the Middle Atlas Mountains in Morocco. The model is estimated
            with 32 time periods. Because estimates from the recent and ancient
            past are less accurate, we set the population size in the first 7
            time periods equal to the size at the 8th time period and the size
            during last 2 time periods equal to the size in the 30th time
            period.
        </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>SouthMiddleAtlas_1D17</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>South Middle Atlas piecewise constant size</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>SouthMiddleAtlas</p></td>
<td><p>0</p></td>
<td><p>Arabidopsis Thaliana South Middle Atlas population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Durvasula et al., 2017. <a class="reference external" href="https://doi.org/10.1073/pnas.1616736114">https://doi.org/10.1073/pnas.1616736114</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>73,989</p></td>
<td><p>Ancestral population size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>121,796</p></td>
<td><p>Pop. size during 1st time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>165,210</p></td>
<td><p>Pop. size during 2nd time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>198,019</p></td>
<td><p>Pop. size during 3rd time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>217,752</p></td>
<td><p>Pop. size during 4th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>228,222</p></td>
<td><p>Pop. size during 5th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>238,593</p></td>
<td><p>Pop. size during 6th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>246,984</p></td>
<td><p>Pop. size during 7th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>241,400</p></td>
<td><p>Pop. size during 8th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>217,331</p></td>
<td><p>Pop. size during 9th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>181,571</p></td>
<td><p>Pop. size during 10th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>143456</p></td>
<td><p>Pop. size during 11th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>111,644</p></td>
<td><p>Pop. size during 12th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>91,813</p></td>
<td><p>Pop. size during 13th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>83,829</p></td>
<td><p>Pop. size during 14th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>83,932</p></td>
<td><p>Pop. size during 15th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>87,661</p></td>
<td><p>Pop. size during 16th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>96,283</p></td>
<td><p>Pop. size during 17th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>110,745</p></td>
<td><p>Pop. size during 18th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>111,132</p></td>
<td><p>Pop. size during 19th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>78,908</p></td>
<td><p>Pop. size during 20th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>1,537,686</p></td>
<td><p>Begining of 1st time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>1,119,341</p></td>
<td><p>Begining of 2nd time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>954,517</p></td>
<td><p>Begining of 3rd time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>813,610</p></td>
<td><p>Begining of 4th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>693,151</p></td>
<td><p>Begining of 5th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>590,173</p></td>
<td><p>Begining of 6th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>502,139</p></td>
<td><p>Begining of 7th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>426,879</p></td>
<td><p>Begining of 8th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>362,541</p></td>
<td><p>Begining of 9th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>307,540</p></td>
<td><p>Begining of 10th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>260,520</p></td>
<td><p>Begining of 11th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>220,324</p></td>
<td><p>Begining of 12th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>185,960</p></td>
<td><p>Begining of 13th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>156,584</p></td>
<td><p>Begining of 14th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>131,471</p></td>
<td><p>Begining of 15th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>110,001</p></td>
<td><p>Begining of 16th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>91,648</p></td>
<td><p>Begining of 17th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>75,958</p></td>
<td><p>Begining of 18th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>62,544</p></td>
<td><p>Begining of 19th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>51,077</p></td>
<td><p>Begining of 20th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time (yrs.)</p></td>
<td><p>1</p></td>
<td><p>Average generation interval</p></td>
</tr>
<tr class="row-even"><td><p>Mutation rate</p></td>
<td><p>7.1e-9</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_aratha_models_southmiddleatlas_1d17.png" src="_images/sec_catalog_aratha_models_southmiddleatlas_1d17.png" />
<hr class="docutils" />
<section id="sec_catalog_aratha_models_african2epoch_1h18">
<span id="sec_catalog_aratha_models_african2epoch_1h18"></span><h4>South Middle Atlas African two epoch model<a class="headerlink" href="#sec_catalog_aratha_models_african2epoch_1h18" title="Permalink to this headline"></a></h4>
<p>
            Model estimated from site frequency spectrum of synonymous
            SNPs from African South Middle Atlas samples using
            Williamson et al. 2005 methodology. Values come from supplementary
            table 1 of Huber et al 2018. Sizes change from N_A -&gt; N_0 and t_1 is
            time of the second epoch.
        </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>African2Epoch_1H18</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>South Middle Atlas African two epoch model</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>SouthMiddleAtlas</p></td>
<td><p>0</p></td>
<td><p>Arabidopsis Thaliana South Middle Atlas population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Huber et al., 2018. <a class="reference external" href="https://doi.org/10.1038/s41467-018-05281-7">https://doi.org/10.1038/s41467-018-05281-7</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>746,148</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>100,218</p></td>
<td><p>Pop. size during second epoch</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>568,344</p></td>
<td><p>Time of second epoch</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time (yrs.)</p></td>
<td><p>1</p></td>
<td><p>Average generation interval</p></td>
</tr>
<tr class="row-even"><td><p>Mutation rate</p></td>
<td><p>7e-9</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_aratha_models_african2epoch_1h18.png" src="_images/sec_catalog_aratha_models_african2epoch_1h18.png" />
<hr class="docutils" />
<section id="sec_catalog_aratha_models_african3epoch_1h18">
<span id="sec_catalog_aratha_models_african3epoch_1h18"></span><h4>South Middle Atlas African three epoch model<a class="headerlink" href="#sec_catalog_aratha_models_african3epoch_1h18" title="Permalink to this headline"></a></h4>
<p>
            Model estimated from site frequency spectrum of synonymous
            SNPs from African (South Middle Atlas) samples using Williamson et
            al. 2005 methodology. Values come from supplementary table 1 of
            Huber et al 2018. Sizes change from N_A -&gt; N_2 -&gt; N_3 and t_2 is
            the time of the second epoch and t_3 is the time of the 3rd epoch.
        </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>African3Epoch_1H18</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>South Middle Atlas African three epoch model</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>SouthMiddleAtlas</p></td>
<td><p>0</p></td>
<td><p>Arabidopsis Thaliana South Middle Atlas population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Huber et al., 2018. <a class="reference external" href="https://doi.org/10.1038/s41467-018-05281-7">https://doi.org/10.1038/s41467-018-05281-7</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>161,744</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>24,076</p></td>
<td><p>Pop. size during second epoch</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>203,077</p></td>
<td><p>Pop. size during Third epoch</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>7,420</p></td>
<td><p>Time of second epoch</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>14,534</p></td>
<td><p>Time of third epoch</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time (yrs.)</p></td>
<td><p>1</p></td>
<td><p>Average generation interval</p></td>
</tr>
<tr class="row-even"><td><p>Mutation rate</p></td>
<td><p>7e-9</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_aratha_models_african3epoch_1h18.png" src="_images/sec_catalog_aratha_models_african3epoch_1h18.png" />
</section>
</section>
<section id="sec_catalog_BosTau">
<span id="sec_catalog_BosTau"></span><h2>Bos taurus<a class="headerlink" href="#sec_catalog_BosTau" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>BosTau</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Bos taurus</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Cattle</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>5 (<a class="reference external" href="https://doi.org/10.1093/molbev/mst125">MacLeod et al., 2013</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>62000 (<a class="reference external" href="https://doi.org/10.1093/molbev/mst125">MacLeod et al., 2013</a>)</p>
</dd>
</dl>
<section id="sec_catalog_BosTau_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_BosTau_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>158534110</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>136231102</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>121005158</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>120000601</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>120089316</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>117806340</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>110682743</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>113319770</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>105454467</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>103308737</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>106982474</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>87216183</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p>83472345</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p>82403003</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p>85007780</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>81013979</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p>73167244</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>18</p></td>
<td><p>65820629</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>19</p></td>
<td><p>63449741</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>20</p></td>
<td><p>71974595</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>21</p></td>
<td><p>69862954</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>22</p></td>
<td><p>60773035</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>23</p></td>
<td><p>52498615</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>24</p></td>
<td><p>62317253</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>25</p></td>
<td><p>42350435</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>26</p></td>
<td><p>51992305</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>27</p></td>
<td><p>45612108</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>28</p></td>
<td><p>45940150</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>29</p></td>
<td><p>51098607</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-odd"><td><p>X</p></td>
<td><p>139009144</p></td>
<td><p>9.26e-09</p></td>
<td><p>1.2e-08</p></td>
</tr>
<tr class="row-even"><td><p>MT</p></td>
<td><p>16338</p></td>
<td><p>0</p></td>
<td><p>1.2e-08</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_BosTau_models">
<h3>Demographic Models<a class="headerlink" href="#sec_catalog_BosTau_models" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_bostau_models_holsteinfriesian_1m13">HolsteinFriesian_1M13</a></p></td>
<td><p>Piecewise size changes in Holstein-Friesian cattle.</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_bostau_models_holsteinfriesian_1m13">
<span id="sec_catalog_bostau_models_holsteinfriesian_1m13"></span><h4>Piecewise size changes in Holstein-Friesian cattle.<a class="headerlink" href="#sec_catalog_bostau_models_holsteinfriesian_1m13" title="Permalink to this headline"></a></h4>
<p>
    The piecewise-constant population size model of Holstein-Friesian cattle
    from MacLeod et al. (2013). Population sizes were estimated from inferred
    runs of homozygosity, with parameter values taken from Figure 4A and Table S1.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>HolsteinFriesian_1M13</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Piecewise size changes in Holstein-Friesian cattle.</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>Holstein_Friesian</p></td>
<td><p>0</p></td>
<td><p>Holstein-Friesian</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>MacLeod et al., 2013. <a class="reference external" href="https://doi.org/10.1093/molbev/mst125">https://doi.org/10.1093/molbev/mst125</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>62,000</p></td>
<td><p>Ancestral population size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>17,000</p></td>
<td><p>Pop. size during 1st time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>10,000</p></td>
<td><p>Pop. size during 2nd time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>7,000</p></td>
<td><p>Pop. size during 3rd time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>3,500</p></td>
<td><p>Pop. size during 4th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>2,500</p></td>
<td><p>Pop. size during 5th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>2,000</p></td>
<td><p>Pop. size during 6th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>1,500</p></td>
<td><p>Pop. size during 7th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>1,000</p></td>
<td><p>Pop. size during 8th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>350</p></td>
<td><p>Pop. size during 9th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>250</p></td>
<td><p>Pop. size during 10th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>120</p></td>
<td><p>Pop. size during 11th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>90</p></td>
<td><p>Pop. size during 12th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>33,154</p></td>
<td><p>Begining of 1st time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>3,354</p></td>
<td><p>Begining of 2nd time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>2,354</p></td>
<td><p>Begining of 3rd time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>1,754</p></td>
<td><p>Begining of 4th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>654</p></td>
<td><p>Begining of 5th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>454</p></td>
<td><p>Begining of 6th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>154</p></td>
<td><p>Begining of 7th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>24</p></td>
<td><p>Begining of 8th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>18</p></td>
<td><p>Begining of 9th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>12</p></td>
<td><p>Begining of 10th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>6</p></td>
<td><p>Begining of 11th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>3</p></td>
<td><p>Begining of 12th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time (yrs.)</p></td>
<td><p>5</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-even"><td><p>Mutation rate</p></td>
<td><p>9.4e-9</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_bostau_models_holsteinfriesian_1m13.png" src="_images/sec_catalog_bostau_models_holsteinfriesian_1m13.png" />
</section>
</section>
<section id="sec_catalog_CaeEle">
<span id="sec_catalog_CaeEle"></span><h2>Caenorhabditis elegans<a class="headerlink" href="#sec_catalog_CaeEle" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>CaeEle</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Caenorhabditis elegans</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>C. elegans</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>0.01 (<a class="reference external" href="https://doi.org/10.7554/eLife.05849">Frézal &amp; Félix, 2015</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>10000 (<a class="reference external" href="https://doi.org/10.1016/j.cub.2005.06.022">Barrière &amp; Félix, 2005</a>; <a class="reference external" href="https://doi.org/10.1534/genetics.105.048207">Cutter, 2006</a>)</p>
</dd>
</dl>
<section id="sec_catalog_CaeEle_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_CaeEle_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>I</p></td>
<td><p>15072434</p></td>
<td><p>3.12163e-11</p></td>
<td><p>1.84e-09</p></td>
</tr>
<tr class="row-odd"><td><p>II</p></td>
<td><p>15279421</p></td>
<td><p>3.52929e-11</p></td>
<td><p>1.84e-09</p></td>
</tr>
<tr class="row-even"><td><p>III</p></td>
<td><p>13783801</p></td>
<td><p>3.9066e-11</p></td>
<td><p>1.84e-09</p></td>
</tr>
<tr class="row-odd"><td><p>IV</p></td>
<td><p>17493829</p></td>
<td><p>2.7121e-11</p></td>
<td><p>1.84e-09</p></td>
</tr>
<tr class="row-even"><td><p>V</p></td>
<td><p>20924180</p></td>
<td><p>2.47057e-11</p></td>
<td><p>1.84e-09</p></td>
</tr>
<tr class="row-odd"><td><p>X</p></td>
<td><p>17718942</p></td>
<td><p>2.94724e-11</p></td>
<td><p>1.84e-09</p></td>
</tr>
<tr class="row-even"><td><p>MtDNA</p></td>
<td><p>13794</p></td>
<td><p>0</p></td>
<td><p>1.05e-07</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_CaeEle_genetic_maps">
<h3>Genetic Maps<a class="headerlink" href="#sec_catalog_CaeEle_genetic_maps" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_caeele_genetic_maps_rockmanriail_ce11">RockmanRIAIL_ce11</a></p></td>
<td><p>2009</p></td>
<td><p>Genetic map from recombinant inbred advanced intercross lines</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_caeele_genetic_maps_rockmanriail_ce11">
<span id="sec_catalog_caeele_genetic_maps_rockmanriail_ce11"></span><h4>RockmanRIAIL_ce11<a class="headerlink" href="#sec_catalog_caeele_genetic_maps_rockmanriail_ce11" title="Permalink to this headline"></a></h4>
<p>
        The authors genotyped 1454 nuclear SNP markers in 236 recombinant
        inbred advanced intercross lines (RIAILs).
        The genetic distances were estimated in r/qtl using the Haldane map
        function, treating observed recombination fractions as though they had
        been observed in a backcross. The marker density is sufficiently high
        that the exact form of map function employed has little effect on
        estimated genetic distances. The tip domains of each chromosome were
        defined by including all markers between the chromosome ends and the
        first recombination breakpoint observed in the RIAILs.
        The genetic map corresponds to the assembly ce11 (GCA_000002985.3).
        </p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Rockman &amp; Kruglyak, 2009. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1000419">https://doi.org/10.1371/journal.pgen.1000419</a></p></li>
</ul>
</section>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_CanFam">
<span id="sec_catalog_CanFam"></span><h2>Canis familiaris<a class="headerlink" href="#sec_catalog_CanFam" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>CanFam</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Canis familiaris</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Dog</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>3</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>13000 (<a class="reference external" href="https://doi.org/10.1038/nature04338">Lindblad-Toh et al., 2005</a>)</p>
</dd>
</dl>
<section id="sec_catalog_CanFam_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_CanFam_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>122678785</p></td>
<td><p>7.636e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>85426708</p></td>
<td><p>8.79852e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>91889043</p></td>
<td><p>8.00087e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>88276631</p></td>
<td><p>8.0523e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>88915250</p></td>
<td><p>9.34433e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>77573801</p></td>
<td><p>8.19219e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>80974532</p></td>
<td><p>7.29347e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>74330416</p></td>
<td><p>8.29131e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>61074082</p></td>
<td><p>9.28772e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>69331447</p></td>
<td><p>9.10715e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>74389097</p></td>
<td><p>7.63945e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>72498081</p></td>
<td><p>7.76106e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p>63241923</p></td>
<td><p>8.41302e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p>60966679</p></td>
<td><p>9.02812e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p>64190966</p></td>
<td><p>7.85675e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>59632846</p></td>
<td><p>8.61406e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p>64289059</p></td>
<td><p>9.71883e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>18</p></td>
<td><p>55844845</p></td>
<td><p>1.02993e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>19</p></td>
<td><p>53741614</p></td>
<td><p>1.04251e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>20</p></td>
<td><p>58134056</p></td>
<td><p>9.99097e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>21</p></td>
<td><p>50858623</p></td>
<td><p>1.0339e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>22</p></td>
<td><p>61439934</p></td>
<td><p>8.61505e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>23</p></td>
<td><p>52294480</p></td>
<td><p>9.12664e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>24</p></td>
<td><p>47698779</p></td>
<td><p>1.1146e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>25</p></td>
<td><p>51628933</p></td>
<td><p>1.15437e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>26</p></td>
<td><p>38964690</p></td>
<td><p>1.20846e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>27</p></td>
<td><p>45876710</p></td>
<td><p>1.12603e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>28</p></td>
<td><p>41182112</p></td>
<td><p>1.24636e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>29</p></td>
<td><p>41845238</p></td>
<td><p>1.10136e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>30</p></td>
<td><p>40214260</p></td>
<td><p>1.16876e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>31</p></td>
<td><p>39895921</p></td>
<td><p>1.13977e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>32</p></td>
<td><p>38810281</p></td>
<td><p>1.15559e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>33</p></td>
<td><p>31377067</p></td>
<td><p>1.33394e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>34</p></td>
<td><p>42124431</p></td>
<td><p>1.04838e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>35</p></td>
<td><p>26524999</p></td>
<td><p>1.42991e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>36</p></td>
<td><p>30810995</p></td>
<td><p>1.18752e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>37</p></td>
<td><p>30902991</p></td>
<td><p>1.38346e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>38</p></td>
<td><p>23914537</p></td>
<td><p>1.43637e-08</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-even"><td><p>X</p></td>
<td><p>123869142</p></td>
<td><p>9.50648e-09</p></td>
<td><p>4e-09</p></td>
</tr>
<tr class="row-odd"><td><p>MT</p></td>
<td><p>16727</p></td>
<td><p>0</p></td>
<td><p>4e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_CanFam_genetic_maps">
<h3>Genetic Maps<a class="headerlink" href="#sec_catalog_CanFam_genetic_maps" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_canfam_genetic_maps_campbell2016_canfam3_1">Campbell2016_CanFam3_1</a></p></td>
<td><p>2016</p></td>
<td><p>Pedigree-based crossover map from 237 individuals</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_canfam_genetic_maps_campbell2016_canfam3_1">
<span id="sec_catalog_canfam_genetic_maps_campbell2016_canfam3_1"></span><h4>Campbell2016_CanFam3_1<a class="headerlink" href="#sec_catalog_canfam_genetic_maps_campbell2016_canfam3_1" title="Permalink to this headline"></a></h4>
<p>
        Sex-averaged crossover frequency map based on 163,400 autosomal SNPs
        genotyped in a pedigree of 237 Labrador Retriever x Greyhound crosses.
        Genotypes were phased without respect to the pedigree, using SHAPEIT2,
        recombinations were called using duoHMM, and genetic distances were
        obtained using Haldane’s map function.
        </p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Campbell et al., 2016. <a class="reference external" href="https://doi.org/10.1534/g3.116.034678">https://doi.org/10.1534/g3.116.034678</a></p></li>
</ul>
</section>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_ChlRei">
<span id="sec_catalog_ChlRei"></span><h2>Chlamydomonas reinhardtii<a class="headerlink" href="#sec_catalog_ChlRei" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>ChlRei</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Chlamydomonas reinhardtii</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Chlamydomonas reinhardtii</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>0.001141552511415525 (<a class="reference external" href="https://doi.org/10.1007/s00425-011-1427-7">Vítová et al, 2011</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>1.3999999999999998e-07 (<a class="reference external" href="https://doi.org/10.1093/molbev/msv272">Ness et al., 2016</a>)</p>
</dd>
</dl>
<section id="sec_catalog_ChlRei_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_ChlRei_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>8033585</p></td>
<td><p>1.21e-10</p></td>
<td><p>9.74e-10</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>9223677</p></td>
<td><p>1.49e-10</p></td>
<td><p>8.62e-10</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>9219486</p></td>
<td><p>1.52e-10</p></td>
<td><p>9.5e-10</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>4091191</p></td>
<td><p>1.47e-10</p></td>
<td><p>9.66e-10</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>3500558</p></td>
<td><p>1.7e-10</p></td>
<td><p>1.17e-09</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>9023763</p></td>
<td><p>1.17e-10</p></td>
<td><p>9.12e-10</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>6421821</p></td>
<td><p>8.66e-11</p></td>
<td><p>9.14e-10</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>5033832</p></td>
<td><p>1.39e-10</p></td>
<td><p>8.98e-10</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>7956127</p></td>
<td><p>1.12e-10</p></td>
<td><p>9.17e-10</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>6576019</p></td>
<td><p>1.97e-10</p></td>
<td><p>9.27e-10</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>3826814</p></td>
<td><p>1.63e-10</p></td>
<td><p>1.03e-09</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>9730733</p></td>
<td><p>9.15e-11</p></td>
<td><p>9.55e-10</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p>5206065</p></td>
<td><p>1.43e-10</p></td>
<td><p>7.56e-10</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p>4157777</p></td>
<td><p>1.9e-10</p></td>
<td><p>8.96e-10</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p>1922860</p></td>
<td><p>3.93e-10</p></td>
<td><p>6.91e-10</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>7783580</p></td>
<td><p>1.71e-10</p></td>
<td><p>9.59e-10</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p>7188315</p></td>
<td><p>1.83e-10</p></td>
<td><p>1.05e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_DroMel">
<span id="sec_catalog_DroMel"></span><h2>Drosophila melanogaster<a class="headerlink" href="#sec_catalog_DroMel" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>DroMel</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Drosophila melanogaster</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>D. melanogaster</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>0.1 (<a class="reference external" href="https://doi.org/10.1371/journal.pgen.0020166">Li et al., 2006</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>1720600 (<a class="reference external" href="https://doi.org/10.1371/journal.pgen.0020166">Li et al., 2006</a>)</p>
</dd>
</dl>
<section id="sec_catalog_DroMel_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_DroMel_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2L</p></td>
<td><p>23513712</p></td>
<td><p>2.40463e-08</p></td>
<td><p>5.49e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2R</p></td>
<td><p>25286936</p></td>
<td><p>2.23459e-08</p></td>
<td><p>5.49e-09</p></td>
</tr>
<tr class="row-even"><td><p>3L</p></td>
<td><p>28110227</p></td>
<td><p>1.7966e-08</p></td>
<td><p>5.49e-09</p></td>
</tr>
<tr class="row-odd"><td><p>3R</p></td>
<td><p>32079331</p></td>
<td><p>1.71642e-08</p></td>
<td><p>5.49e-09</p></td>
</tr>
<tr class="row-even"><td><p>4</p></td>
<td><p>1348131</p></td>
<td><p>0</p></td>
<td><p>5.49e-09</p></td>
</tr>
<tr class="row-odd"><td><p>X</p></td>
<td><p>23542271</p></td>
<td><p>2.89651e-08</p></td>
<td><p>5.49e-09</p></td>
</tr>
<tr class="row-even"><td><p>Y</p></td>
<td><p>3667352</p></td>
<td><p>0</p></td>
<td><p>5.49e-09</p></td>
</tr>
<tr class="row-odd"><td><p>mitochondrion_genome</p></td>
<td><p>19524</p></td>
<td><p>0</p></td>
<td><p>5.49e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_DroMel_genetic_maps">
<h3>Genetic Maps<a class="headerlink" href="#sec_catalog_DroMel_genetic_maps" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_dromel_genetic_maps_comeroncrossover_dm6">ComeronCrossover_dm6</a></p></td>
<td><p>2012</p></td>
<td><p>Crossover map from meioses products of 8 lab crosses</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_dromel_genetic_maps_comeroncrossoverv2_dm6">ComeronCrossoverV2_dm6</a></p></td>
<td><p>2012</p></td>
<td><p>Crossover map from meioses products of 8 lab crosses</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_dromel_genetic_maps_comeroncrossover_dm6">
<span id="sec_catalog_dromel_genetic_maps_comeroncrossover_dm6"></span><h4>ComeronCrossover_dm6<a class="headerlink" href="#sec_catalog_dromel_genetic_maps_comeroncrossover_dm6" title="Permalink to this headline"></a></h4>
<p>
        The crossover map from a study of 8 crosses of 12 highly
        inbred lines of D. melanogaster. This is based on the
        products of 5,860 female meioses from whole genome sequencing data.
        Recombination rates were calculated from the density of individual
        recombination events that were detected in crosses. This map was
        subsequently lifted over to the dm6 assembly.
        </p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Comeron et al, 2012. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002905">https://doi.org/10.1371/journal.pgen.1002905</a></p></li>
</ul>
</section>
<section id="sec_catalog_dromel_genetic_maps_comeroncrossoverv2_dm6">
<span id="sec_catalog_dromel_genetic_maps_comeroncrossoverv2_dm6"></span><h4>ComeronCrossoverV2_dm6<a class="headerlink" href="#sec_catalog_dromel_genetic_maps_comeroncrossoverv2_dm6" title="Permalink to this headline"></a></h4>
<p>
        The crossover map from a study of 8 crosses of 12 highly
        inbred lines of D. melanogaster. This is based on the
        products of 5,860 female meioses from whole genome sequencing data.
        Recombination rates were calculated from the density of individual
        recombination events that were detected in crosses. This map was
        subsequently lifted over to the dm6 assembly using the available
        maintenance code command:

        python liftOver_comeron2012.py             –winLen 1000             –gapThresh 1000000             –useAdjacentAvg             –retainIntermediates
        </p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Comeron et al, 2012. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002905">https://doi.org/10.1371/journal.pgen.1002905</a></p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_DroMel_models">
<h3>Demographic Models<a class="headerlink" href="#sec_catalog_DroMel_models" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_dromel_models_african3epoch_1s16">African3Epoch_1S16</a></p></td>
<td><p>Three epoch African population</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_dromel_models_outofafrica_2l06">OutOfAfrica_2L06</a></p></td>
<td><p>Three epoch model for African and European populations</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_dromel_models_african3epoch_1s16">
<span id="sec_catalog_dromel_models_african3epoch_1s16"></span><h4>Three epoch African population<a class="headerlink" href="#sec_catalog_dromel_models_african3epoch_1s16" title="Permalink to this headline"></a></h4>
<p>
        The three epoch (modern, bottleneck, ancestral) model estimated for a
        single African Drosophila Melanogaster population from Sheehan and Song (2016).
        Population sizes are estimated by a
        deep learning model trained on simulation data. NOTE: Due to differences in
        coalescence units between PSMC (2N) and msms (4N) the number of generations were
        doubled from PSMC estimates when simulating data from msms in the original
        publication. We have faithfully represented the published model here.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>African3Epoch_1S16</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Three epoch African population</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>AFR</p></td>
<td><p>0</p></td>
<td><p>African D. melanogaster population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Sheehan and Song, 2016. <a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1004845">https://doi.org/10.1371/journal.pcbi.1004845</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>100,000</p></td>
<td><p>Reference population size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>652,700</p></td>
<td><p>Ancestral pop. Size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>145,300</p></td>
<td><p>Bottleneck pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>544,200</p></td>
<td><p>Recent pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>2,200,000</p></td>
<td><p>Onset of bottleneck</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>200,000</p></td>
<td><p>Population expansion</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>0.1</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>8.4e-9</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_dromel_models_african3epoch_1s16.png" src="_images/sec_catalog_dromel_models_african3epoch_1s16.png" />
<hr class="docutils" />
<section id="sec_catalog_dromel_models_outofafrica_2l06">
<span id="sec_catalog_dromel_models_outofafrica_2l06"></span><h4>Three epoch model for African and European populations<a class="headerlink" href="#sec_catalog_dromel_models_outofafrica_2l06" title="Permalink to this headline"></a></h4>
<p>
        The three epoch (modern, bottleneck, ancestral) model estimated for two
        Drosophila Melanogaster populations: African (ancestral) and European (derived)
        from Li and Stephan (2006).
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>OutOfAfrica_2L06</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Three epoch model for African and European populations</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>2</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>AFR</p></td>
<td><p>0</p></td>
<td><p>African D. melanogaster population</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>EUR</p></td>
<td><p>0</p></td>
<td><p>European D. melanogaster population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Li et al., 2006. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.0020166">https://doi.org/10.1371/journal.pgen.0020166</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>1,720,600</p></td>
<td><p>Ancestral pop. Size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>8,603,000</p></td>
<td><p>Post-expansion African pop. Size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>2,200</p></td>
<td><p>European bottleneck pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>1,075,000</p></td>
<td><p>Modern European pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>600,000</p></td>
<td><p>Expansion of population in Africa</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>158,000</p></td>
<td><p>African-European divergence</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>154,600</p></td>
<td><p>European pop. Expansion</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time (yrs.)</p></td>
<td><p>0.1</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-even"><td><p>Mutation rate</p></td>
<td><p>1.45e-9</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_dromel_models_outofafrica_2l06.png" src="_images/sec_catalog_dromel_models_outofafrica_2l06.png" />
</section>
<section id="sec_catalog_DroMel_annotations">
<h3>Annotations<a class="headerlink" href="#sec_catalog_DroMel_annotations" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_dromel_annotations_flybase_bdgp6.32.51_exons">FlyBase_BDGP6.32.51_exons</a></p></td>
<td><p>2014</p></td>
<td><p>FlyBase exon annotations on BDGP6</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_dromel_annotations_flybase_bdgp6.32.51_cds">FlyBase_BDGP6.32.51_CDS</a></p></td>
<td><p>2014</p></td>
<td><p>FlyBase CDS annotations on BDGP6</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_dromel_annotations_flybase_bdgp6.32.51_exons">
<span id="sec_catalog_dromel_annotations_flybase_bdgp6.32.51_exons"></span><h4>FlyBase_BDGP6.32.51_exons<a class="headerlink" href="#sec_catalog_dromel_annotations_flybase_bdgp6.32.51_exons" title="Permalink to this headline"></a></h4>
<p>FlyBase exon annotations on BDGP6</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Hoskins et al, 2014. <a class="reference external" href="https://doi.org/10.1101/gr.185579.114">https://doi.org/10.1101/gr.185579.114</a></p></li>
</ul>
</section>
<section id="sec_catalog_dromel_annotations_flybase_bdgp6.32.51_cds">
<span id="sec_catalog_dromel_annotations_flybase_bdgp6.32.51_cds"></span><h4>FlyBase_BDGP6.32.51_CDS<a class="headerlink" href="#sec_catalog_dromel_annotations_flybase_bdgp6.32.51_cds" title="Permalink to this headline"></a></h4>
<p>FlyBase CDS annotations on BDGP6</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Hoskins et al, 2014. <a class="reference external" href="https://doi.org/10.1101/gr.185579.114">https://doi.org/10.1101/gr.185579.114</a></p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_DroMel_dfe">
<h3>Distribution of Fitness Effects (DFEs)<a class="headerlink" href="#sec_catalog_DroMel_dfe" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_dromel_dfes_gamma_h17">Gamma_H17</a></p></td>
<td><p>2017</p></td>
<td><p>Deleterious Gamma DFE</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_dromel_dfes_lognormalpluspositive_r16">LognormalPlusPositive_R16</a></p></td>
<td><p>2016</p></td>
<td><p>Deleterious log-normal and beneficial mixed DFE</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_dromel_dfes_gamma_h17">
<span id="sec_catalog_dromel_dfes_gamma_h17"></span><h4>Gamma_H17<a class="headerlink" href="#sec_catalog_dromel_dfes_gamma_h17" title="Permalink to this headline"></a></h4>
<p>Deleterious Gamma DFE</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Huber et al., 2017. <a class="reference external" href="https://doi.org/10.1073/pnas.1619508114">https://doi.org/10.1073/pnas.1619508114</a></p></li>
</ul>
</section>
<section id="sec_catalog_dromel_dfes_lognormalpluspositive_r16">
<span id="sec_catalog_dromel_dfes_lognormalpluspositive_r16"></span><h4>LognormalPlusPositive_R16<a class="headerlink" href="#sec_catalog_dromel_dfes_lognormalpluspositive_r16" title="Permalink to this headline"></a></h4>
<p>Deleterious log-normal and beneficial mixed DFE</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Ragsdale et al., 2016. <a class="reference external" href="https://doi.org/10.1534/genetics.115.184812">https://doi.org/10.1534/genetics.115.184812</a></p></li>
</ul>
</section>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_DroSec">
<span id="sec_catalog_DroSec"></span><h2>Drosophila sechellia<a class="headerlink" href="#sec_catalog_DroSec" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>DroSec</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Drosophila sechellia</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Drosophila sechellia</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>0.05 (<a class="reference external" href="https://doi.org/10.1534/genetics.108.092080">Legrand et al., 2009</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>100000 (<a class="reference external" href="https://doi.org/10.1534/genetics.108.092080">Legrand et al., 2009</a>)</p>
</dd>
</dl>
<section id="sec_catalog_DroSec_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_DroSec_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2L</p></td>
<td><p>24956976</p></td>
<td><p>2.28e-08</p></td>
<td><p>1.5e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2R</p></td>
<td><p>21536224</p></td>
<td><p>2.51e-08</p></td>
<td><p>1.5e-09</p></td>
</tr>
<tr class="row-even"><td><p>3L</p></td>
<td><p>28131630</p></td>
<td><p>1.88e-08</p></td>
<td><p>1.5e-09</p></td>
</tr>
<tr class="row-odd"><td><p>3R</p></td>
<td><p>30464902</p></td>
<td><p>1.91e-08</p></td>
<td><p>1.5e-09</p></td>
</tr>
<tr class="row-even"><td><p>X</p></td>
<td><p>22909512</p></td>
<td><p>2.85e-08</p></td>
<td><p>1.5e-09</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>1277805</p></td>
<td><p>0</p></td>
<td><p>1.5e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_EscCol">
<span id="sec_catalog_EscCol"></span><h2>Escherichia coli<a class="headerlink" href="#sec_catalog_EscCol" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>EscCol</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Escherichia coli</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>E. coli</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>3.805175e-05 (<a class="reference external" href="https://doi.org/10.1128/JB.01368-07">Sezonov et al., 2007</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>180000000.0 (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1206133/">Hartl, Moriyama, and Sawyer, 1994</a>)</p>
</dd>
</dl>
<section id="sec_catalog_EscCol_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_EscCol_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Chromosome</p></td>
<td><p>4641652</p></td>
<td><p>0</p></td>
<td><p>8.9e-11</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_GasAcu">
<span id="sec_catalog_GasAcu"></span><h2>Gasterosteus aculeatus<a class="headerlink" href="#sec_catalog_GasAcu" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>GasAcu</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Gasterosteus aculeatus</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Three-spined stickleback</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>2 (<a class="reference external" href="https://doi.org/10.1111/mec.13827">Liu et al., 2016</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>10000.0 (<a class="reference external" href="https://doi.org/10.1111/mec.13827">Liu et al., 2016</a>)</p>
</dd>
</dl>
<section id="sec_catalog_GasAcu_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_GasAcu_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>29619991</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>23686546</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>17759012</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>34181212</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>15550311</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>18825451</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>30776923</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>20553084</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>20843631</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>17985176</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>17651971</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>20694444</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p>20748428</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p>16147532</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p>17318724</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>19507025</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p>20195758</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>18</p></td>
<td><p>15939336</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>19</p></td>
<td><p>20580295</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>20</p></td>
<td><p>20445003</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>21</p></td>
<td><p>17421465</p></td>
<td><p>3.11e-08</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-odd"><td><p>Y</p></td>
<td><p>15859692</p></td>
<td><p>0</p></td>
<td><p>3.7e-08</p></td>
</tr>
<tr class="row-even"><td><p>MT</p></td>
<td><p>16543</p></td>
<td><p>0</p></td>
<td><p>3.7e-08</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_HelAnn">
<span id="sec_catalog_HelAnn"></span><h2>Helianthus annuus<a class="headerlink" href="#sec_catalog_HelAnn" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>HelAnn</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Helianthus annuus</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Helianthus annuus</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>1.0 (<a class="reference external" href="https://doi.org/10.1111/j.1558-5646.2008.00415.x">Strasburg and Rieseberg, 2008</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>673968 (<a class="reference external" href="https://doi.org/10.1093/molbev/msq270">Strasburg et al., 2010</a>)</p>
</dd>
</dl>
<section id="sec_catalog_HelAnn_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_HelAnn_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>149502186</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>174800439</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>176490873</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>208320189</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>178169690</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>148147350</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>149542083</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>167167940</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>189665024</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>181411567</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>189830405</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>163781230</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p>173487274</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p>173346949</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p>175671323</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>206736614</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p>195042445</p></td>
<td><p>4e-09</p></td>
<td><p>6.1e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_HelMel">
<span id="sec_catalog_HelMel"></span><h2>Heliconius melpomene<a class="headerlink" href="#sec_catalog_HelMel" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>HelMel</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Heliconius melpomene</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Heliconius melpomene</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>0.0958904109589041 (<a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002752">Pardo-Diaz et al, 2012</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>2111109 (<a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002752">Pardo-Diaz et al, 2012</a>)</p>
</dd>
</dl>
<section id="sec_catalog_HelMel_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_HelMel_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>17206585</p></td>
<td><p>3.17e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>9045316</p></td>
<td><p>5.61e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>10541528</p></td>
<td><p>5.1e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>9662098</p></td>
<td><p>4.97e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>9908586</p></td>
<td><p>5.15e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>14054175</p></td>
<td><p>3.4e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>14308859</p></td>
<td><p>3.76e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>9320449</p></td>
<td><p>5.28e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>8708747</p></td>
<td><p>5.31e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>17965481</p></td>
<td><p>3.16e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>11759272</p></td>
<td><p>4.47e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>16327298</p></td>
<td><p>3.13e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p>18127314</p></td>
<td><p>3.08e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p>9174305</p></td>
<td><p>5.47e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p>10235750</p></td>
<td><p>4.78e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>10083215</p></td>
<td><p>4.71e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p>14773299</p></td>
<td><p>3.94e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>18</p></td>
<td><p>16803890</p></td>
<td><p>3.16e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>19</p></td>
<td><p>16399344</p></td>
<td><p>3.11e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-odd"><td><p>20</p></td>
<td><p>14871695</p></td>
<td><p>3.45e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
<tr class="row-even"><td><p>21</p></td>
<td><p>13359691</p></td>
<td><p>3.71e-08</p></td>
<td><p>2.9e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_HomSap">
<span id="sec_catalog_HomSap"></span><h2>Homo sapiens<a class="headerlink" href="#sec_catalog_HomSap" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>HomSap</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Homo sapiens</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Human</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>30 (<a class="reference external" href="https://doi.org/10.1086/302770">Tremblay and Vézina, 2000</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>10000 (<a class="reference external" href="https://doi.org/10.1093/oxfordjournals.molbev.a039995">Takahata, 1993</a>)</p>
</dd>
</dl>
<section id="sec_catalog_HomSap_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_HomSap_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>248956422</p></td>
<td><p>1.14856e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>242193529</p></td>
<td><p>1.10543e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>198295559</p></td>
<td><p>1.12796e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>190214555</p></td>
<td><p>1.12312e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>181538259</p></td>
<td><p>1.12809e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>170805979</p></td>
<td><p>1.12229e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>159345973</p></td>
<td><p>1.17646e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>145138636</p></td>
<td><p>1.14785e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>138394717</p></td>
<td><p>1.17807e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>133797422</p></td>
<td><p>1.33651e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>135086622</p></td>
<td><p>1.17193e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>133275309</p></td>
<td><p>1.30502e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p>114364328</p></td>
<td><p>1.09149e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p>107043718</p></td>
<td><p>1.11973e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p>101991189</p></td>
<td><p>1.38358e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>90338345</p></td>
<td><p>1.48346e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p>83257441</p></td>
<td><p>1.58249e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>18</p></td>
<td><p>80373285</p></td>
<td><p>1.5076e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>19</p></td>
<td><p>58617616</p></td>
<td><p>1.82201e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>20</p></td>
<td><p>64444167</p></td>
<td><p>1.71783e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>21</p></td>
<td><p>46709983</p></td>
<td><p>1.30452e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>22</p></td>
<td><p>50818468</p></td>
<td><p>1.4445e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>X</p></td>
<td><p>156040895</p></td>
<td><p>1.16466e-08</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-odd"><td><p>Y</p></td>
<td><p>57227415</p></td>
<td><p>0</p></td>
<td><p>1.29e-08</p></td>
</tr>
<tr class="row-even"><td><p>MT</p></td>
<td><p>16569</p></td>
<td><p>0</p></td>
<td><p>1.29e-08</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_HomSap_genetic_maps">
<h3>Genetic Maps<a class="headerlink" href="#sec_catalog_HomSap_genetic_maps" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_hapmapii_grch37">HapMapII_GRCh37</a></p></td>
<td><p>2007</p></td>
<td><p>HapMap Phase II lifted over to GRCh37</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_decodesexaveraged_grch36">DeCodeSexAveraged_GRCh36</a></p></td>
<td><p>2010</p></td>
<td><p>Sex averaged map from deCode family study</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoacb_grch38">PyrhoACB_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for ACB</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoasw_grch38">PyrhoASW_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for ASW</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhobeb_grch38">PyrhoBEB_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for BEB</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhocdx_grch38">PyrhoCDX_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for CDX</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoceu_grch38">PyrhoCEU_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for CEU</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhochb_grch38">PyrhoCHB_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for CHB</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhochs_grch38">PyrhoCHS_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for CHS</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoclm_grch38">PyrhoCLM_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for CLM</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoesn_grch38">PyrhoESN_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for ESN</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhofin_grch38">PyrhoFIN_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for FIN</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhogbr_grch38">PyrhoGBR_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for GBR</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhogih_grch38">PyrhoGIH_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for GIH</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhogwd_grch38">PyrhoGWD_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for GWD</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoibs_grch38">PyrhoIBS_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for IBS</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoitu_grch38">PyrhoITU_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for ITU</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhojpt_grch38">PyrhoJPT_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for JPT</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhokhv_grch38">PyrhoKHV_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for KHV</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrholwk_grch38">PyrhoLWK_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for LWK</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhomsl_grch38">PyrhoMSL_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for MSL</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhomxl_grch38">PyrhoMXL_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for MXL</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhopel_grch38">PyrhoPEL_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for PEL</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhopjl_grch38">PyrhoPJL_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for PJL</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhopur_grch38">PyrhoPUR_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for PUR</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhostu_grch38">PyrhoSTU_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for STU</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhotsi_grch38">PyrhoTSI_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for TSI</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_genetic_maps_pyrhoyri_grch38">PyrhoYRI_GRCh38</a></p></td>
<td><p>2019</p></td>
<td><p>Pyrho population-specific map for YRI</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_homsap_genetic_maps_hapmapii_grch37">
<span id="sec_catalog_homsap_genetic_maps_hapmapii_grch37"></span><h4>HapMapII_GRCh37<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_hapmapii_grch37" title="Permalink to this headline"></a></h4>
<p>
        This genetic map is from the Phase II Hapmap project
        and based on 3.1 million genotyped SNPs
        from 270 individuals across four populations (YRI, CEU, CHB and JPT).
        Genome wide recombination rates were estimated using LDHat.
        This version of the HapMap genetic map was lifted over to GRCh37
        (and adjusted in regions where the genome assembly had rearranged)
        for use in the 1000 Genomes project. Please see the README file on
        the 1000 Genomes download site for details of these adjustments.
        ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/working/20110106_recombination_hotspots
        </p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>The International HapMap Consortium, 2007. <a class="reference external" href="https://doi.org/10.1038/nature06258">https://doi.org/10.1038/nature06258</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_decodesexaveraged_grch36">
<span id="sec_catalog_homsap_genetic_maps_decodesexaveraged_grch36"></span><h4>DeCodeSexAveraged_GRCh36<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_decodesexaveraged_grch36" title="Permalink to this headline"></a></h4>
<p>
        This genetic map is from the deCode study of recombination
        events in 15,257 parent-offspring pairs from Iceland.
        289,658 phased autosomal SNPs were used to call recombinations
        within these families, and recombination rates computed from the
        density of these events. This is the combined male and female
        (sex averaged) map. See
        https://www.decode.com/addendum/ for more details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Kong et al, 2010. <a class="reference external" href="https://doi.org/10.1038/nature09525">https://doi.org/10.1038/nature09525</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhoacb_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhoacb_grch38"></span><h4>PyrhoACB_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhoacb_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the ACB
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhoasw_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhoasw_grch38"></span><h4>PyrhoASW_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhoasw_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the ASW
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhobeb_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhobeb_grch38"></span><h4>PyrhoBEB_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhobeb_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the BEB
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhocdx_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhocdx_grch38"></span><h4>PyrhoCDX_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhocdx_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the CDX
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhoceu_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhoceu_grch38"></span><h4>PyrhoCEU_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhoceu_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the CEU
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhochb_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhochb_grch38"></span><h4>PyrhoCHB_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhochb_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the CHB
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhochs_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhochs_grch38"></span><h4>PyrhoCHS_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhochs_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the CHS
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhoclm_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhoclm_grch38"></span><h4>PyrhoCLM_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhoclm_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the CLM
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhoesn_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhoesn_grch38"></span><h4>PyrhoESN_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhoesn_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the ESN
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhofin_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhofin_grch38"></span><h4>PyrhoFIN_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhofin_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the FIN
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhogbr_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhogbr_grch38"></span><h4>PyrhoGBR_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhogbr_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the GBR
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhogih_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhogih_grch38"></span><h4>PyrhoGIH_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhogih_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the GIH
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhogwd_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhogwd_grch38"></span><h4>PyrhoGWD_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhogwd_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the GWD
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhoibs_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhoibs_grch38"></span><h4>PyrhoIBS_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhoibs_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the IBS
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhoitu_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhoitu_grch38"></span><h4>PyrhoITU_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhoitu_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the ITU
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhojpt_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhojpt_grch38"></span><h4>PyrhoJPT_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhojpt_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the JPT
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhokhv_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhokhv_grch38"></span><h4>PyrhoKHV_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhokhv_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the KHV
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrholwk_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrholwk_grch38"></span><h4>PyrhoLWK_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrholwk_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the LWK
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhomsl_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhomsl_grch38"></span><h4>PyrhoMSL_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhomsl_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the MSL
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhomxl_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhomxl_grch38"></span><h4>PyrhoMXL_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhomxl_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the MXL
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhopel_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhopel_grch38"></span><h4>PyrhoPEL_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhopel_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the PEL
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhopjl_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhopjl_grch38"></span><h4>PyrhoPJL_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhopjl_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the PJL
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhopur_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhopur_grch38"></span><h4>PyrhoPUR_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhopur_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the PUR
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhostu_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhostu_grch38"></span><h4>PyrhoSTU_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhostu_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the STU
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhotsi_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhotsi_grch38"></span><h4>PyrhoTSI_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhotsi_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the TSI
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_genetic_maps_pyrhoyri_grch38">
<span id="sec_catalog_homsap_genetic_maps_pyrhoyri_grch38"></span><h4>PyrhoYRI_GRCh38<a class="headerlink" href="#sec_catalog_homsap_genetic_maps_pyrhoyri_grch38" title="Permalink to this headline"></a></h4>
<p>
        This genetic map was inferred using individuals from the YRI
        population from Phase 3 of the 1000 Genomes Project. Rates
        were estimated using pyrho (https://github.com/popgenmethods/pyrho)
        while using population-specific population size history estimates
        obtained from smc++ (https://github.com/popgenmethods/smcpp).
        Genetic maps are only available for the 22 autosomes.
        See https://doi.org/10.1126/sciadv.aaw9206 for more
        details.</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Spence and Song, 2019. <a class="reference external" href="https://doi.org/10.1126/sciadv.aaw9206">https://doi.org/10.1126/sciadv.aaw9206</a></p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_HomSap_models">
<h3>Demographic Models<a class="headerlink" href="#sec_catalog_HomSap_models" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_outofafricaextendedneandertaladmixturepulse_3i21">OutOfAfricaExtendedNeandertalAdmixturePulse_3I21</a></p></td>
<td><p>Three population out-of-Africa with an extended pulse of         Neandertal admixture into Europeans</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_outofafrica_3g09">OutOfAfrica_3G09</a></p></td>
<td><p>Three population out-of-Africa</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_outofafrica_2t12">OutOfAfrica_2T12</a></p></td>
<td><p>Two population out-of-Africa</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_africa_1t12">Africa_1T12</a></p></td>
<td><p>African population</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_americanadmixture_4b11">AmericanAdmixture_4B11</a></p></td>
<td><p>American admixture</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_outofafricaarchaicadmixture_5r19">OutOfAfricaArchaicAdmixture_5R19</a></p></td>
<td><p>Three population out-of-Africa with archaic admixture</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_zigzag_1s14">Zigzag_1S14</a></p></td>
<td><p>Periodic growth and decline.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_ancienteurasia_9k19">AncientEurasia_9K19</a></p></td>
<td><p>Multi-population model of ancient Eurasia</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_papuansoutofafrica_10j19">PapuansOutOfAfrica_10J19</a></p></td>
<td><p>Out-of-Africa with archaic admixture into Papuans</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_ashksub_7g19">AshkSub_7G19</a></p></td>
<td><p>Ashkenazi Jewish with substructure and European admixture</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_outofafrica_4j17">OutOfAfrica_4J17</a></p></td>
<td><p>4 population out of Africa</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_africa_1b08">Africa_1B08</a></p></td>
<td><p>African-americans population</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_models_ancienteurope_4a21">AncientEurope_4A21</a></p></td>
<td><p>Multi-population model of ancient Europe</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_homsap_models_outofafricaextendedneandertaladmixturepulse_3i21">
<span id="sec_catalog_homsap_models_outofafricaextendedneandertaladmixturepulse_3i21"></span><h4>Three population out-of-Africa with an extended pulse of         Neandertal admixture into Europeans<a class="headerlink" href="#sec_catalog_homsap_models_outofafricaextendedneandertaladmixturepulse_3i21" title="Permalink to this headline"></a></h4>
<p>
        Demographic model of an extended admixture pulse from Neandertals into
        Europeans taken from Iasi et al. (2021), specifically the simple model
        of Supplementary Figure 1a with a gamma-shaped pulse.
        This model simulates 3 populations: Africans, Europeans and Neandertals
        with an Out-of-Africa event. The population sizes are constant
        with an unidirectional admixture from Neandertals into Europeans after
        the split between Europeans and Africans.
        The admixture event is modelled as an 800 generation (20 ky) long
        extended admixture pulse.
        </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>OutOfAfricaExtendedNeandertalAdmixturePulse_3I21</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Three population out-of-Africa with an extended pulse of         Neandertal admixture into Europeans</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>3</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>YRI</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes YRI (Yoruba)</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>CEU</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CEU (Utah Residents (CEPH) with Northern and Western European Ancestry</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>NEA</p></td>
<td><p>0</p></td>
<td><p>Neandertals</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Iasi et al., 2021. <a class="reference external" href="https://doi.org/10.1093/molbev/msab210">https://doi.org/10.1093/molbev/msab210</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>10000</p></td>
<td><p>YRI pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>10000</p></td>
<td><p>CEU pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>10000</p></td>
<td><p>Neandertal pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Total migration rate </p></td>
<td><p>0.029</p></td>
<td><p>Neandertal-CEU total migration rate over the extended admixture pulse</p></td>
</tr>
<tr class="row-even"><td><p>Time (kya)</p></td>
<td><p>290</p></td>
<td><p>Neandertal Human split</p></td>
</tr>
<tr class="row-odd"><td><p>Time (kya)</p></td>
<td><p>73.95</p></td>
<td><p>Time of OOA event</p></td>
</tr>
<tr class="row-even"><td><p>Time (kya)</p></td>
<td><p>50</p></td>
<td><p>Neandertal migrations starts</p></td>
</tr>
<tr class="row-odd"><td><p>Time (kya)</p></td>
<td><p>30</p></td>
<td><p>Neandertal migrations end</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>29</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>2e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_outofafricaextendedneandertaladmixturepulse_3i21.png" src="_images/sec_catalog_homsap_models_outofafricaextendedneandertaladmixturepulse_3i21.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_outofafrica_3g09">
<span id="sec_catalog_homsap_models_outofafrica_3g09"></span><h4>Three population out-of-Africa<a class="headerlink" href="#sec_catalog_homsap_models_outofafrica_3g09" title="Permalink to this headline"></a></h4>
<p>
        The three population Out-of-Africa model from Gutenkunst et al. 2009.
        It describes the ancestral human population in Africa, the out of Africa
        event, and the subsequent European-Asian population split.
        Model parameters are the maximum likelihood values of the
        various parameters given in Table 1 of Gutenkunst et al.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>OutOfAfrica_3G09</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Three population out-of-Africa</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>3</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>YRI</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes YRI (Yoruba)</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>CEU</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CEU (Utah Residents (CEPH) with Northern and Western European Ancestry</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>CHB</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CHB (Han Chinese in Beijing, China)</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Gutenkunst et al., 2009. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1000695">https://doi.org/10.1371/journal.pgen.1000695</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>7,300</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>12,300</p></td>
<td><p>YRI pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>2,100</p></td>
<td><p>OOA pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>1,000</p></td>
<td><p>CEU pop. size after EU/AS divergence</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>510</p></td>
<td><p>CHB pop. size after EU/AS divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.004</p></td>
<td><p>CEU pop. growth rate (per gen.)</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.0055</p></td>
<td><p>CHB pop. growth rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-5)</p></td>
<td><p>25</p></td>
<td><p>YRI-OOA migration rate (per gen.)</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-5)</p></td>
<td><p>3</p></td>
<td><p>YRI-CEU migration rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-5)</p></td>
<td><p>1.9</p></td>
<td><p>YRI-CHB migration rate (per gen.)</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-5)</p></td>
<td><p>9.6</p></td>
<td><p>CEU-CHB migration rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>8,800</p></td>
<td><p>Expansion time of ancestral pop.</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>5,600</p></td>
<td><p>Time of OOA event</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>848</p></td>
<td><p>Time of CEU-CHB split</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>25</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>2.35e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_outofafrica_3g09.png" src="_images/sec_catalog_homsap_models_outofafrica_3g09.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_outofafrica_2t12">
<span id="sec_catalog_homsap_models_outofafrica_2t12"></span><h4>Two population out-of-Africa<a class="headerlink" href="#sec_catalog_homsap_models_outofafrica_2t12" title="Permalink to this headline"></a></h4>
<p>
        The model is derived from the Tennessen et al. analysis of the
        jSFS from European Americans and African Americans.
        It describes the ancestral human population in Africa, the out of Africa event,
        and two distinct periods of subsequent European population growth over the past
        23kya. Model parameters are taken from Fig. S5 in Fu et al.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>OutOfAfrica_2T12</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Two population out-of-Africa</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>2</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>AFR</p></td>
<td><p>0</p></td>
<td><p>African Americans</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>EUR</p></td>
<td><p>0</p></td>
<td><p>European Americans</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Tennessen et al., 2012. <a class="reference external" href="https://doi.org/10.1126/science.1219240">https://doi.org/10.1126/science.1219240</a></p></li>
<li><p>Fu et al., 2013. <a class="reference external" href="https://doi.org/10.1038/nature11690">https://doi.org/10.1038/nature11690</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>7,310</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>14,474</p></td>
<td><p>AFR pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>1,861</p></td>
<td><p>OOA pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>1,032</p></td>
<td><p>EU pop. size after EU/AS divergence</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>9,279</p></td>
<td><p>EU pop. size after 1st expansion</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>501,436</p></td>
<td><p>EU pop. size after 2nd expansion</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>432,125</p></td>
<td><p>AFR pop. size after 1st expansion</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.00307</p></td>
<td><p>EU pop. growth rate 1st expansion</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.0195</p></td>
<td><p>EU pop. growth rate 2st expansion</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.0166</p></td>
<td><p>AFR pop. growth rate 1st expansion</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-5 per gen.)</p></td>
<td><p>15</p></td>
<td><p>AFR-OOA migration rate</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-5 per gen.)</p></td>
<td><p>2.5</p></td>
<td><p>AFR-EU migration rate (both expansions)</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>5,920</p></td>
<td><p>Expansion time of ancestral pop.</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>2,040</p></td>
<td><p>Time of OOA event</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>920</p></td>
<td><p>Beginning of 1st EU growth period</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>204.6</p></td>
<td><p>Beginning of 2nd EU/1st AFR growth period</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>25</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>2.36e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_outofafrica_2t12.png" src="_images/sec_catalog_homsap_models_outofafrica_2t12.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_africa_1t12">
<span id="sec_catalog_homsap_models_africa_1t12"></span><h4>African population<a class="headerlink" href="#sec_catalog_homsap_models_africa_1t12" title="Permalink to this headline"></a></h4>
<p>
        The model is a simplification of the two population Tennesen et al.
        model with the European-American population removed so that we are
        modeling the African population in isolation.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>Africa_1T12</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>African population</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>AFR</p></td>
<td><p>0</p></td>
<td><p>African</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Tennessen et al., 2012. <a class="reference external" href="https://doi.org/10.1126/science.1219240">https://doi.org/10.1126/science.1219240</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>7,310</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>14,474</p></td>
<td><p>AFR pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>432,125</p></td>
<td><p>AFR pop. size after 1st expansion</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.0166</p></td>
<td><p>AFR pop. growth rate 1st expansion</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>5,920</p></td>
<td><p>Expansion time of ancestral pop.</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>204.6</p></td>
<td><p>Beginning of AFR growth period</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>25</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>2.36e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_africa_1t12.png" src="_images/sec_catalog_homsap_models_africa_1t12.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_americanadmixture_4b11">
<span id="sec_catalog_homsap_models_americanadmixture_4b11"></span><h4>American admixture<a class="headerlink" href="#sec_catalog_homsap_models_americanadmixture_4b11" title="Permalink to this headline"></a></h4>
<p>
        Demographic model for American admixture, taken from Browning et al. 2011.
        This model extends the Gravel et al. (2011) model of African/European/Asian
        demographic history to simulate an admixed population with admixture
        occurring 12 generations ago. The admixed population had an initial size
        of 30,000 and grew at a rate of 5% per generation, with 1/6 of the
        population of African ancestry, 1/3 European, and 1/2 Asian. Note that this
        demographic model was not inferred, and the mutation rate that Browning et al.
        used for simulation is smaller than used for inferring the model,
        so the mutation rate provided here is that from Gravel et al.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>AmericanAdmixture_4B11</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>American admixture</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>4</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>AFR</p></td>
<td><p>0</p></td>
<td><p>Contemporary African population</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>EUR</p></td>
<td><p>0</p></td>
<td><p>Contemporary European population</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>ASIA</p></td>
<td><p>0</p></td>
<td><p>Contemporary Asian population</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>ADMIX</p></td>
<td><p>0</p></td>
<td><p>Modern admixed population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Browning et al., 2018. <a class="reference external" href="http://dx.doi.org/10.1371/journal.pgen.1007385">http://dx.doi.org/10.1371/journal.pgen.1007385</a></p></li>
<li><p>Gravel et al., 2011. <a class="reference external" href="https://doi.org/10.1073/pnas.1019276108">https://doi.org/10.1073/pnas.1019276108</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>7,310</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>14,474</p></td>
<td><p>AFR pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>1,861</p></td>
<td><p>OOA pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>1,032</p></td>
<td><p>EU pop. size after EU/AS divergence</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>554</p></td>
<td><p>ASN pop. size after EU/AS divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>30,000</p></td>
<td><p>Initial ADMIX pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.0038</p></td>
<td><p>EU pop. growth rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.0048</p></td>
<td><p>ASN pop. growth rate (per gen.)</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.05</p></td>
<td><p>ADMIX pop. growth rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-5)</p></td>
<td><p>15</p></td>
<td><p>AFR-OOA migration rate (per gen.)</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-5)</p></td>
<td><p>2.5</p></td>
<td><p>AFR-EU migration rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-5)</p></td>
<td><p>0.78</p></td>
<td><p>AFR-ASN migration rate (per gen.)</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-5)</p></td>
<td><p>3.11</p></td>
<td><p>EU-ASN migration rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>5,920</p></td>
<td><p>Expansion time of ancestral pop.</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>2,040</p></td>
<td><p>Time of OOA event</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>920</p></td>
<td><p>Time of EU-ASN split</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>12</p></td>
<td><p>Time of ADMIX population emergence</p></td>
</tr>
<tr class="row-odd"><td><p>ADMIX percentage</p></td>
<td><p> 1/6</p></td>
<td><p>Amount African admixture</p></td>
</tr>
<tr class="row-even"><td><p>ADMIX percentage</p></td>
<td><p> 1/3</p></td>
<td><p>Amount European admixture</p></td>
</tr>
<tr class="row-odd"><td><p>ADMIX percentage</p></td>
<td><p> 1/2</p></td>
<td><p>Amount Asian admixture</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>25</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>2.36e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_americanadmixture_4b11.png" src="_images/sec_catalog_homsap_models_americanadmixture_4b11.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_outofafricaarchaicadmixture_5r19">
<span id="sec_catalog_homsap_models_outofafricaarchaicadmixture_5r19"></span><h4>Three population out-of-Africa with archaic admixture<a class="headerlink" href="#sec_catalog_homsap_models_outofafricaarchaicadmixture_5r19" title="Permalink to this headline"></a></h4>
<p>
        The three population out-of-African model popularized by Gutenkunst et al. (2009)
        and augmented by archaic contributions to both Eurasian and African populations.
        Two archaic populations split early in human history, before the African
        expansion, and contribute to Eurasian populations (putative Neanderthal branch)
        and to the African branch (a deep diverging branch within Africa). Admixture
        is modeled as symmetric migration between the archaic and modern human branches,
        with contribution ending at a given time in the past.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>OutOfAfricaArchaicAdmixture_5R19</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Three population out-of-Africa with archaic admixture</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>5</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>YRI</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes YRI (Yoruba)</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>CEU</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CEU (Utah Residents (CEPH) with Northern and Western European Ancestry</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>CHB</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CHB (Han Chinese in Beijing, China)</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>Neanderthal</p></td>
<td><p>None</p></td>
<td><p>Putative Neanderthals</p></td>
</tr>
<tr class="row-even"><td><p>4</p></td>
<td><p>ArchaicAFR</p></td>
<td><p>None</p></td>
<td><p>Putative Archaic Africans</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Ragsdale and Gravel, 2019. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1008204">https://doi.org/10.1371/journal.pgen.1008204</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>3,600</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>13,900</p></td>
<td><p>YRI pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>880</p></td>
<td><p>OOA pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>2,300</p></td>
<td><p>CEU pop. size after EU/AS divergence</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>650</p></td>
<td><p>CHB pop. size after EU/AS divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.00125</p></td>
<td><p>CEU pop. growth rate (per gen.)</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.00372</p></td>
<td><p>CHB pop. growth rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-5)</p></td>
<td><p>52.2</p></td>
<td><p>YRI-OOA migration rate (per gen.)</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-5)</p></td>
<td><p>2.48</p></td>
<td><p>YRI-CEU migration rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-5)</p></td>
<td><p>0</p></td>
<td><p>YRI-CHB migration rate (per gen.)</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-5)</p></td>
<td><p>11.3</p></td>
<td><p>CEU-CHB migration rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Time (kya)</p></td>
<td><p>300</p></td>
<td><p>Expansion time of ancestral pop.</p></td>
</tr>
<tr class="row-even"><td><p>Time (kya)</p></td>
<td><p>60.7</p></td>
<td><p>Time of OOA event</p></td>
</tr>
<tr class="row-odd"><td><p>Time (kya)</p></td>
<td><p>36</p></td>
<td><p>Time of CEU-CHB split</p></td>
</tr>
<tr class="row-even"><td><p>Time (kya)</p></td>
<td><p>499</p></td>
<td><p>Archaic African split time</p></td>
</tr>
<tr class="row-odd"><td><p>Time (kya)</p></td>
<td><p>125</p></td>
<td><p>Archaic African migration begins</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-5)</p></td>
<td><p>1.98</p></td>
<td><p>Arch Afr-Nean migration rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Time (kya)</p></td>
<td><p>559</p></td>
<td><p>Neanderthal split time</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-5)</p></td>
<td><p>0.825</p></td>
<td><p>OOA pops-Nean migration rate (per gen.)</p></td>
</tr>
<tr class="row-odd"><td><p>Time (kya)</p></td>
<td><p>18.7</p></td>
<td><p>Archaic migrations end</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>29</p></td>
<td><p>Generation time</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_outofafricaarchaicadmixture_5r19.png" src="_images/sec_catalog_homsap_models_outofafricaarchaicadmixture_5r19.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_zigzag_1s14">
<span id="sec_catalog_homsap_models_zigzag_1s14"></span><h4>Periodic growth and decline.<a class="headerlink" href="#sec_catalog_homsap_models_zigzag_1s14" title="Permalink to this headline"></a></h4>
<p>
        A validation model used by Schiffels and Durbin (2014) and Terhorst and
        Terhorst, Kamm, and Song (2017) with periods of exponential growth and
        decline in a single population.
        </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>Zigzag_1S14</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Periodic growth and decline.</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>generic</p></td>
<td><p>0</p></td>
<td><p>Generic expanding and contracting population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Schiffels and Durbin, 2014. <a class="reference external" href="https://doi.org/10.1038/ng.3015">https://doi.org/10.1038/ng.3015</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>7,156</p></td>
<td><p>Ancient pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>71,560</p></td>
<td><p>Recent pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (per gen.)</p></td>
<td><p>8.99448 x 10^(-5)</p></td>
<td><p>Growth rate for 1st growth</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (per gen.)</p></td>
<td><p>-0.00035977</p></td>
<td><p>Growth rate for 1st decline</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.0014391</p></td>
<td><p>Growth rate for 2nd growth</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (per gen.)</p></td>
<td><p>-0.005756</p></td>
<td><p>Growth rate for 2rd decline</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (per gen.)</p></td>
<td><p>0.023025</p></td>
<td><p>Growth rate for 3st growth</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>34,133.31</p></td>
<td><p>Beginning of 1st growth</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>8,533.33</p></td>
<td><p>Beginning of 1st decline</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>2,133.33</p></td>
<td><p>Beginning of 2nd growth</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>533.33</p></td>
<td><p>Beginning of 2nd decline</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>133.33</p></td>
<td><p>Beginning of 3rd growth</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>33.333</p></td>
<td><p>End of 3rd growth</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time</p></td>
<td><p>30</p></td>
<td><p>Generation time in years</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_zigzag_1s14.png" src="_images/sec_catalog_homsap_models_zigzag_1s14.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_ancienteurasia_9k19">
<span id="sec_catalog_homsap_models_ancienteurasia_9k19"></span><h4>Multi-population model of ancient Eurasia<a class="headerlink" href="#sec_catalog_homsap_models_ancienteurasia_9k19" title="Permalink to this headline"></a></h4>
<p>
        This is the best-fitting model of a history of
        multiple ancient and present-day human populations
        sampled across Eurasia over the past 120,000 years.
        The fitting was performed using momi2 (Kamm et al. 2019),
        which uses the multi-population site-frequency spectrum
        as input data. The model includes a ghost admixture event
        (from unsampled basal Eurasians into early European
        farmers), and two admixture events where the source is
        approximately well-known (from Neanderthals into
        Non-Africans and from Western European hunter-gatherers
        into modern Sardinians. There are three present-day
        populations: Sardinians, Han Chinese and African Mbuti.
        Additionally, there are several ancient samples
        obtained from fossils dated at different times in
        the past: the Altai Neanderthal (Prufer et al. 2014),
        a Mesolithic hunter-gatherer (Lazaridis et al. 2014),
        a Neolithic early European sample (Lazaridis et al. 2014),
        and two Palaeolithic modern humans from Siberia - MA1
        (Raghavan et al. 2014) and  Ust’Ishim (Fu et al. 2014).
        All the ancient samples are represented by a single diploid
        genome.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>AncientEurasia_9K19</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Multi-population model of ancient Eurasia</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>9</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>Mbuti</p></td>
<td><p>0</p></td>
<td><p>Present-day African Mbuti</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>LBK</p></td>
<td><p>320</p></td>
<td><p>Early European farmer (EEF)</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>Sardinian</p></td>
<td><p>0</p></td>
<td><p>Present-day Sardinian</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>Loschbour</p></td>
<td><p>300</p></td>
<td><p>Western hunter-gatherer (WHG)</p></td>
</tr>
<tr class="row-even"><td><p>4</p></td>
<td><p>MA1</p></td>
<td><p>960</p></td>
<td><p>Upper Palaeolithic MAl’ta culture</p></td>
</tr>
<tr class="row-odd"><td><p>5</p></td>
<td><p>Han</p></td>
<td><p>0</p></td>
<td><p>Present-day Han Chinese</p></td>
</tr>
<tr class="row-even"><td><p>6</p></td>
<td><p>UstIshim</p></td>
<td><p>1800</p></td>
<td><p>early Palaeolithic Ust’-Ishim</p></td>
</tr>
<tr class="row-odd"><td><p>7</p></td>
<td><p>Neanderthal</p></td>
<td><p>2000</p></td>
<td><p>Altai Neanderthal from Siberia</p></td>
</tr>
<tr class="row-even"><td><p>8</p></td>
<td><p>BasalEurasian</p></td>
<td><p>None</p></td>
<td><p>Basal Eurasians</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Kamm et al., 2019. <a class="reference external" href="https://doi.org/10.1080/01621459.2019.1635482">https://doi.org/10.1080/01621459.2019.1635482</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>17,300</p></td>
<td><p>Mbuti pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>75.7</p></td>
<td><p>EEF pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>15,000</p></td>
<td><p>Sardinian pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>1,920</p></td>
<td><p>Size of WHG, Bazal, Mal’ta and Ust’-Ishim populations</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>6,300</p></td>
<td><p>Han Chinese pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>86.9</p></td>
<td><p>Neanderthal pop. size after exp. decline</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>2,340</p></td>
<td><p>(WHG + Han chinese) pop. size before divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>29,100</p></td>
<td><p>(WHG + Mbuti) pop. size before divergence</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>18,200</p></td>
<td><p>Ancestral pop. size and Nean. size before exp. decline</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>12,000</p></td>
<td><p>(Sardinian + EEF) pop. size before divergence</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>696,000</p></td>
<td><p>Time of WHG and Neanderthal split</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>95,800</p></td>
<td><p>Time of WHG and Mbuti split, start of Nean. decline</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>79,800</p></td>
<td><p>Time of WHG and Bazal split</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>51,500</p></td>
<td><p>Time of WHG and Ust’Ishim split</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>50,400</p></td>
<td><p>Time of WHG and Han Chinese split</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>44,900</p></td>
<td><p>Time of WHG and Mal’ta split</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>37,700</p></td>
<td><p>Time of WHG and EEF split</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>7,690</p></td>
<td><p>Time of EEF and Sardinian split</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>56,800</p></td>
<td><p>Time of Neanderthal to Eurasian (WHG) admixture</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>33,700</p></td>
<td><p>Time of Bazal to EEF admixture</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>1,230</p></td>
<td><p>Time of WHG to Sardinian admixture</p></td>
</tr>
<tr class="row-odd"><td><p>ADMIX percentage</p></td>
<td><p>2.96</p></td>
<td><p>Amount of Neanderthal admixture in Eurasian pop.</p></td>
</tr>
<tr class="row-even"><td><p>ADMIX percentage</p></td>
<td><p>9.36</p></td>
<td><p>Amount of Bazal admixture in EEF pop.</p></td>
</tr>
<tr class="row-odd"><td><p>ADMIX percentage</p></td>
<td><p>3.17</p></td>
<td><p>Amount of WHG admixture in Sardinian pop.</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>8,000</p></td>
<td><p>Time of EEF samples</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>7,500</p></td>
<td><p>Time of WHG samples</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>24,000</p></td>
<td><p>Time of MAl’ta samples</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>45,000</p></td>
<td><p>Time of Ust’-Ishim samples</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>50,000</p></td>
<td><p>Time of Neanderthal samples</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time (yrs.)</p></td>
<td><p>25</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-even"><td><p>Mutation rate</p></td>
<td><p>1.22e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_ancienteurasia_9k19.png" src="_images/sec_catalog_homsap_models_ancienteurasia_9k19.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_papuansoutofafrica_10j19">
<span id="sec_catalog_homsap_models_papuansoutofafrica_10j19"></span><h4>Out-of-Africa with archaic admixture into Papuans<a class="headerlink" href="#sec_catalog_homsap_models_papuansoutofafrica_10j19" title="Permalink to this headline"></a></h4>
<p>
        A ten population model of out-of-Africa, including two pulses of
        Denisovan admixture into Papuans, and several pulses of Neandertal
        admixture into non-Africans.
        Most parameters are from Jacobs et al. (2019), Table S5 and Figure S5.
        This model is an extension of one from Malaspinas et al. (2016), thus
        some parameters are inherited from there.
        </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>PapuansOutOfAfrica_10J19</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Out-of-Africa with archaic admixture into Papuans</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>10</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>YRI</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes YRI (Yoruba)</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>CEU</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CEU (Utah Residents (CEPH) with Northern and Western European Ancestry</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>CHB</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CHB (Han Chinese in Beijing, China)</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>Papuan</p></td>
<td><p>0</p></td>
<td><p>Papuans from Indonesia and New Guinea</p></td>
</tr>
<tr class="row-even"><td><p>4</p></td>
<td><p>DenA</p></td>
<td><p>2058</p></td>
<td><p>Altai Denisovan (sampling) lineage</p></td>
</tr>
<tr class="row-odd"><td><p>5</p></td>
<td><p>NeaA</p></td>
<td><p>2612</p></td>
<td><p>Altai Neandertal (sampling) lineage</p></td>
</tr>
<tr class="row-even"><td><p>6</p></td>
<td><p>Den1</p></td>
<td><p>None</p></td>
<td><p>Denisovan D1 (introgressing) lineage</p></td>
</tr>
<tr class="row-odd"><td><p>7</p></td>
<td><p>Den2</p></td>
<td><p>None</p></td>
<td><p>Denisovan D2 (introgressing) lineage</p></td>
</tr>
<tr class="row-even"><td><p>8</p></td>
<td><p>Nea1</p></td>
<td><p>None</p></td>
<td><p>Neandertal N1 (introgressing) lineage</p></td>
</tr>
<tr class="row-odd"><td><p>9</p></td>
<td><p>Ghost</p></td>
<td><p>None</p></td>
<td><p>Out-of-Africa lineage</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Jacobs et al., 2019. <a class="reference external" href="https://doi.org/10.1016/j.cell.2019.02.035">https://doi.org/10.1016/j.cell.2019.02.035</a></p></li>
<li><p>Malaspinas et al., 2016. <a class="reference external" href="https://doi.org/10.1038/nature18299">https://doi.org/10.1038/nature18299</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>48,433</p></td>
<td><p>African pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>6,962</p></td>
<td><p>European pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>9,025</p></td>
<td><p>East Asian pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>8,834</p></td>
<td><p>Papuan pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>5,083</p></td>
<td><p>Altai Denisovan pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>826</p></td>
<td><p>Altai Neandertal pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>13,249</p></td>
<td><p>Introgressing Denisovan D1 pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>13,249</p></td>
<td><p>Introgressing Denisovan D2 pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>13,249</p></td>
<td><p>Introgressing Neandertal pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>8,516</p></td>
<td><p>Ghost (out-of-Africa lineage) pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>12,971</p></td>
<td><p>(European + East Asian) pop. size before divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>41,563</p></td>
<td><p>(Ghost + African) pop. size before divergence</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>13,249</p></td>
<td><p>(Denisovan + Neandertal) pop. size before divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>32,671</p></td>
<td><p>(Human + Archaic) pop. size before divergence</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>100</p></td>
<td><p>(Altai Denisovan + Introgressing Denisovan D1) pop. size before divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>100</p></td>
<td><p>(Altai Denisovan + Introgressing Denisovan D2) pop. size before divergence</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>13,249</p></td>
<td><p>(Altai Neandertal + Introgressing Neandertal) pop. size before divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>37,497</p></td>
<td><p>Time of European and East Asian split</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>50,982</p></td>
<td><p>Time of Ghost and (European + East Asian) split</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>51,736</p></td>
<td><p>Time of Ghost and Papuan split</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>64,322</p></td>
<td><p>Time of Ghost and African split</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>97,875</p></td>
<td><p>Time of Altai Neandertal and Introgressing Neandertal split</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>282,750</p></td>
<td><p>Time of Altai Denisovan and Introgressing Denisovan D1 split</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>362,500</p></td>
<td><p>Time of Altai Denisovan and Introgressing Denisovan D2 split</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>437,610</p></td>
<td><p>Time of Denisovan and Neandertal split</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>586,525</p></td>
<td><p>Time of Human and Archaic split</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>2,231</p></td>
<td><p>(European + East Asian) bottleneck pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>243</p></td>
<td><p>Papuan bottleneck pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>1,394</p></td>
<td><p>Ghost (out-of-Africa) bottleneck pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>48,111</p></td>
<td><p>Time of the (European + East Asian) bottleneck</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>48,865</p></td>
<td><p>Time of the Papuan bottleneck</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>61,451</p></td>
<td><p>Time of the Ghost (out-of-Africa) bottleneck</p></td>
</tr>
<tr class="row-even"><td><p>ADMIX percentage</p></td>
<td><p>2.2</p></td>
<td><p>Amount of Denisovan D1 admixture in Papaun pop.</p></td>
</tr>
<tr class="row-odd"><td><p>ADMIX percentage</p></td>
<td><p>1.8</p></td>
<td><p>Amount of Denisovan D2 admixture in Papaun pop.</p></td>
</tr>
<tr class="row-even"><td><p>ADMIX percentage</p></td>
<td><p>2.4</p></td>
<td><p>Amount of Neandertal admixture in Ghost pop.</p></td>
</tr>
<tr class="row-odd"><td><p>ADMIX percentage</p></td>
<td><p>1.1</p></td>
<td><p>Amount of Neandertal admixture in (European + East Asian) pop.</p></td>
</tr>
<tr class="row-even"><td><p>ADMIX percentage</p></td>
<td><p>0.2</p></td>
<td><p>Amount of Neandertal admixture in Papuan pop.</p></td>
</tr>
<tr class="row-odd"><td><p>ADMIX percentage</p></td>
<td><p>0.2</p></td>
<td><p>Amount of Neandertal admixture in East Asian pop.</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>29,800</p></td>
<td><p>Time of Denisovan D1 to Papuan admixture</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>45,700</p></td>
<td><p>Time of Denisovan D2 to Papuan admixture</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>53,737</p></td>
<td><p>Time of Neandertal to Ghost admixture</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>45,414</p></td>
<td><p>Time of Neandertal to (European + East Asian) admixture</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>40,948</p></td>
<td><p>Time of Neandertal to Papuan admixture</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>25,607</p></td>
<td><p>Time of Neandertal to East Asian admixture</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-4)</p></td>
<td><p>1.79</p></td>
<td><p>Ghost–African migration rate</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-4)</p></td>
<td><p>4.42</p></td>
<td><p>Ghost–European migration rate</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-4)</p></td>
<td><p>31.4</p></td>
<td><p>European–East Asian migration rate</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-4)</p></td>
<td><p>57.2</p></td>
<td><p>East Asian–Papuan migration rate</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (x10^-4)</p></td>
<td><p>5.72</p></td>
<td><p>(European + East Asian)–Papuan migration rate</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (x10^-4)</p></td>
<td><p>4.42</p></td>
<td><p>(European + East Asian)–Ghost migration rate</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>37,497</p></td>
<td><p>(European + East Asian)–Papuan migration begins</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>37,497</p></td>
<td><p>(European + East Asian)–Ghost migration begins</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>29</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>1.4e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_papuansoutofafrica_10j19.png" src="_images/sec_catalog_homsap_models_papuansoutofafrica_10j19.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_ashksub_7g19">
<span id="sec_catalog_homsap_models_ashksub_7g19"></span><h4>Ashkenazi Jewish with substructure and European admixture<a class="headerlink" href="#sec_catalog_homsap_models_ashksub_7g19" title="Permalink to this headline"></a></h4>
<p>
    This was the best fit model of Ashkenazi Jewish demographic history from
    Gladstein and Hammer 2019, shown in Figure 1, labeled “Substructure Model”.
    Model choice and parameter estimation were performed with Approximate
    Bayesian Computation. Parameter values are based on the mode from ABC found
    in Table S3 of Gladstein and Hammer 2019. In this model, the ancestors of
    Europeans and Middle Eastern populations diverge. Non-Ashkenazi Jewish
    populations then diverge from the Middle Eastern population. The Ashkenazi
    Jews then diverge from the other Jewish populations and experience a
    substantial reduction in population size and a single pulse of gene flow
    from Europeans (corresponding to their arrival in Europe). After the gene
    flow from Europeans to the Ashkenazi Jews, the Ashkenazi Jews split into
    two groups, the Western and Eastern. Finally, the Western Ashkenazi Jews
    experience moderate instantaneous population size increase, and the
    Eastern experience a massive population size increase. In addition to the
    demographic model Gladstein and Hammer 2019 also incorporated an SNP array
    ascertainment scheme into the simulation. This demographic model does not
    include the SNP array ascertainment scheme. It should be noted that
    Gladstein and Hammer 2019 simulated with a mutation rate of 2.5e-8.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>AshkSub_7G19</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Ashkenazi Jewish with substructure and European admixture</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>7</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>YRI</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes YRI (Yoruba)</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>CHB</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CHB (Han Chinese in Beijing, China)</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>CEU</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CEU (Utah Residents (CEPH) with Northern and Western European Ancestry</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>ME</p></td>
<td><p>0</p></td>
<td><p>Middle Eastern</p></td>
</tr>
<tr class="row-even"><td><p>4</p></td>
<td><p>J</p></td>
<td><p>0</p></td>
<td><p>non-Ashkenazi Jewish</p></td>
</tr>
<tr class="row-odd"><td><p>5</p></td>
<td><p>WAJ</p></td>
<td><p>0</p></td>
<td><p>Western Ashkenazi Jewish</p></td>
</tr>
<tr class="row-even"><td><p>6</p></td>
<td><p>EAJ</p></td>
<td><p>0</p></td>
<td><p>Eastern Ashkenazi Jewish</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Gladstein and Hammer, 2019. <a class="reference external" href="https://doi.org/10.1093/molbev/msz047">https://doi.org/10.1093/molbev/msz047</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>7,300</p></td>
<td><p>Ancestral African pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>18,197</p></td>
<td><p>YRI pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>4,073</p></td>
<td><p>CHB pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>33,113</p></td>
<td><p>CEU pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>436,515</p></td>
<td><p>Middle Eastern pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>354,813</p></td>
<td><p>non-AJ Jewish pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>6,606</p></td>
<td><p> Western AJ pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>1,949,844</p></td>
<td><p> Eastern AJ founder pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>8,800</p></td>
<td><p>Expansion time of YRI ancestral pop.</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>2,105</p></td>
<td><p>Time of OOA event</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>850</p></td>
<td><p>Time of CEU-CHB split</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>481</p></td>
<td><p>Time of Middle Eastern-CEU split</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>211</p></td>
<td><p>Time of Jewish-Middle Eastern split</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>29</p></td>
<td><p>Time of AJ-Jewish split</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>28</p></td>
<td><p>Time of geneflow (not inferred)</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>14</p></td>
<td><p>Time of Eastern-Western AJ split</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>13</p></td>
<td><p>Time of AJ growth (not inferred)</p></td>
</tr>
<tr class="row-odd"><td><p>ADMIX percentage</p></td>
<td><p> 0.17</p></td>
<td><p> European to AJ gene flow</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>25</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>2.5e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_ashksub_7g19.png" src="_images/sec_catalog_homsap_models_ashksub_7g19.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_outofafrica_4j17">
<span id="sec_catalog_homsap_models_outofafrica_4j17"></span><h4>4 population out of Africa<a class="headerlink" href="#sec_catalog_homsap_models_outofafrica_4j17" title="Permalink to this headline"></a></h4>
<p>
        Demographic model for a four population out-of-Africa history,
        taken from Jouganous et al. (2017). Parameter values were taken
        from table 4 in the main text. This model was fit based on joint
        allele frequecy spectrum (AFS) data from 1000 Genomes exomes
        from the YRI, CEU, CHB, and JPT poulation samples. The demography
        follows the previous three-populations out-of-Africa models
        with an additional population split in Asia leading to the
        Japanese (JPT) population. Parameter values were estimated with
        the program Moments assuming a mutation rate of 1.44e-8 and a
        generation time of 29 years.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>OutOfAfrica_4J17</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>4 population out of Africa</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>4</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>YRI</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes YRI (Yoruba)</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>CEU</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CEU (Utah Residents (CEPH) with Northern and Western European Ancestry</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>CHB</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes CHB (Han Chinese in Beijing, China)</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>JPT</p></td>
<td><p>0</p></td>
<td><p>1000 Genomes JPT (Japanese in Tokyo, Japan)</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Jouganous et al., 2017. <a class="reference external" href="https://doi.org/10.1534/genetics.117.200493">https://doi.org/10.1534/genetics.117.200493</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size (individuals)</p></td>
<td><p>11293</p></td>
<td><p>A (ancestral) population size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size (individuals)</p></td>
<td><p>23721</p></td>
<td><p>YRI population size</p></td>
</tr>
<tr class="row-even"><td><p>Population size (individuals)</p></td>
<td><p>2831</p></td>
<td><p>B (OOA) population size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size (individuals)</p></td>
<td><p>2512</p></td>
<td><p>CEU (European) initial pop. size after EU/AS divergence</p></td>
</tr>
<tr class="row-even"><td><p>Population size (individuals)</p></td>
<td><p>1019</p></td>
<td><p>CHB (Asian) initial pop. size after EU/AS divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Population size (individuals)</p></td>
<td><p>4384</p></td>
<td><p>JPT (Japanese ) pop. size after split from CHB</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (percent per generation)</p></td>
<td><p>0</p></td>
<td><p>A growth rate</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (percent per generation)</p></td>
<td><p>0</p></td>
<td><p>YRI growth rate</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (percent per generation)</p></td>
<td><p>0</p></td>
<td><p>B growth rate</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (percent per generation)</p></td>
<td><p>0.16</p></td>
<td><p>CEU growth rate</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (percent per generation)</p></td>
<td><p>0.26</p></td>
<td><p>CHB growth rate</p></td>
</tr>
<tr class="row-odd"><td><p>Growth rate (percent per generation)</p></td>
<td><p>1.29</p></td>
<td><p>JPT growth rate</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (fraction per generation)</p></td>
<td><p>16.8e-5</p></td>
<td><p>YRI-B migration rate</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (fraction per generation)</p></td>
<td><p>1.14e-5</p></td>
<td><p>YRI-CEU migration rate</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (fraction per generation)</p></td>
<td><p>0.56e-5</p></td>
<td><p>YRI-CHB migration rate</p></td>
</tr>
<tr class="row-odd"><td><p>Migration rate (fraction per generation)</p></td>
<td><p>4.75e-5</p></td>
<td><p>CEU-CHB migration rate</p></td>
</tr>
<tr class="row-even"><td><p>Migration rate (fraction per generation)</p></td>
<td><p>3.3e-5</p></td>
<td><p>CHB-JPT migration rate</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (thousands of years ago)</p></td>
<td><p>357</p></td>
<td><p>Expansion time of ancestral population</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (thousands of years ago)</p></td>
<td><p>119</p></td>
<td><p>Time of OOA event</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (thousands of years ago)</p></td>
<td><p>46</p></td>
<td><p>Time of CEU-CHB split</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (thousands of years ago)</p></td>
<td><p>9</p></td>
<td><p>Time of CHB-JPT split</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time (years)</p></td>
<td><p>29</p></td>
<td><p>Years per generation</p></td>
</tr>
<tr class="row-even"><td><p>Mutation rate</p></td>
<td><p>1.44e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_outofafrica_4j17.png" src="_images/sec_catalog_homsap_models_outofafrica_4j17.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_africa_1b08">
<span id="sec_catalog_homsap_models_africa_1b08"></span><h4>African-americans population<a class="headerlink" href="#sec_catalog_homsap_models_africa_1b08" title="Permalink to this headline"></a></h4>
<p>
        African-American two-epoch instantaneous growth model from Boyko et al
        2008, fit to the synonymous SFS for the 11 of 15 African Americans showing
        the least European ancestry, using coalescent simulations with
        recombination with the maximum likelihood method of Williamson et al 2005;
        times were calibrated assuming 3e5 generations since human-chimp divergence
        and fitting the number of synonymous human-chimp differences. Mutation and
        recombination rates were assumed to be the same (1.8e-8).
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>Africa_1B08</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>African-americans population</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>African_Americans</p></td>
<td><p>0</p></td>
<td><p>African-Americans from Boyko et al 2008</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Boyko et al., 2008. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1000083">https://doi.org/10.1371/journal.pgen.1000083</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>7,778</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>25,636</p></td>
<td><p>African-americans current pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Epoch Time (gen.)</p></td>
<td><p>6,809</p></td>
<td><p>instantaneous expansion time of ancestral pop.</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>1.80E-08</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_africa_1b08.png" src="_images/sec_catalog_homsap_models_africa_1b08.png" />
<hr class="docutils" />
<section id="sec_catalog_homsap_models_ancienteurope_4a21">
<span id="sec_catalog_homsap_models_ancienteurope_4a21"></span><h4>Multi-population model of ancient Europe<a class="headerlink" href="#sec_catalog_homsap_models_ancienteurope_4a21" title="Permalink to this headline"></a></h4>
<p>
    Population structure that has existed over the last 45,000 years in Europe, leading
    to modern Europeans. The model demonstrates the divergence of a Basal European
    Lineages into four ancient populations; Western, Eastern and Caucasus Hunter-
    Gatherers and Anatolian Farmers. Migration of Anatolian farmers into Western Europe
    and admixture with Western Hunter-Gatherers produces the European Neolithic Farmers.
    In West Asia the admixture of Eastern Hunter-Gatherers and Caucasus Hunter-
    Gatherers leads to the formation of the Yamnaya Steppe population. The Yamnaya
    migrate into Western Europe to admixture with the Neolithic farmers giving rise to
    Bronze Age europeans. There is only an exponential growth in population size from
    then to the Present-day. Samples are taken at multiple point throughout history
    from each population.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>AncientEurope_4A21</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Multi-population model of ancient Europe</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>10</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>OOA</p></td>
<td><p>1500</p></td>
<td><p>Basal/OOA</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>NE</p></td>
<td><p>600</p></td>
<td><p>Northern European</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>WA</p></td>
<td><p>800</p></td>
<td><p>West Asian</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>CHG</p></td>
<td><p>300</p></td>
<td><p>Caucasus Hunter-gathers</p></td>
</tr>
<tr class="row-even"><td><p>4</p></td>
<td><p>ANA</p></td>
<td><p>260</p></td>
<td><p>Anatolian</p></td>
</tr>
<tr class="row-odd"><td><p>5</p></td>
<td><p>WHG</p></td>
<td><p>250</p></td>
<td><p>Western Hunter-gathers</p></td>
</tr>
<tr class="row-even"><td><p>6</p></td>
<td><p>EHG</p></td>
<td><p>250</p></td>
<td><p>Eastern Hunter-gathers</p></td>
</tr>
<tr class="row-odd"><td><p>7</p></td>
<td><p>YAM</p></td>
<td><p>160</p></td>
<td><p>Yamnaya</p></td>
</tr>
<tr class="row-even"><td><p>8</p></td>
<td><p>NEO</p></td>
<td><p>180</p></td>
<td><p>Neolithic</p></td>
</tr>
<tr class="row-odd"><td><p>9</p></td>
<td><p>Bronze</p></td>
<td><p>135</p></td>
<td><p>Bronze Age</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Allentoft et al., 2022. <a class="reference external" href="https://doi.org/10.1101/2022.05.04.490594">https://doi.org/10.1101/2022.05.04.490594</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>50,000</p></td>
<td><p>Bronze Age pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>5,000</p></td>
<td><p>Yamnaya pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>10,000</p></td>
<td><p>Western Hunter-Gatherer pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>10,000</p></td>
<td><p>Eastern Hunter-Gatherer pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>50,000</p></td>
<td><p>Neolithic Farmer pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>10,000</p></td>
<td><p>Caucasus Hunter-Gatherer pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>5,000</p></td>
<td><p>Northern European pop. size (ancestor of WHG and EHG)</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>5,000</p></td>
<td><p>West Asian pop. size (ancestor of Anatolian Farmers and CHG)</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>140</p></td>
<td><p>Time of Yamnaya and Neolithic Farmer admixture</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>180</p></td>
<td><p>Time of EHG and CHG admixture</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>200</p></td>
<td><p>Time of Anatolian Farmer and WHG admixture</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>600</p></td>
<td><p>Time of WHG and EHG divergence</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>800</p></td>
<td><p>Time of Anatolian Farmers and CHG divergence</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>1500</p></td>
<td><p>Time of Basal European split</p></td>
</tr>
<tr class="row-even"><td><p>Growth rate (per gen.)</p></td>
<td><p>6.7</p></td>
<td><p>Growth rate from Bronze Age to present day</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>0</p></td>
<td><p>Time of present day samples</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>135</p></td>
<td><p>Time of Bronze Age samples</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>160</p></td>
<td><p>Time of Yamnaya samples</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>180</p></td>
<td><p>Time of Neolithic Farmer samples</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>250</p></td>
<td><p>Time of Western Hunter-Gatherer samples</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>250</p></td>
<td><p>Time of Eastern Hunter-Gatherer samples</p></td>
</tr>
<tr class="row-odd"><td><p>Time (gen.)</p></td>
<td><p>260</p></td>
<td><p>Time of Anatolian Farmer samples</p></td>
</tr>
<tr class="row-even"><td><p>Time (gen.)</p></td>
<td><p>300</p></td>
<td><p>Time of Caucasus Hunter-Gatherer samples</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time (yrs.)</p></td>
<td><p>29</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-even"><td><p>Mutation rate</p></td>
<td><p>1.25e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_homsap_models_ancienteurope_4a21.png" src="_images/sec_catalog_homsap_models_ancienteurope_4a21.png" />
</section>
<section id="sec_catalog_HomSap_annotations">
<h3>Annotations<a class="headerlink" href="#sec_catalog_HomSap_annotations" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_annotations_ensembl_havana_104_exons">ensembl_havana_104_exons</a></p></td>
<td><p>2018</p></td>
<td><p>Ensembl Havana exon annotations on GRCh38</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_annotations_ensembl_havana_104_cds">ensembl_havana_104_CDS</a></p></td>
<td><p>2018</p></td>
<td><p>Ensembl Havana CDS annotations on GRCh38</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_homsap_annotations_ensembl_havana_104_exons">
<span id="sec_catalog_homsap_annotations_ensembl_havana_104_exons"></span><h4>ensembl_havana_104_exons<a class="headerlink" href="#sec_catalog_homsap_annotations_ensembl_havana_104_exons" title="Permalink to this headline"></a></h4>
<p>Ensembl Havana exon annotations on GRCh38</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Hunt et al, 2018. <a class="reference external" href="https://doi.org/10.1093/database/bay119">https://doi.org/10.1093/database/bay119</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_annotations_ensembl_havana_104_cds">
<span id="sec_catalog_homsap_annotations_ensembl_havana_104_cds"></span><h4>ensembl_havana_104_CDS<a class="headerlink" href="#sec_catalog_homsap_annotations_ensembl_havana_104_cds" title="Permalink to this headline"></a></h4>
<p>Ensembl Havana CDS annotations on GRCh38</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Hunt et al, 2018. <a class="reference external" href="https://doi.org/10.1093/database/bay119">https://doi.org/10.1093/database/bay119</a></p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_HomSap_dfe">
<h3>Distribution of Fitness Effects (DFEs)<a class="headerlink" href="#sec_catalog_HomSap_dfe" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_homsap_dfes_gamma_k17">Gamma_K17</a></p></td>
<td><p>2017</p></td>
<td><p>Deleterious Gamma DFE</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_homsap_dfes_gamma_h17">Gamma_H17</a></p></td>
<td><p>2017</p></td>
<td><p>Deleterious Gamma DFE</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_homsap_dfes_gamma_k17">
<span id="sec_catalog_homsap_dfes_gamma_k17"></span><h4>Gamma_K17<a class="headerlink" href="#sec_catalog_homsap_dfes_gamma_k17" title="Permalink to this headline"></a></h4>
<p>Deleterious Gamma DFE</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Kim et al., 2017. <a class="reference external" href="https://doi.org/10.1534/genetics.116.197145">https://doi.org/10.1534/genetics.116.197145</a></p></li>
</ul>
</section>
<section id="sec_catalog_homsap_dfes_gamma_h17">
<span id="sec_catalog_homsap_dfes_gamma_h17"></span><h4>Gamma_H17<a class="headerlink" href="#sec_catalog_homsap_dfes_gamma_h17" title="Permalink to this headline"></a></h4>
<p>Deleterious Gamma DFE</p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Huber et al., 2017. <a class="reference external" href="https://doi.org/10.1073/pnas.1619508114">https://doi.org/10.1073/pnas.1619508114</a></p></li>
</ul>
</section>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_PapAnu">
<span id="sec_catalog_PapAnu"></span><h2>Papio anubis<a class="headerlink" href="#sec_catalog_PapAnu" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>PapAnu</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Papio anubis</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Olive baboon</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>11 (<a class="reference external" href="https://doi.org/10.1371/journal.pbio.3000838">Wu et. al., 2020</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>335505 (<a class="reference external" href="https://doi.org/10.1093/gbe/evac040">Wall et. al., 2022</a>)</p>
</dd>
</dl>
<section id="sec_catalog_PapAnu_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_PapAnu_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>218172882</p></td>
<td><p>9.92638e-09</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>193660750</p></td>
<td><p>9.60544e-09</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>184919515</p></td>
<td><p>9.02238e-09</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>182120902</p></td>
<td><p>9.82513e-09</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>173900761</p></td>
<td><p>9.5798e-09</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>167138247</p></td>
<td><p>1.04979e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>161768468</p></td>
<td><p>1.11888e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>140274886</p></td>
<td><p>1.10899e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>127591819</p></td>
<td><p>1.13288e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>126462689</p></td>
<td><p>1.17532e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>125913696</p></td>
<td><p>1.18403e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>123343450</p></td>
<td><p>1.0824e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p>106849001</p></td>
<td><p>1.24677e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p>106654974</p></td>
<td><p>1.27419e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p>91985775</p></td>
<td><p>1.26084e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>91184193</p></td>
<td><p>1.47616e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p>74525926</p></td>
<td><p>1.5241e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>18</p></td>
<td><p>72894408</p></td>
<td><p>1.36841e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>19</p></td>
<td><p>72123344</p></td>
<td><p>1.30374e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>20</p></td>
<td><p>50021108</p></td>
<td><p>1.6772e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-even"><td><p>X</p></td>
<td><p>142711496</p></td>
<td><p>1.18898e-08</p></td>
<td><p>5.7e-09</p></td>
</tr>
<tr class="row-odd"><td><p>Y</p></td>
<td><p>8309886</p></td>
<td><p>0</p></td>
<td><p>5.7e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_PapAnu_models">
<h3>Demographic Models<a class="headerlink" href="#sec_catalog_PapAnu_models" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_papanu_models_singlepopsmcpp_1w22">SinglePopSMCpp_1W22</a></p></td>
<td><p>SMC++ estimates of N(t) for Papio Anubis individuals</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_papanu_models_singlepopsmcpp_1w22">
<span id="sec_catalog_papanu_models_singlepopsmcpp_1w22"></span><h4>SMC++ estimates of N(t) for Papio Anubis individuals<a class="headerlink" href="#sec_catalog_papanu_models_singlepopsmcpp_1w22" title="Permalink to this headline"></a></h4>
<p>
        These estimates were obtained from a sample of Papio Anubis
        individuals from the colony housed at the Southwest National
        Primate Research Center (SNPRC). SMC++ was run with a subset of
        36 individuals from the population.
        </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>SinglePopSMCpp_1W22</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>SMC++ estimates of N(t) for Papio Anubis individuals</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>PAnubis_SNPRC</p></td>
<td><p>0</p></td>
<td><p>Papio Anubis population from SNPRC</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Wall et. al., 2022. <a class="reference external" href="https://doi.org/10.1093/gbe/evac040">https://doi.org/10.1093/gbe/evac040</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>93362</p></td>
<td><p>Ancestral population size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>55968</p></td>
<td><p>Pop. size during 1st time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>72998</p></td>
<td><p>Pop. size during 2nd time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>30714</p></td>
<td><p>Pop. size during 3rd time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>41841</p></td>
<td><p>Pop. size during 4th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>51822</p></td>
<td><p>Pop. size during 5th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>120758</p></td>
<td><p>Pop. size during 6th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>335505</p></td>
<td><p>Pop. size during 7th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>2046585</p></td>
<td><p>Begining of 1st time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>1219140</p></td>
<td><p>Begining of 2nd time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>782017</p></td>
<td><p>Begining of 3rd time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>432614</p></td>
<td><p>Begining of 4th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>165305</p></td>
<td><p>Begining of 5th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Time (yrs.)</p></td>
<td><p>5101</p></td>
<td><p>Begining of 6th time interval</p></td>
</tr>
<tr class="row-even"><td><p>Time (yrs.)</p></td>
<td><p>2434</p></td>
<td><p>Begining of 7th time interval</p></td>
</tr>
<tr class="row-odd"><td><p>Generation time (yrs.)</p></td>
<td><p>11</p></td>
<td><p>Average generation interval</p></td>
</tr>
<tr class="row-even"><td><p>Mutation rate</p></td>
<td><p>5.7e-9</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_papanu_models_singlepopsmcpp_1w22.png" src="_images/sec_catalog_papanu_models_singlepopsmcpp_1w22.png" />
</section>
</section>
<section id="sec_catalog_PonAbe">
<span id="sec_catalog_PonAbe"></span><h2>Pongo abelii<a class="headerlink" href="#sec_catalog_PonAbe" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>PonAbe</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Pongo abelii</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Sumatran orangutan</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>25 (<a class="reference external" href="https://doi.org/10.1093/acprof:oso/9780199213276.003.0005">Wich et al., 2008</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>17900.0 (<a class="reference external" href="http://doi.org/10.1038/nature09687">Locke et al., 2011</a>)</p>
</dd>
</dl>
<section id="sec_catalog_PonAbe_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_PonAbe_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>227913704</p></td>
<td><p>5.72097e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>2A</p></td>
<td><p>109511694</p></td>
<td><p>6.74694e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>2B</p></td>
<td><p>129937803</p></td>
<td><p>6.08833e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>193656255</p></td>
<td><p>5.77615e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>4</p></td>
<td><p>189387572</p></td>
<td><p>6.06176e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>5</p></td>
<td><p>179185813</p></td>
<td><p>5.87839e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>6</p></td>
<td><p>169501136</p></td>
<td><p>5.89605e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>7</p></td>
<td><p>145408105</p></td>
<td><p>6.82979e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>8</p></td>
<td><p>144036388</p></td>
<td><p>6.34313e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>9</p></td>
<td><p>112206110</p></td>
<td><p>7.04756e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>10</p></td>
<td><p>132178492</p></td>
<td><p>6.94812e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>11</p></td>
<td><p>128122151</p></td>
<td><p>5.79185e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>12</p></td>
<td><p>132184051</p></td>
<td><p>6.04316e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>13</p></td>
<td><p>98475126</p></td>
<td><p>6.40817e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>14</p></td>
<td><p>88963417</p></td>
<td><p>6.11075e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>15</p></td>
<td><p>82547911</p></td>
<td><p>6.6315e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>16</p></td>
<td><p>68237989</p></td>
<td><p>6.76066e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>17</p></td>
<td><p>75914007</p></td>
<td><p>8.01983e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>18</p></td>
<td><p>75923960</p></td>
<td><p>6.40065e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>19</p></td>
<td><p>57575784</p></td>
<td><p>9.09647e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>20</p></td>
<td><p>60841859</p></td>
<td><p>6.31424e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>21</p></td>
<td><p>34683425</p></td>
<td><p>7.58867e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>22</p></td>
<td><p>35308119</p></td>
<td><p>9.83628e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-odd"><td><p>X</p></td>
<td><p>151242693</p></td>
<td><p>6.40288e-09</p></td>
<td><p>1.5e-08</p></td>
</tr>
<tr class="row-even"><td><p>MT</p></td>
<td><p>16499</p></td>
<td><p>0</p></td>
<td><p>1.5e-08</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
<hr class="docutils" />
<section id="sec_catalog_PonAbe_genetic_maps">
<h3>Genetic Maps<a class="headerlink" href="#sec_catalog_PonAbe_genetic_maps" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Year</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_ponabe_genetic_maps_naterpa_ponabe3">NaterPA_PonAbe3</a></p></td>
<td><p>2017</p></td>
<td><p>From Nater et al. (2017) for Pongo abelii</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#sec_catalog_ponabe_genetic_maps_naterpp_ponabe3">NaterPP_PonAbe3</a></p></td>
<td><p>2017</p></td>
<td><p>From Nater et al. (2017) for Pongo pygmaeus</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_ponabe_genetic_maps_naterpa_ponabe3">
<span id="sec_catalog_ponabe_genetic_maps_naterpa_ponabe3"></span><h4>NaterPA_PonAbe3<a class="headerlink" href="#sec_catalog_ponabe_genetic_maps_naterpa_ponabe3" title="Permalink to this headline"></a></h4>
<p>
        This genetic map is from the Nater et al. (2017) study, inferred using
        LDhat from n=15 whole-genome sequenced Sumatran orangutan individuals.
        See https://doi.org/10.1016/j.cub.2017.09.047 for more details.
        Lifted over from assembly PonAbe2 (as used in Nater et al.) to PonAbe3.
        </p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Nater et al., 2017. <a class="reference external" href="https://doi.org/10.1016/j.cub.2017.09.047">https://doi.org/10.1016/j.cub.2017.09.047</a></p></li>
</ul>
</section>
<section id="sec_catalog_ponabe_genetic_maps_naterpp_ponabe3">
<span id="sec_catalog_ponabe_genetic_maps_naterpp_ponabe3"></span><h4>NaterPP_PonAbe3<a class="headerlink" href="#sec_catalog_ponabe_genetic_maps_naterpp_ponabe3" title="Permalink to this headline"></a></h4>
<p>
        This genetic map is from the Nater et al. (2017) study, inferred using
        LDhat from n=20 whole-genome sequenced Bornean orangutan individuals.
        See https://doi.org/10.1016/j.cub.2017.09.047 for more details.
        Lifted over from assembly PonAbe2 (as used in Nater et al.) to PonAbe3.
        </p>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Nater et al., 2017. <a class="reference external" href="https://doi.org/10.1016/j.cub.2017.09.047">https://doi.org/10.1016/j.cub.2017.09.047</a></p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="sec_catalog_PonAbe_models">
<h3>Demographic Models<a class="headerlink" href="#sec_catalog_PonAbe_models" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#sec_catalog_ponabe_models_twospecies_2l11">TwoSpecies_2L11</a></p></td>
<td><p>Two population orangutan model</p></td>
</tr>
</tbody>
</table>
<section id="sec_catalog_ponabe_models_twospecies_2l11">
<span id="sec_catalog_ponabe_models_twospecies_2l11"></span><h4>Two population orangutan model<a class="headerlink" href="#sec_catalog_ponabe_models_twospecies_2l11" title="Permalink to this headline"></a></h4>
<p>
        The two orang-utan species, Sumatran (Pongo abelii) and Bornean (Pongo
        pygmaeus) inferred from the joint-site frequency spectrum with ten
        individuals from each population. This model is an isolation-with-
        migration model, with exponential growth or decay in each population
        after the split. The Sumatran population grows in size, while the
        Bornean population slightly declines.
    </p>
<p class="rubric">Details</p>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>TwoSpecies_2L11</p>
</dd>
<dt class="field-even">Description</dt>
<dd class="field-even"><p>Two population orangutan model</p>
</dd>
<dt class="field-odd">Num populations</dt>
<dd class="field-odd"><p>2</p>
</dd>
</dl>
<p class="rubric">Populations</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Index</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Sampling time</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>Bornean</p></td>
<td><p>0</p></td>
<td><p>Pongo pygmaeus (Bornean) population</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>Sumatran</p></td>
<td><p>0</p></td>
<td><p>Pongo abelii (Sumatran) population</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Citations</p>
<ul class="simple">
<li><p>Locke et al., 2011. <a class="reference external" href="http://doi.org/10.1038/nature09687">http://doi.org/10.1038/nature09687</a></p></li>
</ul>
<p class="rubric">Demographic Model parameters</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter Type (units)</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>17,934</p></td>
<td><p>Ancestral pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>10,617</p></td>
<td><p>Pongo pygmaeus pop. size at split</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>7,317</p></td>
<td><p>Pongo abelii pop. size at split</p></td>
</tr>
<tr class="row-odd"><td><p>Population size</p></td>
<td><p>8,805</p></td>
<td><p>Modern Pongo pygmaeus pop. size</p></td>
</tr>
<tr class="row-even"><td><p>Population size</p></td>
<td><p>37,661</p></td>
<td><p>Modern Pongo abelii pop. size</p></td>
</tr>
<tr class="row-odd"><td><p>Epoch Time (gen.)</p></td>
<td><p>20,157</p></td>
<td><p>Species divergence time</p></td>
</tr>
<tr class="row-even"><td><p>Generation time (yrs.)</p></td>
<td><p>20</p></td>
<td><p>Generation time</p></td>
</tr>
<tr class="row-odd"><td><p>Mutation rate</p></td>
<td><p>2e-8</p></td>
<td><p>Per-base per-generation mutation rate</p></td>
</tr>
</tbody>
</table>
</section>
<img alt="_images/sec_catalog_ponabe_models_twospecies_2l11.png" src="_images/sec_catalog_ponabe_models_twospecies_2l11.png" />
</section>
</section>
<section id="sec_catalog_StrAga">
<span id="sec_catalog_StrAga"></span><h2>Streptococcus agalactiae<a class="headerlink" href="#sec_catalog_StrAga" title="Permalink to this headline"></a></h2>
<dl class="field-list simple">
<dt class="field-odd">ID</dt>
<dd class="field-odd"><p>StrAga</p>
</dd>
<dt class="field-even">Name</dt>
<dd class="field-even"><p>Streptococcus agalactiae</p>
</dd>
<dt class="field-odd">Common name</dt>
<dd class="field-odd"><p>Group B Streptococcus</p>
</dd>
<dt class="field-even">Generation time</dt>
<dd class="field-even"><p>0.0027397260273972603 (<a class="reference external" href="https://doi.org/10.1086/284168">Savageau M.A., 1983</a>)</p>
</dd>
<dt class="field-odd">Population size</dt>
<dd class="field-odd"><p>140000 (<a class="reference external" href="https://doi.org/10.1038/ncomms5544">Da Cunha et al, 2014</a>)</p>
</dd>
</dl>
<section id="sec_catalog_StrAga_genome">
<h3>Genome<a class="headerlink" href="#sec_catalog_StrAga_genome" title="Permalink to this headline"></a></h3>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Length</p></th>
<th class="head"><p>Recombination rate</p></th>
<th class="head"><p>Mutation rate</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>2065074</p></td>
<td><p>0</p></td>
<td><p>1.53e-09</p></td>
</tr>
</tbody>
</table>
<p>Mutation and recombination rates are in units of per bp and per generation.</p>
</section>
</section>
<hr class="docutils" />
<section id="generic-models">
<h2>Generic models<a class="headerlink" href="#generic-models" title="Permalink to this headline"></a></h2>
<p>In addition to the species-specific models listed in this catalog, <code class="docutils literal notranslate"><span class="pre">stdpopsim</span></code> offers
a number of generic demographic models that can be run with any species.
These are described in more detail in the <a class="reference internal" href="api.html#sec-api-generic-models"><span class="std std-ref">API</span></a>.
Simulations using these generic models must be run via the Python interface; see our
<a class="reference internal" href="tutorial.html#sec-python-tute"><span class="std std-ref">Python tutorial</span></a> to learn how to do this.</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="api.html#stdpopsim.PiecewiseConstantSize" title="stdpopsim.PiecewiseConstantSize"><code class="xref py py-meth docutils literal notranslate"><span class="pre">stdpopsim.PiecewiseConstantSize()</span></code></a></p></li>
<li><p><a class="reference internal" href="api.html#stdpopsim.IsolationWithMigration" title="stdpopsim.IsolationWithMigration"><code class="xref py py-meth docutils literal notranslate"><span class="pre">stdpopsim.IsolationWithMigration()</span></code></a></p></li>
</ul>
</div></blockquote>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorial.html" class="btn btn-neutral float-right" title="Tutorials" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019-2020, PopSim Consortium.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>