

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Command-line options &mdash; stdpopsim 0.1.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API" href="api.html" />
    <link rel="prev" title="Tutorials" href="tutorial.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> stdpopsim
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="catalog.html">Catalog</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Command-line options</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#msprime specific parameters">msprime specific parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#SLiM specific parameters">SLiM specific parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Sub-commands:">Sub-commands:</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#StrAga">StrAga</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat1">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat1">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#AnoGam">AnoGam</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat2">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat2">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#CaeEle">CaeEle</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat3">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat3">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#DroSec">DroSec</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat4">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat4">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#CanFam">CanFam</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat5">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat5">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#DroMel">DroMel</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat6">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat6">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#AedAeg">AedAeg</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat7">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat7">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#BosTau">BosTau</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat8">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat8">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#HomSap">HomSap</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat9">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat9">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#EscCol">EscCol</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat10">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat10">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#PonAbe">PonAbe</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat11">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat11">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#AnaPla">AnaPla</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat12">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat12">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#HelMel">HelMel</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat13">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat13">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#AnoCar">AnoCar</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat14">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat14">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#AraTha">AraTha</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat15">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat15">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#ChlRei">ChlRei</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat16">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments_repeat16">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#download-genetic-maps">download-genetic-maps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Positional Arguments_repeat17">Positional Arguments</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelogs.html">Changelogs</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">stdpopsim</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Command-line options</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/cli_arguments.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="command-line-options">
<span id="sec-cli-args"></span><h1>Command-line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h1>
<p><p>Command line interface for stdpopsim.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">stdpopsim</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">V</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="o">|</span> <span class="o">-</span><span class="n">q</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">CACHE_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="p">{</span><span class="n">msprime</span><span class="p">,</span><span class="n">slim</span><span class="p">}]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">msprime</span><span class="o">-</span><span class="n">model</span> <span class="p">{</span><span class="n">hudson</span><span class="p">,</span><span class="n">dtwf</span><span class="p">,</span><span class="n">smc</span><span class="p">,</span><span class="n">smc_prime</span><span class="p">}]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">msprime</span><span class="o">-</span><span class="n">change</span><span class="o">-</span><span class="n">model</span> <span class="n">T</span> <span class="n">MODEL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">slim</span><span class="o">-</span><span class="n">path</span> <span class="n">PATH</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">slim</span><span class="o">-</span><span class="n">script</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">slim</span><span class="o">-</span><span class="n">scaling</span><span class="o">-</span><span class="n">factor</span> <span class="n">Q</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">slim</span><span class="o">-</span><span class="n">burn</span><span class="o">-</span><span class="ow">in</span> <span class="n">X</span><span class="p">]</span>
                 <span class="p">{</span><span class="n">StrAga</span><span class="p">,</span><span class="n">AnoGam</span><span class="p">,</span><span class="n">CaeEle</span><span class="p">,</span><span class="n">DroSec</span><span class="p">,</span><span class="n">CanFam</span><span class="p">,</span><span class="n">DroMel</span><span class="p">,</span><span class="n">AedAeg</span><span class="p">,</span><span class="n">BosTau</span><span class="p">,</span><span class="n">HomSap</span><span class="p">,</span><span class="n">EscCol</span><span class="p">,</span><span class="n">PonAbe</span><span class="p">,</span><span class="n">AnaPla</span><span class="p">,</span><span class="n">HelMel</span><span class="p">,</span><span class="n">AnoCar</span><span class="p">,</span><span class="n">AraTha</span><span class="p">,</span><span class="n">ChlRei</span><span class="p">,</span><span class="n">download</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span><span class="p">}</span>
                 <span class="o">...</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h2>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h2>
<dl class="option-list">
<dt><kbd>subcommand</kbd></dt>
<dd><p>Possible choices: StrAga, AnoGam, CaeEle, DroSec, CanFam, DroMel, AedAeg, BosTau, HomSap, EscCol, PonAbe, AnaPla, HelMel, AnoCar, AraTha, ChlRei, download-genetic-maps</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments">
<h2>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h2>
<dl class="option-list">
<dt><kbd>-V, --version</kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Increase logging verbosity (can use be used multiple times).</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-q, --quiet</kbd></dt>
<dd><p>Do not write any non-essential messages</p>
</dd>
<dt><kbd>-c, --cache-dir</kbd></dt>
<dd><p>Set the cache directory to the specified value. Note that this can also be set using the environment variable STDPOPSIM_CACHE. If both the environment variable and this option are set, the option takes precedence. Default: /home/runner/.cache/stdpopsim</p>
</dd>
<dt><kbd>-e, --engine</kbd></dt>
<dd><p>Possible choices: msprime, slim</p>
<p>Specify a simulation engine.</p>
<p>Default: “msprime”</p>
</dd>
</dl>
</div>
<div class="section" id="msprime specific parameters">
<h2>msprime specific parameters<a class="headerlink" href="#msprime specific parameters" title="Permalink to this headline">¶</a></h2>
<dl class="option-list">
<dt><kbd>--msprime-model</kbd></dt>
<dd><p>Possible choices: hudson, dtwf, smc, smc_prime</p>
<p>Specify the simulation model used by msprime. See msprime API documentation for details.</p>
<p>Default: “hudson”</p>
</dd>
<dt><kbd>--msprime-change-model</kbd></dt>
<dd><p>Change to the specified simulation MODEL at generation T. This option may provided multiple times.</p>
<p>Default: []</p>
</dd>
</dl>
</div>
<div class="section" id="SLiM specific parameters">
<h2>SLiM specific parameters<a class="headerlink" href="#SLiM specific parameters" title="Permalink to this headline">¶</a></h2>
<dl class="option-list">
<dt><kbd>--slim-path</kbd></dt>
<dd><p>Full path to <a href="#id1"><span class="problematic" id="id2">`</span></a>slim’ executable.</p>
</dd>
<dt><kbd>--slim-script</kbd></dt>
<dd><p>Write script to stdout and exit without running SLiM.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--slim-scaling-factor</kbd></dt>
<dd><p>Rescale model parameters by Q to speed up simulation. See SLiM manual: <cite>5.5 Rescaling population sizes to improve simulation performance</cite>. [default=1].</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--slim-burn-in</kbd></dt>
<dd><p>Length of the burn-in phase, in units of N generations [default=10].</p>
<p>Default: 10</p>
</dd>
</dl>
</div>
<div class="section" id="Sub-commands:">
<h2>Sub-commands:<a class="headerlink" href="#Sub-commands:" title="Permalink to this headline">¶</a></h2>
<div class="section" id="StrAga">
<h3>StrAga<a class="headerlink" href="#StrAga" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Streptococcus agalactiae using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Streptococcus agalactiae:
Generation time: 0.0027397260273972603
Population size: 140000
Mutation rate: 1.53e-09
Recombination rate: 0</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">StrAga</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat1">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat1" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat1">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat1" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, I, chr1</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="AnoGam">
<h3>AnoGam<a class="headerlink" href="#AnoGam" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Anopheles gambiae using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Anopheles gambiae:
Generation time: 0.09090909090909091
Population size: 6000000.0
Mutation rate: 5.49e-09
Recombination rate: 1.044e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AnoGam</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat2">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat2" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat2">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat2" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 2L, chr2L, 2R, chr2R, 3L, chr3L, 3R, chr3R, X, chrX, Mt</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 2L, 2R, 3L, 3R, X, Mt. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: GAS_1A17</p>
<p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:GAS_1A17. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="CaeEle">
<h3>CaeEle<a class="headerlink" href="#CaeEle" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Caenorhabditis elegans using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Caenorhabditis elegans:
Generation time: 0.01
Population size: 10000
Mutation rate: 1.84e-09
Recombination rate: 3.325e-11</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">CaeEle</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat3">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat3" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat3">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat3" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: I, II, III, IV, V, X, MtDNA</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: I, II, III, IV, V, X, MtDNA. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="DroSec">
<h3>DroSec<a class="headerlink" href="#DroSec" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Drosophila sechellia using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Drosophila sechellia:
Generation time: 0.05
Population size: 100000
Mutation rate: 1.5e-09
Recombination rate: 2.165e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">DroSec</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat4">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat4" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat4">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat4" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 2L, chr2L, 2R, chr2R, 3L, chr3L, 3R, chr3R, X, chrX, 4, chr4</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 2L, 2R, 3L, 3R, X, 4. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="CanFam">
<h3>CanFam<a class="headerlink" href="#CanFam" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Canis familiaris using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Canis familiaris:
Generation time: 3
Population size: 13000
Mutation rate: 4e-09
Recombination rate: 9.506e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">CanFam</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat5">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat5" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat5">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat5" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: Campbell2016_CanFam3_1</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://stdpopsim.readthedocs.io/en/latest/catalog.html">https://stdpopsim.readthedocs.io/en/latest/catalog.html</a>&gt; Available maps: Campbell2016_CanFam3_1.</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, 23, chr23, 24, chr24, 25, chr25, 26, chr26, 27, chr27, 28, chr28, 29, chr29, 30, chr30, 31, chr31, 32, chr32, 33, chr33, 34, chr34, 35, chr35, 36, chr36, 37, chr37, 38, chr38, X, chrX, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, X, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="DroMel">
<h3>DroMel<a class="headerlink" href="#DroMel" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Drosophila melanogaster using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Drosophila melanogaster:
Generation time: 0.1
Population size: 1720600
Mutation rate: 5.49e-09
Recombination rate: 2.104e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">DroMel</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat6">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat6" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat6">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat6" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: ComeronCrossover_dm6, ComeronCrossoverV2_dm6</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://stdpopsim.readthedocs.io/en/latest/catalog.html">https://stdpopsim.readthedocs.io/en/latest/catalog.html</a>&gt; Available maps: ComeronCrossover_dm6, ComeronCrossoverV2_dm6.</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 2L, chr2L, 2R, chr2R, 3L, chr3L, 3R, chr3R, 4, chr4, X, chrX, Y, chrY, mitochondrion_genome</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 2L, 2R, 3L, 3R, 4, X, Y, mitochondrion_genome. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: African3Epoch_1S16, OutOfAfrica_2L06</p>
<p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:African3Epoch_1S16, OutOfAfrica_2L06. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="AedAeg">
<h3>AedAeg<a class="headerlink" href="#AedAeg" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Aedes aegypti using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Aedes aegypti:
Generation time: 0.06666666666666667
Population size: 1000000.0
Mutation rate: 3.5e-09
Recombination rate: 2.782e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AedAeg</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat7">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat7" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat7">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat7" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="BosTau">
<h3>BosTau<a class="headerlink" href="#BosTau" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Bos taurus using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Bos taurus:
Generation time: 5
Population size: 90
Mutation rate: 1.2e-08
Recombination rate: 9.26e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">BosTau</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat8">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat8" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat8">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat8" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, 23, chr23, 24, chr24, 25, chr25, 26, chr26, 27, chr27, 28, chr28, 29, chr29, X, chrX, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, X, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: HolsteinFriesian_1M13</p>
<p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:HolsteinFriesian_1M13. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="HomSap">
<h3>HomSap<a class="headerlink" href="#HomSap" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Homo sapiens using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Homo sapiens:
Generation time: 30
Population size: 10000
Mutation rate: 1.29e-08
Recombination rate: 1.205e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">HomSap</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat9">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat9" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat9">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat9" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: HapMapII_GRCh37, DeCodeSexAveraged_GRCh36, PyrhoACB_GRCh38, PyrhoASW_GRCh38, PyrhoBEB_GRCh38, PyrhoCDX_GRCh38, PyrhoCEU_GRCh38, PyrhoCHB_GRCh38, PyrhoCHS_GRCh38, PyrhoCLM_GRCh38, PyrhoESN_GRCh38, PyrhoFIN_GRCh38, PyrhoGBR_GRCh38, PyrhoGIH_GRCh38, PyrhoGWD_GRCh38, PyrhoIBS_GRCh38, PyrhoITU_GRCh38, PyrhoJPT_GRCh38, PyrhoKHV_GRCh38, PyrhoLWK_GRCh38, PyrhoMSL_GRCh38, PyrhoMXL_GRCh38, PyrhoPEL_GRCh38, PyrhoPJL_GRCh38, PyrhoPUR_GRCh38, PyrhoSTU_GRCh38, PyrhoTSI_GRCh38, PyrhoYRI_GRCh38</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://stdpopsim.readthedocs.io/en/latest/catalog.html">https://stdpopsim.readthedocs.io/en/latest/catalog.html</a>&gt; Available maps: HapMapII_GRCh37, DeCodeSexAveraged_GRCh36, PyrhoACB_GRCh38, PyrhoASW_GRCh38, PyrhoBEB_GRCh38, PyrhoCDX_GRCh38, PyrhoCEU_GRCh38, PyrhoCHB_GRCh38, PyrhoCHS_GRCh38, PyrhoCLM_GRCh38, PyrhoESN_GRCh38, PyrhoFIN_GRCh38, PyrhoGBR_GRCh38, PyrhoGIH_GRCh38, PyrhoGWD_GRCh38, PyrhoIBS_GRCh38, PyrhoITU_GRCh38, PyrhoJPT_GRCh38, PyrhoKHV_GRCh38, PyrhoLWK_GRCh38, PyrhoMSL_GRCh38, PyrhoMXL_GRCh38, PyrhoPEL_GRCh38, PyrhoPJL_GRCh38, PyrhoPUR_GRCh38, PyrhoSTU_GRCh38, PyrhoTSI_GRCh38, PyrhoYRI_GRCh38.</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, X, chrX, Y, chrY, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, X, Y, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: OutOfAfrica_3G09, OutOfAfrica_2T12, Africa_1T12, AmericanAdmixture_4B11, OutOfAfricaArchaicAdmixture_5R19, Zigzag_1S14, AncientEurasia_9K19, PapuansOutOfAfrica_10J19, AshkSub_7G19, OutOfAfrica_4J17, Africa_1B08</p>
<p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:OutOfAfrica_3G09, OutOfAfrica_2T12, Africa_1T12, AmericanAdmixture_4B11, OutOfAfricaArchaicAdmixture_5R19, Zigzag_1S14, AncientEurasia_9K19, PapuansOutOfAfrica_10J19, AshkSub_7G19, OutOfAfrica_4J17, Africa_1B08. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="EscCol">
<h3>EscCol<a class="headerlink" href="#EscCol" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Escherichia coli using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Escherichia coli:
Generation time: 3.805175e-05
Population size: 180000000.0
Mutation rate: 8.9e-11
Recombination rate: 0</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">EscCol</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat10">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat10" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat10">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat10" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: Chromosome</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: Chromosome. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="PonAbe">
<h3>PonAbe<a class="headerlink" href="#PonAbe" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Pongo abelii using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Pongo abelii:
Generation time: 20
Population size: 17900.0
Mutation rate: 1.5e-08
Recombination rate: 5.65e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">PonAbe</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat11">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat11" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat11">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat11" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: NaterPA_PonAbe2, NaterPP_PonAbe2</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://stdpopsim.readthedocs.io/en/latest/catalog.html">https://stdpopsim.readthedocs.io/en/latest/catalog.html</a>&gt; Available maps: NaterPA_PonAbe2, NaterPP_PonAbe2.</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2a, chr2a, 2b, chr2b, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, X, chrX, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2a, 2b, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, X, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: TwoSpecies_2L11</p>
<p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:TwoSpecies_2L11. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="AnaPla">
<h3>AnaPla<a class="headerlink" href="#AnaPla" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Anas platyrhynchos using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Anas platyrhynchos:
Generation time: 4
Population size: 156000
Mutation rate: 4.83e-09
Recombination rate: 1.47e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AnaPla</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat12">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat12" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat12">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat12" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, 23, chr23, 24, chr24, 25, chr25, 26, chr26, 27, chr27, 28, chr28, 29, chr29, 30, chr30, Z, chrZ, 31, chr31, 32, chr32, 33, chr33, 34, chr34, 35, chr35, 36, chr36, 37, chr37, 38, chr38, 39, chr39, 40, chr40</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, Z, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: MallardBlackDuck_2L19</p>
<p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:MallardBlackDuck_2L19. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="HelMel">
<h3>HelMel<a class="headerlink" href="#HelMel" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Heliconius melpomene using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Heliconius melpomene:
Generation time: 0.095
Population size: 2100000.0
Mutation rate: 2.9e-09
Recombination rate: 4.601e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">HelMel</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat13">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat13" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat13">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat13" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="AnoCar">
<h3>AnoCar<a class="headerlink" href="#AnoCar" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Anolis carolinensis using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Anolis carolinensis:
Generation time: 1.5
Population size: 3050000.0
Mutation rate: 2.1e-10
Recombination rate: 1e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AnoCar</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat14">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat14" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat14">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat14" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, LGa, LGb, LGc, LGd, LGf, LGg, LGh, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, LGa, LGb, LGc, LGd, LGf, LGg, LGh, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="AraTha">
<h3>AraTha<a class="headerlink" href="#AraTha" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Arabidopsis thaliana using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Arabidopsis thaliana:
Generation time: 1.0
Population size: 10000
Mutation rate: 7e-09
Recombination rate: 8.029e-10</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AraTha</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat15">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat15" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat15">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat15" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: SalomeAveraged_TAIR7</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://stdpopsim.readthedocs.io/en/latest/catalog.html">https://stdpopsim.readthedocs.io/en/latest/catalog.html</a>&gt; Available maps: SalomeAveraged_TAIR7.</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, Mt, Pt</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, Mt, Pt. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: SouthMiddleAtlas_1D17, African2Epoch_1H18, African3Epoch_1H18</p>
<p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:SouthMiddleAtlas_1D17, African2Epoch_1H18, African3Epoch_1H18. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="ChlRei">
<h3>ChlRei<a class="headerlink" href="#ChlRei" title="Permalink to this headline">¶</a></h3>
<p>Run simulations for Chlamydomonas reinhardtii using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Chlamydomonas reinhardtii:
Generation time: 0.001141552511415525
Population size: 1.3999999999999998e-07
Mutation rate: 9.351e-10
Recombination rate: 1.459e-10</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">ChlRei</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat16">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat16" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two samples must be specified. The number of arguments that will be accepted depends on the simulation model that is specified: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments_repeat16">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments_repeat16" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length, using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single populationconstant size model is used. Available models:. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="download-genetic-maps">
<h3>download-genetic-maps<a class="headerlink" href="#download-genetic-maps" title="Permalink to this headline">¶</a></h3>
<p>Download genetic maps and store them in the cache directory. Maps are downloaded regardless of whether they are already in the cache or not. Please use the –cache-dir option to download maps to a specific directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">download</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span>
                                <span class="p">[</span><span class="n">species</span><span class="p">]</span> <span class="p">[</span><span class="n">genetic_maps</span> <span class="p">[</span><span class="n">genetic_maps</span> <span class="o">...</span><span class="p">]]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments_repeat17">
<h4>Positional Arguments<a class="headerlink" href="#Positional Arguments_repeat17" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>species</kbd></dt>
<dd><p>Download genetic maps for this species. If not specified download all known genetic maps.</p>
</dd>
<dt><kbd>genetic_maps</kbd></dt>
<dd><p>If specified, download these genetic maps. If no maps are provided, download all maps for this species.</p>
</dd>
</dl>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="api.html" class="btn btn-neutral float-right" title="API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="tutorial.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2019-2020, PopSim Consortium.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>