<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Command-line options &mdash; stdpopsim 0.2.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API" href="api.html" />
    <link rel="prev" title="Tutorials" href="tutorial.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            stdpopsim
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="catalog.html">Catalog</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Command-line options</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#positional-arguments">Positional Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#named-arguments">Named Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#msprime-specific-parameters">msprime specific parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#slim-specific-parameters">SLiM specific parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Sub-commands">Sub-commands</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#AedAeg">AedAeg</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat1">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat1">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#AnaPla">AnaPla</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat2">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat2">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#AnoCar">AnoCar</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat3">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat3">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#AnoGam">AnoGam</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat4">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat4">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#ApiMel">ApiMel</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat5">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat5">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#AraTha">AraTha</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat6">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat6">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#BosTau">BosTau</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat7">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat7">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#CaeEle">CaeEle</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat8">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat8">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#CanFam">CanFam</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat9">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat9">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#ChlRei">ChlRei</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat10">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat10">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#DroMel">DroMel</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat11">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat11">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#DroSec">DroSec</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat12">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat12">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#EscCol">EscCol</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat13">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat13">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#GasAcu">GasAcu</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat14">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat14">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#HelAnn">HelAnn</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat15">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat15">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#HelMel">HelMel</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat16">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat16">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#HomSap">HomSap</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat17">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat17">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#MusMus">MusMus</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat18">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat18">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#OrySat">OrySat</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat19">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat19">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#PanTro">PanTro</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat20">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat20">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#PapAnu">PapAnu</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat21">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat21">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#PhoSin">PhoSin</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat22">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat22">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#PonAbe">PonAbe</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat23">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat23">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#StrAga">StrAga</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat24">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments_repeat24">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#download-genetic-maps">download-genetic-maps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments_repeat25">Positional Arguments</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelogs.html">Changelogs</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">stdpopsim</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Command-line options</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/cli_arguments.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="command-line-options">
<span id="sec-cli-args"></span><h1>Command-line options<a class="headerlink" href="#command-line-options" title="Permalink to this heading"></a></h1>
<p><p>Command line interface for stdpopsim.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">stdpopsim</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">V</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="o">|</span> <span class="o">-</span><span class="n">q</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">CACHE_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="p">{</span><span class="n">msprime</span><span class="p">,</span><span class="n">slim</span><span class="p">}]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">msprime</span><span class="o">-</span><span class="n">model</span> <span class="p">{</span><span class="n">hudson</span><span class="p">,</span><span class="n">dtwf</span><span class="p">,</span><span class="n">smc</span><span class="p">,</span><span class="n">smc_prime</span><span class="p">}]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">msprime</span><span class="o">-</span><span class="n">change</span><span class="o">-</span><span class="n">model</span> <span class="n">T</span> <span class="n">MODEL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">slim</span><span class="o">-</span><span class="n">path</span> <span class="n">PATH</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">slim</span><span class="o">-</span><span class="n">script</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">slim</span><span class="o">-</span><span class="n">scaling</span><span class="o">-</span><span class="n">factor</span> <span class="n">Q</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">slim</span><span class="o">-</span><span class="n">burn</span><span class="o">-</span><span class="ow">in</span> <span class="n">X</span><span class="p">]</span>
                 <span class="p">{</span><span class="n">AedAeg</span><span class="p">,</span><span class="n">AnaPla</span><span class="p">,</span><span class="n">AnoCar</span><span class="p">,</span><span class="n">AnoGam</span><span class="p">,</span><span class="n">ApiMel</span><span class="p">,</span><span class="n">AraTha</span><span class="p">,</span><span class="n">BosTau</span><span class="p">,</span><span class="n">CaeEle</span><span class="p">,</span><span class="n">CanFam</span><span class="p">,</span><span class="n">ChlRei</span><span class="p">,</span><span class="n">DroMel</span><span class="p">,</span><span class="n">DroSec</span><span class="p">,</span><span class="n">EscCol</span><span class="p">,</span><span class="n">GasAcu</span><span class="p">,</span><span class="n">HelAnn</span><span class="p">,</span><span class="n">HelMel</span><span class="p">,</span><span class="n">HomSap</span><span class="p">,</span><span class="n">MusMus</span><span class="p">,</span><span class="n">OrySat</span><span class="p">,</span><span class="n">PanTro</span><span class="p">,</span><span class="n">PapAnu</span><span class="p">,</span><span class="n">PhoSin</span><span class="p">,</span><span class="n">PonAbe</span><span class="p">,</span><span class="n">StrAga</span><span class="p">,</span><span class="n">download</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span><span class="p">}</span>
                 <span class="o">...</span>
</pre></div>
</div>
<section id="positional-arguments">
<h2>Positional Arguments<a class="headerlink" href="#positional-arguments" title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>subcommand</kbd></dt>
<dd><p>Possible choices: AedAeg, AnaPla, AnoCar, AnoGam, ApiMel, AraTha, BosTau, CaeEle, CanFam, ChlRei, DroMel, DroSec, EscCol, GasAcu, HelAnn, HelMel, HomSap, MusMus, OrySat, PanTro, PapAnu, PhoSin, PonAbe, StrAga, download-genetic-maps</p>
</dd>
</dl>
</section>
<section id="named-arguments">
<h2>Named Arguments<a class="headerlink" href="#named-arguments" title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>-V, --version</kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Increase logging verbosity (can use be used multiple times).</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-q, --quiet</kbd></dt>
<dd><p>Do not write any non-essential messages</p>
</dd>
<dt><kbd>-c, --cache-dir</kbd></dt>
<dd><p>Set the cache directory to the specified value. Note that this can also be set using the environment variable STDPOPSIM_CACHE. If both the environment variable and this option are set, the option takes precedence. Default: /home/runner/.cache/stdpopsim</p>
</dd>
<dt><kbd>-e, --engine</kbd></dt>
<dd><p>Possible choices: msprime, slim</p>
<p>Specify a simulation engine.</p>
<p>Default: “msprime”</p>
</dd>
</dl>
</section>
<section id="msprime-specific-parameters">
<h2>msprime specific parameters<a class="headerlink" href="#msprime-specific-parameters" title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>--msprime-model</kbd></dt>
<dd><p>Possible choices: hudson, dtwf, smc, smc_prime</p>
<p>Specify the simulation model used by msprime. See msprime API documentation for details.</p>
<p>Default: “hudson”</p>
</dd>
<dt><kbd>--msprime-change-model</kbd></dt>
<dd><p>Change to the specified simulation MODEL at generation T. This option may provided multiple times.</p>
<p>Default: []</p>
</dd>
</dl>
</section>
<section id="slim-specific-parameters">
<h2>SLiM specific parameters<a class="headerlink" href="#slim-specific-parameters" title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>--slim-path</kbd></dt>
<dd><p>Full path to <a href="#id1"><span class="problematic" id="id2">`</span></a>slim’ executable.</p>
</dd>
<dt><kbd>--slim-script</kbd></dt>
<dd><p>Write script to stdout and exit without running SLiM.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--slim-scaling-factor</kbd></dt>
<dd><p>Rescale model parameters by Q to speed up simulation. See SLiM manual: <cite>5.5 Rescaling population sizes to improve simulation performance</cite>. [default=1].</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--slim-burn-in</kbd></dt>
<dd><p>Length of the burn-in phase, in units of N generations [default=10].</p>
<p>Default: 10</p>
</dd>
</dl>
</section>
<section id="Sub-commands">
<h2>Sub-commands<a class="headerlink" href="#Sub-commands" title="Permalink to this heading"></a></h2>
<section id="AedAeg">
<h3>AedAeg<a class="headerlink" href="#AedAeg" title="Permalink to this heading"></a></h3>
<p>Run simulations for Aedes aegypti using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Aedes aegypti:
Generation time: 0.06666666666666667
Population size: 1000000.0
Mutation rate: 3.5e-09
Recombination rate: 2.782e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AedAeg</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat1">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat1" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat1">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat1" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="AnaPla">
<h3>AnaPla<a class="headerlink" href="#AnaPla" title="Permalink to this heading"></a></h3>
<p>Run simulations for Anas platyrhynchos using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Anas platyrhynchos:
Generation time: 4
Population size: 156000
Mutation rate: 4.83e-09
Recombination rate: 1.433e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AnaPla</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat2">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat2" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat2">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat2" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, 23, chr23, 24, chr24, 25, chr25, 26, chr26, 27, chr27, 28, chr28, 29, chr29, 30, chr30, Z, chrZ, 31, chr31, 32, chr32, 33, chr33, 34, chr34, 35, chr35, 36, chr36, 37, chr37, 38, chr38, 39, chr39, 40, chr40</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, Z, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: MallardBlackDuck_2L19</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: MallardBlackDuck_2L19. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="AnoCar">
<h3>AnoCar<a class="headerlink" href="#AnoCar" title="Permalink to this heading"></a></h3>
<p>Run simulations for Anolis carolinensis using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Anolis carolinensis:
Generation time: 1.5
Population size: 3050000.0
Mutation rate: 2.1e-10
Recombination rate: 1e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AnoCar</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat3">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat3" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat3">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat3" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, LGa, LGb, LGc, LGd, LGf, LGg, LGh, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, LGa, LGb, LGc, LGd, LGf, LGg, LGh, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="AnoGam">
<h3>AnoGam<a class="headerlink" href="#AnoGam" title="Permalink to this heading"></a></h3>
<p>Run simulations for Anopheles gambiae using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Anopheles gambiae:
Generation time: 0.09090909090909091
Population size: 1000000.0
Mutation rate: 3.5e-09
Recombination rate: 1.447e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AnoGam</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat4">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat4" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat4">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat4" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 2L, chr2L, 2R, chr2R, 3L, chr3L, 3R, chr3R, X, chrX, Mt</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 2L, 2R, 3L, 3R, X, Mt. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: GabonAg1000G_1A17</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: GabonAg1000G_1A17. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="ApiMel">
<h3>ApiMel<a class="headerlink" href="#ApiMel" title="Permalink to this heading"></a></h3>
<p>Run simulations for Apis mellifera using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Apis mellifera:
Generation time: 2
Population size: 200000.0
Mutation rate: 3.4e-09
Recombination rate: 2.318e-07</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">ApiMel</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat5">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat5" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat5">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat5" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: CM009931.2, NC_037638.1, 1, CM009932.2, NC_037639.1, 2, CM009933.2, NC_037640.1, 3, CM009934.2, NC_037641.1, 4, CM009935.2, NC_037642.1, 5, CM009936.2, NC_037643.1, 6, CM009937.2, NC_037644.1, 7, CM009938.2, NC_037645.1, 8, CM009939.2, NC_037646.1, 9, CM009940.2, NC_037647.1, 10, CM009941.2, NC_037648.1, 11, CM009942.2, NC_037649.1, 12, CM009943.2, NC_037650.1, 13, CM009944.2, NC_037651.1, 14, CM009945.2, NC_037652.1, 15, CM009946.2, NC_037653.1, 16, CM009947.2, NC_001566.1, MT</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: CM009931.2, CM009932.2, CM009933.2, CM009934.2, CM009935.2, CM009936.2, CM009937.2, CM009938.2, CM009939.2, CM009940.2, CM009941.2, CM009942.2, CM009943.2, CM009944.2, CM009945.2, CM009946.2, CM009947.2. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="AraTha">
<h3>AraTha<a class="headerlink" href="#AraTha" title="Permalink to this heading"></a></h3>
<p>Run simulations for Arabidopsis thaliana using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Arabidopsis thaliana:
Generation time: 1.0
Population size: 10000
Mutation rate: 7e-09
Recombination rate: 8.029e-10</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">AraTha</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">dfes</span> <span class="p">[</span><span class="n">HELP_DFES</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">annotations</span> <span class="p">[</span><span class="n">HELP_ANNOTATIONS</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">annotation</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat6">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat6" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat6">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat6" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: SalomeAveraged_TAIR10</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html">https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html</a>&gt; Available maps: SalomeAveraged_TAIR10.</p>
</dd>
<dt><kbd>--help-dfes</kbd></dt>
<dd><p>Print list of DFEs and exit. If a DFE ID is given as an argument, show help for this DFE. Otherwise show help for all available DFEs</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, Mt, Pt</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, Mt, Pt. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: SouthMiddleAtlas_1D17, African2Epoch_1H18, African3Epoch_1H18</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: SouthMiddleAtlas_1D17, African2Epoch_1H18, African3Epoch_1H18. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Possible choices: Gamma_H18</p>
<p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:Gamma_H18. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>--help-annotations</kbd></dt>
<dd><p>Print list of annotations and exit. If a annotation ID is given as an argument, show help for this annotation. Otherwise show help for all available annotations</p>
</dd>
<dt><kbd>--dfe-annotation</kbd></dt>
<dd><p>Possible choices: araport_11_exons, araport_11_CDS</p>
<p>Specify the intervals over which a given DFE acts using a genomic annotation track Available maps: araport_11_exons, araport_11_CDS.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="BosTau">
<h3>BosTau<a class="headerlink" href="#BosTau" title="Permalink to this heading"></a></h3>
<p>Run simulations for Bos taurus using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Bos taurus:
Generation time: 5
Population size: 62000
Mutation rate: 1.2e-08
Recombination rate: 9.26e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">BosTau</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat7">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat7" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat7">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat7" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, 23, chr23, 24, chr24, 25, chr25, 26, chr26, 27, chr27, 28, chr28, 29, chr29, X, chrX, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, X, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: HolsteinFriesian_1M13</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: HolsteinFriesian_1M13. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="CaeEle">
<h3>CaeEle<a class="headerlink" href="#CaeEle" title="Permalink to this heading"></a></h3>
<p>Run simulations for Caenorhabditis elegans using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Caenorhabditis elegans:
Generation time: 0.01
Population size: 10000
Mutation rate: 1.854e-09
Recombination rate: 3.053e-11</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">CaeEle</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat8">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat8" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat8">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat8" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: RockmanRIAIL_ce11</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html">https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html</a>&gt; Available maps: RockmanRIAIL_ce11.</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: I, II, III, IV, V, X, MtDNA</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: I, II, III, IV, V, X, MtDNA. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="CanFam">
<h3>CanFam<a class="headerlink" href="#CanFam" title="Permalink to this heading"></a></h3>
<p>Run simulations for Canis familiaris using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Canis familiaris:
Generation time: 3
Population size: 13000
Mutation rate: 4e-09
Recombination rate: 9.506e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">CanFam</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat9">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat9" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat9">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat9" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: Campbell2016_CanFam3_1</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html">https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html</a>&gt; Available maps: Campbell2016_CanFam3_1.</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, 23, chr23, 24, chr24, 25, chr25, 26, chr26, 27, chr27, 28, chr28, 29, chr29, 30, chr30, 31, chr31, 32, chr32, 33, chr33, 34, chr34, 35, chr35, 36, chr36, 37, chr37, 38, chr38, X, chrX, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, X, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="ChlRei">
<h3>ChlRei<a class="headerlink" href="#ChlRei" title="Permalink to this heading"></a></h3>
<p>Run simulations for Chlamydomonas reinhardtii using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Chlamydomonas reinhardtii:
Generation time: 0.001141552511415525
Population size: 1.3999999999999998e-07
Mutation rate: 9.351e-10
Recombination rate: 1.459e-10</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">ChlRei</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat10">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat10" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat10">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat10" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="DroMel">
<h3>DroMel<a class="headerlink" href="#DroMel" title="Permalink to this heading"></a></h3>
<p>Run simulations for Drosophila melanogaster using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Drosophila melanogaster:
Generation time: 0.1
Population size: 1720600
Mutation rate: 5.49e-09
Recombination rate: 2.085e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">DroMel</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">dfes</span> <span class="p">[</span><span class="n">HELP_DFES</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">annotations</span> <span class="p">[</span><span class="n">HELP_ANNOTATIONS</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">annotation</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat11">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat11" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat11">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat11" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: ComeronCrossover_dm6, ComeronCrossoverV2_dm6</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html">https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html</a>&gt; Available maps: ComeronCrossover_dm6, ComeronCrossoverV2_dm6.</p>
</dd>
<dt><kbd>--help-dfes</kbd></dt>
<dd><p>Print list of DFEs and exit. If a DFE ID is given as an argument, show help for this DFE. Otherwise show help for all available DFEs</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 2L, chr2L, 2R, chr2R, 3L, chr3L, 3R, chr3R, 4, chr4, X, chrX, Y, chrY, mitochondrion_genome</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 2L, 2R, 3L, 3R, 4, X, Y, mitochondrion_genome. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: African3Epoch_1S16, OutOfAfrica_2L06</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: African3Epoch_1S16, OutOfAfrica_2L06. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Possible choices: Gamma_H17, LognormalPlusPositive_R16</p>
<p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:Gamma_H17, LognormalPlusPositive_R16. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>--help-annotations</kbd></dt>
<dd><p>Print list of annotations and exit. If a annotation ID is given as an argument, show help for this annotation. Otherwise show help for all available annotations</p>
</dd>
<dt><kbd>--dfe-annotation</kbd></dt>
<dd><p>Possible choices: FlyBase_BDGP6.32.51_exons, FlyBase_BDGP6.32.51_CDS</p>
<p>Specify the intervals over which a given DFE acts using a genomic annotation track Available maps: FlyBase_BDGP6.32.51_exons, FlyBase_BDGP6.32.51_CDS.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="DroSec">
<h3>DroSec<a class="headerlink" href="#DroSec" title="Permalink to this heading"></a></h3>
<p>Run simulations for Drosophila sechellia using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Drosophila sechellia:
Generation time: 0.05
Population size: 100000
Mutation rate: 1.5e-09
Recombination rate: 2.223e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">DroSec</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat12">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat12" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat12">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat12" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 2L, chr2L, 2R, chr2R, 3L, chr3L, 3R, chr3R, X, chrX, 4, chr4</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 2L, 2R, 3L, 3R, X, 4. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="EscCol">
<h3>EscCol<a class="headerlink" href="#EscCol" title="Permalink to this heading"></a></h3>
<p>Run simulations for Escherichia coli using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Escherichia coli:
Generation time: 3.805175e-05
Population size: 180000000.0
Mutation rate: 8.9e-11
Recombination rate: 8.9e-11
Bacterial recombination with tract length: 542</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">EscCol</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat13">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat13" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat13">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat13" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: Chromosome</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: Chromosome. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="GasAcu">
<h3>GasAcu<a class="headerlink" href="#GasAcu" title="Permalink to this heading"></a></h3>
<p>Run simulations for Gasterosteus aculeatus using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Gasterosteus aculeatus:
Generation time: 2
Population size: 10000.0
Mutation rate: 3.7e-08
Recombination rate: 3.001e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">GasAcu</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat14">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat14" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat14">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat14" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, Y, MT</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, Y, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="HelAnn">
<h3>HelAnn<a class="headerlink" href="#HelAnn" title="Permalink to this heading"></a></h3>
<p>Run simulations for Helianthus annuus using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Helianthus annuus:
Generation time: 1.0
Population size: 673968
Mutation rate: 6.1e-09
Recombination rate: 4e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">HelAnn</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat15">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat15" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat15">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat15" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="HelMel">
<h3>HelMel<a class="headerlink" href="#HelMel" title="Permalink to this heading"></a></h3>
<p>Run simulations for Heliconius melpomene using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Heliconius melpomene:
Generation time: 0.0958904109589041
Population size: 2111109
Mutation rate: 2.9e-09
Recombination rate: 3.966e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">HelMel</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat16">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat16" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat16">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat16" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="HomSap">
<h3>HomSap<a class="headerlink" href="#HomSap" title="Permalink to this heading"></a></h3>
<p>Run simulations for Homo sapiens using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Homo sapiens:
Generation time: 30
Population size: 10000
Mutation rate: 1.29e-08
Recombination rate: 1.253e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">HomSap</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">dfes</span> <span class="p">[</span><span class="n">HELP_DFES</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">annotations</span> <span class="p">[</span><span class="n">HELP_ANNOTATIONS</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">annotation</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat17">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat17" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat17">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat17" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: HapMapII_GRCh37, HapMapII_GRCh38, DeCodeSexAveraged_GRCh36, DeCodeSexAveraged_GRCh38, PyrhoACB_GRCh38, PyrhoASW_GRCh38, PyrhoBEB_GRCh38, PyrhoCDX_GRCh38, PyrhoCEU_GRCh38, PyrhoCHB_GRCh38, PyrhoCHS_GRCh38, PyrhoCLM_GRCh38, PyrhoESN_GRCh38, PyrhoFIN_GRCh38, PyrhoGBR_GRCh38, PyrhoGIH_GRCh38, PyrhoGWD_GRCh38, PyrhoIBS_GRCh38, PyrhoITU_GRCh38, PyrhoJPT_GRCh38, PyrhoKHV_GRCh38, PyrhoLWK_GRCh38, PyrhoMSL_GRCh38, PyrhoMXL_GRCh38, PyrhoPEL_GRCh38, PyrhoPJL_GRCh38, PyrhoPUR_GRCh38, PyrhoSTU_GRCh38, PyrhoTSI_GRCh38, PyrhoYRI_GRCh38</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html">https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html</a>&gt; Available maps: HapMapII_GRCh37, HapMapII_GRCh38, DeCodeSexAveraged_GRCh36, DeCodeSexAveraged_GRCh38, PyrhoACB_GRCh38, PyrhoASW_GRCh38, PyrhoBEB_GRCh38, PyrhoCDX_GRCh38, PyrhoCEU_GRCh38, PyrhoCHB_GRCh38, PyrhoCHS_GRCh38, PyrhoCLM_GRCh38, PyrhoESN_GRCh38, PyrhoFIN_GRCh38, PyrhoGBR_GRCh38, PyrhoGIH_GRCh38, PyrhoGWD_GRCh38, PyrhoIBS_GRCh38, PyrhoITU_GRCh38, PyrhoJPT_GRCh38, PyrhoKHV_GRCh38, PyrhoLWK_GRCh38, PyrhoMSL_GRCh38, PyrhoMXL_GRCh38, PyrhoPEL_GRCh38, PyrhoPJL_GRCh38, PyrhoPUR_GRCh38, PyrhoSTU_GRCh38, PyrhoTSI_GRCh38, PyrhoYRI_GRCh38.</p>
</dd>
<dt><kbd>--help-dfes</kbd></dt>
<dd><p>Print list of DFEs and exit. If a DFE ID is given as an argument, show help for this DFE. Otherwise show help for all available DFEs</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, X, chrX, Y, chrY, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, X, Y, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: OutOfAfricaExtendedNeandertalAdmixturePulse_3I21, OutOfAfrica_3G09, OutOfAfrica_2T12, Africa_1T12, AmericanAdmixture_4B11, OutOfAfricaArchaicAdmixture_5R19, Zigzag_1S14, AncientEurasia_9K19, PapuansOutOfAfrica_10J19, AshkSub_7G19, OutOfAfrica_4J17, Africa_1B08, AncientEurope_4A21</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: OutOfAfricaExtendedNeandertalAdmixturePulse_3I21, OutOfAfrica_3G09, OutOfAfrica_2T12, Africa_1T12, AmericanAdmixture_4B11, OutOfAfricaArchaicAdmixture_5R19, Zigzag_1S14, AncientEurasia_9K19, PapuansOutOfAfrica_10J19, AshkSub_7G19, OutOfAfrica_4J17, Africa_1B08, AncientEurope_4A21. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Possible choices: Gamma_K17, Gamma_H17, LogNormal_H17, Mixed_K23</p>
<p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:Gamma_K17, Gamma_H17, LogNormal_H17, Mixed_K23. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>--help-annotations</kbd></dt>
<dd><p>Print list of annotations and exit. If a annotation ID is given as an argument, show help for this annotation. Otherwise show help for all available annotations</p>
</dd>
<dt><kbd>--dfe-annotation</kbd></dt>
<dd><p>Possible choices: ensembl_havana_104_exons, ensembl_havana_104_CDS</p>
<p>Specify the intervals over which a given DFE acts using a genomic annotation track Available maps: ensembl_havana_104_exons, ensembl_havana_104_CDS.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="MusMus">
<h3>MusMus<a class="headerlink" href="#MusMus" title="Permalink to this heading"></a></h3>
<p>Run simulations for Mus musculus using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Mus musculus:
Generation time: 0.75
Population size: 500000
Mutation rate: 5.4e-09
Recombination rate: 5.567e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">MusMus</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat18">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat18" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat18">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat18" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, X, chrX, Y, chrY, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, X, Y, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: DomesticusEurope_1F22, MusculusKorea_1F22, CastaneusIndia_1F22</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: DomesticusEurope_1F22, MusculusKorea_1F22, CastaneusIndia_1F22. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="OrySat">
<h3>OrySat<a class="headerlink" href="#OrySat" title="Permalink to this heading"></a></h3>
<p>Run simulations for Oryza sativa using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Oryza sativa:
Generation time: 1
Population size: 46875
Mutation rate: 3.2e-09
Recombination rate: 8.955e-10</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">OrySat</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat19">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat19" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat19">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat19" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, Mt, Pt, 2, chr2, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, Mt, Pt, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: BottleneckMigration_3C07</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: BottleneckMigration_3C07. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="PanTro">
<h3>PanTro<a class="headerlink" href="#PanTro" title="Permalink to this heading"></a></h3>
<p>Run simulations for Pan troglodytes using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Pan troglodytes:
Generation time: 24.6
Population size: 16781
Mutation rate: 1.6e-08
Recombination rate: 1.2e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">PanTro</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat20">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat20" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat20">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat20" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2A, chr2A, 2B, chr2B, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, X, chrX, Y, chrY</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2A, 2B, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, X, Y. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: BonoboGhost_4K19</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: BonoboGhost_4K19. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="PapAnu">
<h3>PapAnu<a class="headerlink" href="#PapAnu" title="Permalink to this heading"></a></h3>
<p>Run simulations for Papio anubis using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Papio anubis:
Generation time: 11
Population size: 335505
Mutation rate: 5.7e-09
Recombination rate: 1.124e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">PapAnu</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat21">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat21" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat21">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat21" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: Pyrho_PAnubis1_0</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html">https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html</a>&gt; Available maps: Pyrho_PAnubis1_0.</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, X, Y</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, X, Y. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: SinglePopSMCpp_1W22</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: SinglePopSMCpp_1W22. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="PhoSin">
<h3>PhoSin<a class="headerlink" href="#PhoSin" title="Permalink to this heading"></a></h3>
<p>Run simulations for Phocoena sinus using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Phocoena sinus:
Generation time: 11.9
Population size: 3500
Mutation rate: 5.83e-09
Recombination rate: 1e-08</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">PhoSin</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">dfes</span> <span class="p">[</span><span class="n">HELP_DFES</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">annotations</span> <span class="p">[</span><span class="n">HELP_ANNOTATIONS</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">annotation</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat22">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat22" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat22">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat22" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>--help-dfes</kbd></dt>
<dd><p>Print list of DFEs and exit. If a DFE ID is given as an argument, show help for this DFE. Otherwise show help for all available DFEs</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, X</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, X. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: Vaquita2Epoch_1R22</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: Vaquita2Epoch_1R22. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Possible choices: Gamma_R22</p>
<p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:Gamma_R22. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>--help-annotations</kbd></dt>
<dd><p>Print list of annotations and exit. If a annotation ID is given as an argument, show help for this annotation. Otherwise show help for all available annotations</p>
</dd>
<dt><kbd>--dfe-annotation</kbd></dt>
<dd><p>Possible choices: Phocoena_sinus.mPhoSin1.pri.110_exons, Phocoena_sinus.mPhoSin1.pri.110_CDS</p>
<p>Specify the intervals over which a given DFE acts using a genomic annotation track Available maps: Phocoena_sinus.mPhoSin1.pri.110_exons, Phocoena_sinus.mPhoSin1.pri.110_CDS.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="PonAbe">
<h3>PonAbe<a class="headerlink" href="#PonAbe" title="Permalink to this heading"></a></h3>
<p>Run simulations for Pongo abelii using up-to-date genome information,
genetic maps and simulation models from the literature. NOTE: By
default, the tskit ‘.trees’ binary file is written to stdout,so you
should either redirect this to a file or use the ‘–output’ option to
specify a filename.</p>
<p>Default population parameters for Pongo abelii:
Generation time: 25
Population size: 17900.0
Mutation rate: 1.5e-08
Recombination rate: 6.403e-09</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">PonAbe</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="n">HELP_GENETIC_MAPS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat23">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat23" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat23">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat23" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>--help-genetic-maps</kbd></dt>
<dd><p>Print list of genetic maps and exit. If a genetic map ID is given as an argument, show help for this map. Otherwise show help for all available genetic maps</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --genetic-map</kbd></dt>
<dd><p>Possible choices: NaterPA_PonAbe3, NaterPP_PonAbe3</p>
<p>Specify a particular genetic map. By default, a chromosome-specific uniform recombination rate is used. These default rates are listed in the catalog: &lt;<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html">https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html</a>&gt; Available maps: NaterPA_PonAbe3, NaterPP_PonAbe3.</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, chr1, 2A, 2a, chr2a, chr2A, 2B, 2b, chr2b, chr2B, 3, chr3, 4, chr4, 5, chr5, 6, chr6, 7, chr7, 8, chr8, 9, chr9, 10, chr10, 11, chr11, 12, chr12, 13, chr13, 14, chr14, 15, chr15, 16, chr16, 17, chr17, 18, chr18, 19, chr19, 20, chr20, 21, chr21, 22, chr22, X, chrX, MT, chrM</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1, 2A, 2B, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, X, MT. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Possible choices: TwoSpecies_2L11</p>
<p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: TwoSpecies_2L11. Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="StrAga">
<h3>StrAga<a class="headerlink" href="#StrAga" title="Permalink to this heading"></a></h3>
<p>Run simulations for Streptococcus agalactiae using up-to-date genome
information, genetic maps and simulation models from the literature.
NOTE: By default, the tskit ‘.trees’ binary file is written to
stdout,so you should either redirect this to a file or use the ‘–
output’ option to specify a filename.</p>
<p>Default population parameters for Streptococcus agalactiae:
Generation time: 0.0027397260273972603
Population size: 140000
Mutation rate: 1.53e-09
Recombination rate: 1.53e-10
Bacterial recombination with tract length: 120000</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">StrAga</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">help</span><span class="o">-</span><span class="n">models</span> <span class="p">[</span><span class="n">HELP_MODELS</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BIBTEX_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">D</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">L</span> <span class="n">LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INCLUSION_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXCLUSION_MASK</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LENGTH_MULTIPLIER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">left</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">right</span> <span class="n">RIGHT</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">mutation</span><span class="o">-</span><span class="n">ids</span><span class="o">-</span><span class="k">as</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">interval</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dfe</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span>
                 <span class="n">samples</span> <span class="p">[</span><span class="n">samples</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat24">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat24" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>samples</kbd></dt>
<dd><p>The number of samples to draw from each population. At least two genomes must be sampled, say, for one diploid individual. The number of arguments accepted depends on the specified simulation model: for a model that has n populations, we can specify the number of samples to draw from each of these populations.We do not need to provide sample numbers of each of the populations; those that are omitted are set to zero.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat24">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat24" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>--help-models</kbd></dt>
<dd><p>Print descriptions of simulation models and exit. If a model ID is provided as an argument show help for this model; otherwise show help for all available models</p>
</dd>
<dt><kbd>-b, --bibtex-file</kbd></dt>
<dd><p>Write citations to a given bib file. This will overwrite the file.</p>
</dd>
<dt><kbd>-D, --dry-run</kbd></dt>
<dd><p>Do not run actual simulation</p>
<p>Default: False</p>
</dd>
<dt><kbd>-c, --chromosome</kbd></dt>
<dd><p>Possible choices: 1, I, chr1</p>
<p>Simulate a specific chromosome. If no chromosome is given, simulate a generic contig of given length, specified using –length. Options: 1. Default=None.</p>
</dd>
<dt><kbd>-L, --length</kbd></dt>
<dd><p>Simulate a default contig of given length.</p>
</dd>
<dt><kbd>-i, --inclusion-mask</kbd></dt>
<dd><p>Path to inclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-e, --exclusion-mask</kbd></dt>
<dd><p>Path to exclusion mask specified in bed format.</p>
</dd>
<dt><kbd>-l, --length-multiplier</kbd></dt>
<dd><p>Simulate a sequence of length l times the named chromosome’s length using the named chromosome’s mutation and recombination rates.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--left</kbd></dt>
<dd><p>The leftmost coordinate (inclusive) of the contig on the named chromosome. Defaults to 0 (the beginning of the chromosome).</p>
</dd>
<dt><kbd>--right</kbd></dt>
<dd><p>The rightmost coordinate (exclusive) of the contig on the named chromosome. Defaults to the length of the chromosome.</p>
</dd>
<dt><kbd>-s, --seed</kbd></dt>
<dd><p>The random seed to use for simulations. If not specified a high-quality random seed will be generated automatically. For msprime, seeds must be &gt; 0 and &lt; 2^32.</p>
</dd>
<dt><kbd>--keep-mutation-ids-as-alleles</kbd></dt>
<dd><p>If set, mutations in SLiM simulations will have alleles coded as numeric ids rather than random nucleotides.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-d, --demographic-model</kbd></dt>
<dd><p>Specify a simulation model. If no model is specified, a single population constant size model is used. Available models: . Please see –help-models for details of these models.</p>
</dd>
<dt><kbd>--dfe</kbd></dt>
<dd><p>Specify a Distribution of Fitness Effects (DFE) model. If no DFE is specified, all mutations are neutral. Available DFE models:. Please see –help-dfes for details of these DFE models.</p>
</dd>
<dt><kbd>--dfe-interval</kbd></dt>
<dd><p>Specify the interval where selection (given a DFE) is simulated. Anything outside of the interval is simulated as neutral. If no interval is specified, selection is simulated across the entire contig. The interval is written as ‘start,end’ (separated by a comma, with no space, for instance: –dfe-interval 1000,2000.</p>
</dd>
<dt><kbd>--dfe-bed-file</kbd></dt>
<dd><p>A bed file specifing the intervals where selection (given a DFE) is simulated. Non-overlapping intervals belonging to the same chromosome are required. If no interval is specified, selection is simulated across the entire contig. See also –dfe-interval and –dfe-annotation.</p>
</dd>
<dt><kbd>-o, --output</kbd></dt>
<dd><p>Where to write the output tree sequence file. Defaults to stdout if not specified</p>
</dd>
</dl>
</section>
</section>
<section id="download-genetic-maps">
<h3>download-genetic-maps<a class="headerlink" href="#download-genetic-maps" title="Permalink to this heading"></a></h3>
<p>Download genetic maps and store them in the cache directory. Maps are downloaded regardless of whether they are already in the cache or not. Please use the –cache-dir option to download maps to a specific directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdpopsim</span> <span class="n">download</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">maps</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="n">species</span><span class="p">]</span> <span class="p">[</span><span class="n">genetic_maps</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat25">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat25" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>species</kbd></dt>
<dd><p>Download genetic maps for this species. If not specified download all known genetic maps.</p>
</dd>
<dt><kbd>genetic_maps</kbd></dt>
<dd><p>If specified, download these genetic maps. If no maps are provided, download all maps for this species.</p>
</dd>
</dl>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tutorial.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api.html" class="btn btn-neutral float-right" title="API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019-2022, PopSim Consortium.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>