

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>stdpopsim.genomes &mdash; stdpopsim 0.1.1 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home" alt="Documentation Home"> stdpopsim
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../catalog.html">Catalog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli_arguments.html">Command-line options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelogs.html">Changelogs</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">stdpopsim</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>stdpopsim.genomes</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for stdpopsim.genomes</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Infrastructure for defining information about species&#39; genomes.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">attr</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">tskit</span>
<span class="kn">import</span> <span class="nn">stdpopsim</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="Genome"><a class="viewcode-back" href="../../api.html#stdpopsim.Genome">[docs]</a><span class="nd">@attr</span><span class="o">.</span><span class="n">s</span>
<span class="k">class</span> <span class="nc">Genome</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class representing the genome for a species.</span>

<span class="sd">    :ivar chromosomes: A list of :class:`.Chromosome` objects.</span>
<span class="sd">    :vartype chromosomes: list</span>
<span class="sd">    :ivar mutation_rate_citations: A list of :class:`.Citation` objects</span>
<span class="sd">        providing justification for the mutation rate estimate.</span>
<span class="sd">    :vartype mutation_rate_citations: list</span>
<span class="sd">    :ivar recombination_rate_citations: A list of :class:`.Citation` objects</span>
<span class="sd">        providing justification for the recombination rate estimate.</span>
<span class="sd">    :vartype recombination_rate_citations: list</span>
<span class="sd">    :ivar assembly_citations: A list of :class:`.Citation` objects</span>
<span class="sd">        providing reference to the source of the genome assembly.</span>
<span class="sd">    :vartype assembly_citations: list</span>
<span class="sd">    :ivar length: The total length of the genome.</span>
<span class="sd">    :vartype length: int</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>
    <span class="n">mutation_rate_citations</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">recombination_rate_citations</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">assembly_citations</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">assembly_name</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">assembly_accession</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">length</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__attrs_post_init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">+=</span> <span class="n">chromosome</span><span class="o">.</span><span class="n">length</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">s</span> <span class="o">=</span> <span class="s2">&quot;Chromosomes:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">length_sorted</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="o">-</span><span class="n">x</span><span class="o">.</span><span class="n">length</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">length_sorted</span><span class="p">:</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">s</span>

<div class="viewcode-block" id="Genome.get_chromosome"><a class="viewcode-back" href="../../api.html#stdpopsim.Genome.get_chromosome">[docs]</a>    <span class="k">def</span> <span class="nf">get_chromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="nb">id</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the chromosome with the specified ``id``.</span>

<span class="sd">        :param str id: The string ID of the chromosome.</span>
<span class="sd">            A complete list of chromosome IDs for each species can be found in the</span>
<span class="sd">            &quot;Genome&quot; subsection for the species in the :ref:`sec_catalog`.</span>
<span class="sd">        :rtype: :class:`Chromosome`</span>
<span class="sd">        :return: A :class:`Chromosome` that defines properties of the chromosome</span>
<span class="sd">            such as length, mutation rate, and recombination rate.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">chrom</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="nb">id</span> <span class="ow">or</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">chrom</span><span class="o">.</span><span class="n">synonyms</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">chrom</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Chromosome not found&quot;</span><span class="p">)</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">mean_recombination_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The length-weighted mean recombination rate across all chromosomes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mean_recombination_rate</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
            <span class="n">normalized_weight</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">length</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
            <span class="n">cont</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">recombination_rate</span> <span class="o">*</span> <span class="n">normalized_weight</span>
            <span class="n">mean_recombination_rate</span> <span class="o">+=</span> <span class="n">cont</span>
        <span class="k">return</span> <span class="n">mean_recombination_rate</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">mean_mutation_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The length-weighted mean mutation rate across all chromosomes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mean_mutation_rate</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
            <span class="n">normalized_weight</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">length</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
            <span class="n">cont</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">mutation_rate</span> <span class="o">*</span> <span class="n">normalized_weight</span>
            <span class="n">mean_mutation_rate</span> <span class="o">+=</span> <span class="n">cont</span>
        <span class="k">return</span> <span class="n">mean_mutation_rate</span></div>


<div class="viewcode-block" id="Chromosome"><a class="viewcode-back" href="../../api.html#stdpopsim.Chromosome">[docs]</a><span class="nd">@attr</span><span class="o">.</span><span class="n">s</span>
<span class="k">class</span> <span class="nc">Chromosome</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class representing a single chromosome for a species.</span>

<span class="sd">    :ivar str ~.id: The string identifier for the chromosome.</span>
<span class="sd">    :ivar int length: The length of the chromosome.</span>
<span class="sd">    :ivar float mutation_rate: The mutation rate used when simulating this</span>
<span class="sd">        chromosome.</span>
<span class="sd">    :ivar float recombination_rate: The recombination rate used when simulating</span>
<span class="sd">        this chromosome (if not using a genetic map).</span>
<span class="sd">    :ivar synonyms: List of synonyms that may be used when requesting this</span>
<span class="sd">        chromosome by ID, e.g. from the command line interface.</span>
<span class="sd">    :vartype synonyms: list of str</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">id</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">length</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">recombination_rate</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mutation_rate</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">synonyms</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<span class="nd">@attr</span><span class="o">.</span><span class="n">s</span><span class="p">(</span><span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">class</span> <span class="nc">GenomicElementType</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    TODO: docstring.</span>
<span class="sd">    :ivar mutation_type_ids: Indices for the mutation types that can occur in</span>
<span class="sd">        genomic elements of this type.</span>
<span class="sd">    :vartype mutation_type_ids: list</span>
<span class="sd">    :ivar proportions: Proportions (out of the mutation rate) in which each</span>
<span class="sd">        mutation type can happen in genomic elements of this type.</span>
<span class="sd">    :vartype: proportions: list</span>
<span class="sd">    :ivar intervals: (n, 2) numpy array with [left, right) genomic intervals</span>
<span class="sd">        that will be of this GenomicElementType. The intervals are used to</span>
<span class="sd">        specify SLiM GenomicElements. Intervals should be non-overlapping and</span>
<span class="sd">        in asceding order.</span>
<span class="sd">    :vartype intervals: numpy.array, dtype=np.int32</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">mutation_type_ids</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="k">lambda</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">type</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>
    <span class="n">proportions</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="k">lambda</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">type</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>
    <span class="n">intervals</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__attrs_post_init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proportions</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_type_ids</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;The list of proportions and mutation type ids &quot;</span>
                <span class="s2">&quot;should be of the same length.&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">([</span><span class="ow">not</span> <span class="p">(</span><span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">prop</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">prop</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proportions</span><span class="p">]):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Proportions must lie within the [0,1] interval.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proportions</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The sum of proportions can not be greater than 1.&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">intervals</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">util</span><span class="o">.</span><span class="n">safe_np_int_cast</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">intervals</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">intervals</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">intervals</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Left positions cannot be greater than or equal to right positions.&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">intervals</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">intervals</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;A valid array of intervals must be provided.&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="Contig"><a class="viewcode-back" href="../../api.html#stdpopsim.Contig">[docs]</a><span class="nd">@attr</span><span class="o">.</span><span class="n">s</span>
<span class="k">class</span> <span class="nc">Contig</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class representing a contiguous region of genome that is to be</span>
<span class="sd">    simulated. This contains the information about mutation rates</span>
<span class="sd">    and recombination rates that are needed to simulate this region.</span>
<span class="sd">    Genomic element types can be provided</span>

<span class="sd">    :ivar mutation_rate: The rate of mutation per base per generation.</span>
<span class="sd">    :vartype mutation_rate: float</span>
<span class="sd">    :ivar recombination_map: The recombination map for the region. See the</span>
<span class="sd">        `msprime documentation</span>
<span class="sd">        &lt;https://msprime.readthedocs.io/en/stable/api.html#msprime.RecombinationMap&gt;`_</span>
<span class="sd">        for more details.</span>
<span class="sd">    :vartype recombination_map: msprime.simulations.RecombinationMap</span>
<span class="sd">    :ivar genomic_element_types: a list of stdpopsim.GenomicElementType</span>
<span class="sd">        objects. By default, the only element is a single genomic element type</span>
<span class="sd">        that spans the entire contiguous region.</span>
<span class="sd">    :vartype genomic_element_types: list</span>
<span class="sd">    :ivar mutation_types: a list of stdpopsim.ext.MutationType object. By</span>
<span class="sd">        default, the only mutation type is neutral. See</span>
<span class="sd">        stdpopsim.MutationType for more details.</span>

<span class="sd">    .. note::</span>

<span class="sd">        To run stdpopsim simulations with alternative, user-specified mutation</span>
<span class="sd">        or recombination rates, a new contig can be created based on an existing</span>
<span class="sd">        one. For instance, the following will create a ``new_contig`` that,</span>
<span class="sd">        when simulated, will have double the mutation rate of the ``old_contig``:</span>

<span class="sd">        .. code-block:: python</span>

<span class="sd">            new_contig = stdpopsim.Contig(</span>
<span class="sd">                mutation_rate=old_contig.mutation_rate * 2,</span>
<span class="sd">                recombination_map=old_contig.recombination_map,</span>
<span class="sd">                genetic_map=old_contig.genetic_map,</span>
<span class="sd">            )</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">recombination_map</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mutation_rate</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">genetic_map</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">genomic_element_types</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="nb">list</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mutation_types</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="nb">list</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">recombination_map</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">recombination_map</span><span class="o">.</span><span class="n">get_length</span><span class="p">()</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">is_slim_incompatible</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>

    <span class="nd">@genomic_element_types</span><span class="o">.</span><span class="n">default</span>
    <span class="k">def</span> <span class="nf">_get_genetic_map_sized_genomic_element</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Default is a genetic map sized genomic element</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_slim_incompatible</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">GenomicElementType</span><span class="p">(</span><span class="n">intervals</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)]]))]</span>

    <span class="nd">@mutation_types</span><span class="o">.</span><span class="n">default</span>
    <span class="k">def</span> <span class="nf">_get_default_mut_types</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># One neutral mutation type, simulated with msprime</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_slim_incompatible</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">stdpopsim</span><span class="o">.</span><span class="n">ext</span><span class="o">.</span><span class="n">MutationType</span><span class="p">()]</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">slim_fractions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_slim_incompatible</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="n">props</span> <span class="o">=</span> <span class="p">[</span><span class="nb">sum</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">proportions</span><span class="p">)</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genomic_element_types</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">([</span><span class="n">p</span> <span class="o">&gt;</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="mf">1e-15</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">props</span><span class="p">]):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;The sum of the proportions within genomic element types cannot&quot;</span>
                <span class="s2">&quot; exceed 1.&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">props</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genomic_intervals</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns an (n, 3) numpy array with n intervals (left, right, type)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_slim_incompatible</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="n">all_intervals</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">((</span><span class="n">g</span><span class="o">.</span><span class="n">intervals</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">((</span><span class="n">g</span><span class="o">.</span><span class="n">intervals</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">i</span><span class="p">)))</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genomic_element_types</span><span class="p">)</span>
            <span class="p">],</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">sorter</span> <span class="o">=</span> <span class="n">all_intervals</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">argsort</span><span class="p">()</span>
        <span class="n">all_intervals</span> <span class="o">=</span> <span class="n">all_intervals</span><span class="p">[</span><span class="n">sorter</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">all_intervals</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Genomic elements are outside the genomic region&quot;</span>
                <span class="s2">&quot; implied by the recombination map.&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">all_intervals</span><span class="p">[</span><span class="mi">1</span><span class="p">:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">all_intervals</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;You cannot have overlapping genomic elements.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">all_intervals</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">msp_mutation_rate_map</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        breaks = [0,10,20] means [0,10) and [10,20)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_slim_incompatible</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="n">breaks</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">rates</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genomic_intervals</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">start</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">breaks</span><span class="p">:</span>
                <span class="n">breaks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">start</span><span class="p">)</span>
                <span class="n">rates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_rate</span><span class="p">)</span>
            <span class="n">breaks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">end</span><span class="p">)</span>
            <span class="n">rates</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="mi">1</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">slim_fractions</span><span class="p">[</span><span class="n">t</span><span class="p">])</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_rate</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">breaks</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)):</span>
            <span class="n">breaks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">))</span>
            <span class="n">rates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_rate</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">breaks</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">rates</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">breaks</span><span class="p">,</span> <span class="n">rates</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">slim_mutation_rate_map</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        breaks = [10,20] means [0,10] and (10,20]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_slim_incompatible</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="c1"># making sure the mut types are still concordant with the ge types</span>
        <span class="n">_check_mutation_types</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_types</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genomic_element_types</span><span class="p">)</span>
        <span class="n">breaks</span><span class="p">,</span> <span class="n">rates</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">msp_mutation_rate_map</span>
        <span class="k">assert</span> <span class="n">breaks</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
        <span class="n">slim_breaks</span> <span class="o">=</span> <span class="p">[</span><span class="n">b</span> <span class="o">-</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">breaks</span><span class="p">[</span><span class="mi">1</span><span class="p">:]]</span>
        <span class="n">slim_rates</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_rate</span> <span class="o">-</span> <span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">rates</span><span class="p">]</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">slim_breaks</span><span class="p">,</span> <span class="n">slim_rates</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_mutation_type</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutation_type</span><span class="p">,</span> <span class="n">genomic_element_type_id</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">proportion</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">genomic_element_type_id</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genomic_element_types</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Genomic element type of id </span><span class="si">{</span><span class="n">genomic_element_type_id</span><span class="si">}</span><span class="s2"> does not exist.&quot;</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_types</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation_type</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genomic_element_types</span><span class="p">[</span><span class="n">genomic_element_type_id</span><span class="p">]</span><span class="o">.</span><span class="n">mutation_type_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_types</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genomic_element_types</span><span class="p">[</span><span class="n">genomic_element_type_id</span><span class="p">]</span><span class="o">.</span><span class="n">proportions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="n">proportion</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">gmap</span> <span class="o">=</span> <span class="s2">&quot;None&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genetic_map</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">genetic_map</span><span class="o">.</span><span class="n">id</span>
        <span class="n">s</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;Contig(length=</span><span class="si">{:.2G}</span><span class="s2">, recombination_rate=</span><span class="si">{:.2G}</span><span class="s2">, &quot;</span>
            <span class="s2">&quot;mutation_rate=</span><span class="si">{:.2G}</span><span class="s2">, genetic_map=</span><span class="si">{}</span><span class="s2">)&quot;</span>
        <span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">recombination_map</span><span class="o">.</span><span class="n">mean_recombination_rate</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mutation_rate</span><span class="p">,</span>
            <span class="n">gmap</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">s</span></div>


<span class="k">def</span> <span class="nf">_check_mutation_types</span><span class="p">(</span><span class="n">mutation_types</span><span class="p">,</span> <span class="n">genomic_element_types</span><span class="p">):</span>
    <span class="n">mut_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">mtid</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genomic_element_types</span> <span class="k">for</span> <span class="n">mtid</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">mutation_type_ids</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">mutation_types</span><span class="p">)))</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="n">mut_ids</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;The mutation type ids in the genomic elements &quot;</span>
            <span class="s2">&quot;must match the list of mutation types.&quot;</span><span class="p">,</span>
            <span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">mutation_types</span><span class="p">))),</span>
            <span class="nb">set</span><span class="p">(</span><span class="n">mut_ids</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="n">neutral_mut_ids</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">i</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">m</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mutation_types</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">distribution_type</span> <span class="o">==</span> <span class="s2">&quot;f&quot;</span> <span class="ow">and</span> <span class="n">m</span><span class="o">.</span><span class="n">distribution_args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
    <span class="p">]</span>
    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">g</span><span class="o">.</span><span class="n">proportions</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">mutation_type_ids</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="o">!=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genomic_element_types</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">neutral_mut_ids</span>
        <span class="p">]</span>
    <span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;Warning: you defined neutral mutation type(s) &quot;</span>
            <span class="s2">&quot;which have non-zero proportions. This means &quot;</span>
            <span class="s2">&quot;neutral mutations will be simulated in SLiM, what &quot;</span>
            <span class="s2">&quot;can considerably slow down your simulations. If &quot;</span>
            <span class="s2">&quot;your neutral mutation type(s) are assigned &quot;</span>
            <span class="s2">&quot;proportions equal to zero, then they will be &quot;</span>
            <span class="s2">&quot;added onto the tree sequence using msprime.&quot;</span>
        <span class="p">)</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2019-2020, PopSim Consortium

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>