<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>stdpopsim.genomes &mdash; stdpopsim 0.1.1 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> stdpopsim
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../catalog.html">Catalog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli_arguments.html">Command-line options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelogs.html">Changelogs</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">stdpopsim</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>stdpopsim.genomes</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for stdpopsim.genomes</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Infrastructure for defining information about species&#39; genomes and genomic</span>
<span class="sd">regions to be simulated.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">attr</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">msprime</span>

<span class="kn">import</span> <span class="nn">stdpopsim</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="Genome"><a class="viewcode-back" href="../../api.html#stdpopsim.Genome">[docs]</a><span class="nd">@attr</span><span class="o">.</span><span class="n">s</span>
<span class="k">class</span> <span class="nc">Genome</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class representing the genome for a species.</span>

<span class="sd">    :ivar chromosomes: A list of :class:`.Chromosome` objects.</span>
<span class="sd">    :vartype chromosomes: list</span>
<span class="sd">    :ivar citations: A list of :class:`.Citation` objects</span>
<span class="sd">        providing the source for the genome assembly,</span>
<span class="sd">        mutation rate and recombination rate estimates.</span>
<span class="sd">    :vartype citations: list</span>
<span class="sd">    :ivar length: The total length of the genome.</span>
<span class="sd">    :vartype length: int</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># TODO document the assembly_name and accession</span>

    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>
    <span class="n">assembly_name</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">assembly_accession</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">citations</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<div class="viewcode-block" id="Genome.from_data"><a class="viewcode-back" href="../../api.html#stdpopsim.Genome.from_data">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">from_data</span><span class="p">(</span><span class="n">genome_data</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="p">,</span> <span class="n">citations</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Construct a Genome object from the specified dictionary of</span>
<span class="sd">        genome information from Ensembl, recombination_rate and</span>
<span class="sd">        mutation_rate dictionaries.</span>

<span class="sd">        This method is for internal use only.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">chr_names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;chromosomes&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="n">recombination_rate</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">==</span> <span class="n">chr_names</span>
        <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="n">mutation_rate</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">==</span> <span class="n">chr_names</span>
        <span class="n">chromosomes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;chromosomes&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">chromosomes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">Chromosome</span><span class="p">(</span>
                    <span class="nb">id</span><span class="o">=</span><span class="n">name</span><span class="p">,</span>
                    <span class="n">length</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">],</span>
                    <span class="n">synonyms</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">],</span>
                    <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">[</span><span class="n">name</span><span class="p">],</span>
                    <span class="n">recombination_rate</span><span class="o">=</span><span class="n">recombination_rate</span><span class="p">[</span><span class="n">name</span><span class="p">],</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">Genome</span><span class="p">(</span>
            <span class="n">chromosomes</span><span class="o">=</span><span class="n">chromosomes</span><span class="p">,</span>
            <span class="n">assembly_name</span><span class="o">=</span><span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;assembly_name&quot;</span><span class="p">],</span>
            <span class="n">assembly_accession</span><span class="o">=</span><span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;assembly_accession&quot;</span><span class="p">],</span>
            <span class="n">citations</span><span class="o">=</span><span class="n">citations</span><span class="p">,</span>
        <span class="p">)</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">chrom</span><span class="o">.</span><span class="n">length</span> <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">s</span> <span class="o">=</span> <span class="s2">&quot;Chromosomes:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">length_sorted</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="o">-</span><span class="n">x</span><span class="o">.</span><span class="n">length</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">length_sorted</span><span class="p">:</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">s</span>

<div class="viewcode-block" id="Genome.get_chromosome"><a class="viewcode-back" href="../../api.html#stdpopsim.Genome.get_chromosome">[docs]</a>    <span class="k">def</span> <span class="nf">get_chromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="nb">id</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the chromosome with the specified ``id``.</span>

<span class="sd">        :param str id: The string ID of the chromosome.</span>
<span class="sd">            A complete list of chromosome IDs for each species can be found in the</span>
<span class="sd">            &quot;Genome&quot; subsection for the species in the :ref:`sec_catalog`.</span>
<span class="sd">        :rtype: :class:`Chromosome`</span>
<span class="sd">        :return: A :class:`Chromosome` that defines properties of the chromosome</span>
<span class="sd">            such as length, mutation rate, and recombination rate.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">chrom</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="nb">id</span> <span class="ow">or</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">chrom</span><span class="o">.</span><span class="n">synonyms</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">chrom</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Chromosome not found&quot;</span><span class="p">)</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">mean_recombination_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The length-weighted mean recombination rate across all chromosomes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
        <span class="n">mean_recombination_rate</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
            <span class="n">normalized_weight</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">length</span> <span class="o">/</span> <span class="n">length</span>
            <span class="n">cont</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">recombination_rate</span> <span class="o">*</span> <span class="n">normalized_weight</span>
            <span class="n">mean_recombination_rate</span> <span class="o">+=</span> <span class="n">cont</span>
        <span class="k">return</span> <span class="n">mean_recombination_rate</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">mean_mutation_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The length-weighted mean mutation rate across all chromosomes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">length</span>
        <span class="n">mean_mutation_rate</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
            <span class="n">normalized_weight</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">length</span> <span class="o">/</span> <span class="n">length</span>
            <span class="n">cont</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">mutation_rate</span> <span class="o">*</span> <span class="n">normalized_weight</span>
            <span class="n">mean_mutation_rate</span> <span class="o">+=</span> <span class="n">cont</span>
        <span class="k">return</span> <span class="n">mean_mutation_rate</span></div>


<div class="viewcode-block" id="Chromosome"><a class="viewcode-back" href="../../api.html#stdpopsim.Chromosome">[docs]</a><span class="nd">@attr</span><span class="o">.</span><span class="n">s</span>
<span class="k">class</span> <span class="nc">Chromosome</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class representing a single chromosome for a species.</span>

<span class="sd">    :ivar str ~.id: The string identifier for the chromosome.</span>
<span class="sd">    :ivar int length: The length of the chromosome.</span>
<span class="sd">    :ivar float mutation_rate: The mutation rate used when simulating this</span>
<span class="sd">        chromosome.</span>
<span class="sd">    :ivar float recombination_rate: The recombination rate used when simulating</span>
<span class="sd">        this chromosome (if not using a genetic map).</span>
<span class="sd">    :ivar synonyms: List of synonyms that may be used when requesting this</span>
<span class="sd">        chromosome by ID, e.g. from the command line interface.</span>
<span class="sd">    :vartype synonyms: list of str</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">id</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">length</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">recombination_rate</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mutation_rate</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">synonyms</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">,</span> <span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<div class="viewcode-block" id="Contig"><a class="viewcode-back" href="../../api.html#stdpopsim.Contig">[docs]</a><span class="nd">@attr</span><span class="o">.</span><span class="n">s</span><span class="p">(</span><span class="n">kw_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">class</span> <span class="nc">Contig</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class representing a contiguous region of genome that is to be simulated.</span>
<span class="sd">    This contains the information about mutation rates, distributions of</span>
<span class="sd">    fitness effects (DFEs), and recombination rates that are needed to simulate</span>
<span class="sd">    this region.</span>

<span class="sd">    Information about targets of selection are contained in the ``dfe_list``</span>
<span class="sd">    and ``interval_list``. These must be of the same length, and the k-th DFE</span>
<span class="sd">    applies to the k-th interval; see :meth:`.add_dfe` for more information.</span>

<span class="sd">    :ivar mutation_rate: The rate of mutation per base per generation.</span>
<span class="sd">    :vartype mutation_rate: float</span>
<span class="sd">    :ivar recombination_map: The recombination map for the region. See the</span>
<span class="sd">        :class:`msprime.RateMap` for details.</span>
<span class="sd">    :vartype recombination_map: msprime.RateMap</span>
<span class="sd">    :ivar mask_intervals: Intervals to keep in simulated tree sequence, as a list</span>
<span class="sd">        of (left_position, right_position), such that intervals are non-overlapping</span>
<span class="sd">        and in ascending order. Should have shape Nx2, where N is the number of</span>
<span class="sd">        intervals.</span>
<span class="sd">    :vartype mask_intervals: array-like (?)</span>
<span class="sd">    :ivar exclude: If True, ``mask_intervals`` specify regions to exclude. If False,</span>
<span class="sd">        ``mask_intervals`` specify regions in keep.</span>
<span class="sd">    :vartype exclude: bool</span>
<span class="sd">    :ivar dfe_list: a list of :class:`.DFE` objects.</span>
<span class="sd">        By default, the only DFE is completely neutral.</span>
<span class="sd">    :vartype dfe_list: list</span>
<span class="sd">    :ivar interval_list: A list of :class:`np.array` objects containing integers.</span>
<span class="sd">        By default, the inital interval list spans the whole chromosome with the</span>
<span class="sd">        neutral DFE.</span>

<span class="sd">    .. note::</span>
<span class="sd">        To run stdpopsim simulations with alternative, user-specified mutation</span>
<span class="sd">        or recombination rates, a new contig can be created based on an existing</span>
<span class="sd">        one. For instance, the following will create a ``new_contig`` that,</span>
<span class="sd">        when simulated, will have double the mutation rate of the ``old_contig``:</span>

<span class="sd">        .. code-block:: python</span>

<span class="sd">            new_contig = stdpopsim.Contig(</span>
<span class="sd">                mutation_rate=old_contig.mutation_rate * 2,</span>
<span class="sd">                recombination_map=old_contig.recombination_map,</span>
<span class="sd">                genetic_map=old_contig.genetic_map,</span>
<span class="sd">            )</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">recombination_map</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">()</span>
    <span class="n">mutation_rate</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">genetic_map</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">inclusion_mask</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">exclusion_mask</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">dfe_list</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>
    <span class="n">interval_list</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">(</span><span class="n">factory</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__attrs_post_init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_dfe</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)]]),</span>
            <span class="n">stdpopsim</span><span class="o">.</span><span class="n">dfe</span><span class="o">.</span><span class="n">neutral_dfe</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">basic_contig</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
        <span class="n">recomb_map</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">RateMap</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">Contig</span><span class="p">(</span><span class="n">recombination_map</span><span class="o">=</span><span class="n">recomb_map</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">)</span>

<div class="viewcode-block" id="Contig.species_contig"><a class="viewcode-back" href="../../api.html#stdpopsim.Contig.species_contig">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">species_contig</span><span class="p">(</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">species</span><span class="p">,</span>
        <span class="n">chromosome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">genetic_map</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">length_multiplier</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">mutation_rate</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">inclusion_mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">exclusion_mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build a Contig for a species.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO: add non-autosomal support</span>
        <span class="n">non_autosomal_lower</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;m&quot;</span><span class="p">,</span> <span class="s2">&quot;mt&quot;</span><span class="p">,</span> <span class="s2">&quot;chrx&quot;</span><span class="p">,</span> <span class="s2">&quot;chry&quot;</span><span class="p">,</span> <span class="s2">&quot;chrm&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">chromosome</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">in</span> <span class="n">non_autosomal_lower</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
                <span class="n">stdpopsim</span><span class="o">.</span><span class="n">NonAutosomalWarning</span><span class="p">(</span>
                    <span class="s2">&quot;Non-autosomal simulations are not yet supported. See &quot;</span>
                    <span class="s2">&quot;https://github.com/popsim-consortium/stdpopsim/issues/383 and &quot;</span>
                    <span class="s2">&quot;https://github.com/popsim-consortium/stdpopsim/issues/406&quot;</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">genetic_map</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot use genetic map with generic contig&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">length_multiplier</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot use length multiplier for generic contig&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">inclusion_mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">exclusion_mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot use mask with generic contig&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Must specify sequence length of generic contig&quot;</span><span class="p">)</span>
            <span class="n">L_tot</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">r_tot</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">u_tot</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">chrom_data</span> <span class="ow">in</span> <span class="n">species</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">chrom_data</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">non_autosomal_lower</span><span class="p">:</span>
                    <span class="n">L_tot</span> <span class="o">+=</span> <span class="n">chrom_data</span><span class="o">.</span><span class="n">length</span>
                    <span class="n">r_tot</span> <span class="o">+=</span> <span class="n">chrom_data</span><span class="o">.</span><span class="n">length</span> <span class="o">*</span> <span class="n">chrom_data</span><span class="o">.</span><span class="n">recombination_rate</span>
                    <span class="n">u_tot</span> <span class="o">+=</span> <span class="n">chrom_data</span><span class="o">.</span><span class="n">length</span> <span class="o">*</span> <span class="n">chrom_data</span><span class="o">.</span><span class="n">mutation_rate</span>
            <span class="k">if</span> <span class="n">mutation_rate</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">mutation_rate</span> <span class="o">=</span> <span class="n">u_tot</span> <span class="o">/</span> <span class="n">L_tot</span>
            <span class="n">r</span> <span class="o">=</span> <span class="n">r_tot</span> <span class="o">/</span> <span class="n">L_tot</span>
            <span class="n">contig</span> <span class="o">=</span> <span class="n">Contig</span><span class="o">.</span><span class="n">basic_contig</span><span class="p">(</span>
                <span class="n">length</span><span class="o">=</span><span class="n">length</span><span class="p">,</span>
                <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span>
                <span class="n">recombination_rate</span><span class="o">=</span><span class="n">r</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot specify sequence length for named contig&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">inclusion_mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">exclusion_mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot specify both inclusion and exclusion masks&quot;</span><span class="p">)</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="n">species</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">get_chromosome</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">genetic_map</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Making flat chromosome </span><span class="si">{</span><span class="n">length_multiplier</span><span class="si">}</span><span class="s2"> * </span><span class="si">{</span><span class="n">chrom</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">gm</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">recomb_map</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">RateMap</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span>
                    <span class="nb">round</span><span class="p">(</span><span class="n">chrom</span><span class="o">.</span><span class="n">length</span> <span class="o">*</span> <span class="n">length_multiplier</span><span class="p">),</span> <span class="n">chrom</span><span class="o">.</span><span class="n">recombination_rate</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">length_multiplier</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot use length multiplier with empirical maps&quot;</span><span class="p">)</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Getting map for </span><span class="si">{</span><span class="n">chrom</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> from </span><span class="si">{</span><span class="n">genetic_map</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">gm</span> <span class="o">=</span> <span class="n">species</span><span class="o">.</span><span class="n">get_genetic_map</span><span class="p">(</span><span class="n">genetic_map</span><span class="p">)</span>
                <span class="n">recomb_map</span> <span class="o">=</span> <span class="n">gm</span><span class="o">.</span><span class="n">get_chromosome_map</span><span class="p">(</span><span class="n">chrom</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

            <span class="n">inclusion_intervals</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">exclusion_intervals</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="n">inclusion_mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">length_multiplier</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot use length multiplier with mask&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">inclusion_mask</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="n">inclusion_intervals</span> <span class="o">=</span> <span class="n">stdpopsim</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span>
                        <span class="n">inclusion_mask</span><span class="p">,</span> <span class="n">chromosome</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">inclusion_intervals</span> <span class="o">=</span> <span class="n">inclusion_mask</span>
            <span class="k">if</span> <span class="n">exclusion_mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">length_multiplier</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot use length multiplier with mask&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">exclusion_mask</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="n">exclusion_intervals</span> <span class="o">=</span> <span class="n">stdpopsim</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span>
                        <span class="n">exclusion_mask</span><span class="p">,</span> <span class="n">chromosome</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">exclusion_intervals</span> <span class="o">=</span> <span class="n">exclusion_mask</span>

            <span class="k">if</span> <span class="n">mutation_rate</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">mutation_rate</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">mutation_rate</span>

            <span class="n">contig</span> <span class="o">=</span> <span class="n">stdpopsim</span><span class="o">.</span><span class="n">Contig</span><span class="p">(</span>
                <span class="n">recombination_map</span><span class="o">=</span><span class="n">recomb_map</span><span class="p">,</span>
                <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span>
                <span class="n">genetic_map</span><span class="o">=</span><span class="n">gm</span><span class="p">,</span>
                <span class="n">inclusion_mask</span><span class="o">=</span><span class="n">inclusion_intervals</span><span class="p">,</span>
                <span class="n">exclusion_mask</span><span class="o">=</span><span class="n">exclusion_intervals</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">contig</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">recombination_map</span><span class="o">.</span><span class="n">sequence_length</span>

<div class="viewcode-block" id="Contig.dfe_breakpoints"><a class="viewcode-back" href="../../api.html#stdpopsim.Contig.dfe_breakpoints">[docs]</a>    <span class="k">def</span> <span class="nf">dfe_breakpoints</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns two things: the sorted vector of endpoints of all intervals across</span>
<span class="sd">        all DFEs in the contig, and a vector of integer labels for these intervals,</span>
<span class="sd">        saying which DFE goes with which interval.</span>
<span class="sd">        This provides a complementary method to tell which bit of the contig</span>
<span class="sd">        has which DFE attached, which may be more convenient than the list of two-column</span>
<span class="sd">        arrays provided by interval_list.</span>

<span class="sd">        Suppose there are n+1 unique interval endpoints across all the DFE intervals</span>
<span class="sd">        in :attr:`.interval_list`.  (If the intervals in that list</span>
<span class="sd">        cover the whole genome, the number of intervals is n.)</span>
<span class="sd">        This method returns a tuple of two things: ``breaks, dfe_labels``.</span>
<span class="sd">        &quot;breaks&quot; is the array containing those n+1 unique endpoints, in increasing order,</span>
<span class="sd">        and &quot;dfe&quot; is the array of length n containing the index of the DFE</span>
<span class="sd">        the applies to that interval. So, ``breaks[0]`` is always 0, and</span>
<span class="sd">        ``breaks[n+1]`` is always the length of the contig, and</span>
<span class="sd">        ``dfe_labels[k] = j`` if</span>
<span class="sd">        ``[breaks[k], breaks[k+1]]`` is an interval in ``contig.interval_list[j]``,</span>
<span class="sd">        i.e., if ``contig.dfe_list[j]`` applies to the interval starting at</span>
<span class="sd">        ``breaks[k]``.  Some intervals may not be covered by a DFE, in which</span>
<span class="sd">        case they will have the label ``-1`` (beware of python indexing!).</span>

<span class="sd">        :return: A tuple (breaks, dfe_labels).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">breaks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interval_list</span> <span class="o">+</span> <span class="p">[[[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">)]]])</span>  <span class="c1"># also sorted</span>
        <span class="p">)</span>
        <span class="n">dfe_labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">breaks</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">intervals</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interval_list</span><span class="p">):</span>
            <span class="n">dfe_labels</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">breaks</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">intervals</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">assume_unique</span><span class="o">=</span><span class="kc">True</span><span class="p">)]</span> <span class="o">=</span> <span class="n">j</span>
        <span class="k">return</span> <span class="n">breaks</span><span class="p">,</span> <span class="n">dfe_labels</span></div>

<div class="viewcode-block" id="Contig.clear_dfes"><a class="viewcode-back" href="../../api.html#stdpopsim.Contig.clear_dfes">[docs]</a>    <span class="k">def</span> <span class="nf">clear_dfes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Removes all DFEs from the contig (as well as the corresponding list of</span>
<span class="sd">        intervals).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dfe_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">interval_list</span> <span class="o">=</span> <span class="p">[]</span></div>

<div class="viewcode-block" id="Contig.add_dfe"><a class="viewcode-back" href="../../api.html#stdpopsim.Contig.add_dfe">[docs]</a>    <span class="k">def</span> <span class="nf">add_dfe</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">intervals</span><span class="p">,</span> <span class="n">DFE</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adds the provided DFE to the contig, applying to the regions of the</span>
<span class="sd">        contig specified in ``intervals``. These intervals are also *removed*</span>
<span class="sd">        from the intervals of any previously-present DFEs - in other words,</span>
<span class="sd">        more recently-added DFEs take precedence. Each DFE added in this way</span>
<span class="sd">        carries its own MutationTypes: no mutation types are shared between</span>
<span class="sd">        DFEs added by different calls to ``add_dfe( )``, even if the same DFE</span>
<span class="sd">        object is added more than once.</span>

<span class="sd">        For instance, if we do</span>
<span class="sd">        ```</span>
<span class="sd">        a1 = np.array([[0, 100]])</span>
<span class="sd">        a2 = np.array([[50, 120]])</span>
<span class="sd">        contig.add_dfe(a1, dfe1)</span>
<span class="sd">        contig.add_dfe(a2, dfe2)</span>
<span class="sd">        ```</span>
<span class="sd">        then ``dfe1`` applies to the region from 0 to 50 and ``dfe2`` applies to the</span>
<span class="sd">        region from 50 to 120.</span>

<span class="sd">        :param array intervals: A valid set of intervals.</span>
<span class="sd">        :param DFE dfe: A DFE object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stdpopsim</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">check_intervals_validity</span><span class="p">(</span><span class="n">intervals</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">ints</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interval_list</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">interval_list</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">stdpopsim</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">mask_intervals</span><span class="p">(</span><span class="n">ints</span><span class="p">,</span> <span class="n">intervals</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dfe_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">DFE</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">interval_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">intervals</span><span class="p">)</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">is_neutral</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns True if the contig has no non-neutral mutation types.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">all</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">is_neutral</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dfe_list</span> <span class="k">for</span> <span class="n">mt</span> <span class="ow">in</span> <span class="n">d</span><span class="o">.</span><span class="n">mutation_types</span><span class="p">)</span>

<div class="viewcode-block" id="Contig.mutation_types"><a class="viewcode-back" href="../../api.html#stdpopsim.Contig.mutation_types">[docs]</a>    <span class="k">def</span> <span class="nf">mutation_types</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Provides information about the MutationTypes assigned to this Contig,</span>
<span class="sd">        along with information about which DFE they correspond to. This is</span>
<span class="sd">        useful because when simulating with SLiM, the mutation types are</span>
<span class="sd">        assigned numeric IDs in the order provided here (e.g., in the order</span>
<span class="sd">        encountered when iterating over the DFEs); this method provides an easy</span>
<span class="sd">        way to map back from their numeric ID to the DFE that each MutationType</span>
<span class="sd">        corresponds to.</span>

<span class="sd">        This method returns a list of dictionaries of length equal to the</span>
<span class="sd">        number of MutationTypes in all DFEs of the contig, each dictionary</span>
<span class="sd">        containing three things: ``&quot;dfe_id&quot;``: the ID of the DFE this</span>
<span class="sd">        MutationType comes from; ``mutation_type``: the mutation type; and</span>
<span class="sd">        ``id``: the index in the list.</span>

<span class="sd">        For instance, if ``muts`` is a list of mutation objects in a SLiM tree</span>
<span class="sd">        sequence, then the following code will print the IDs of the DFEs that</span>
<span class="sd">        each comes from:</span>

<span class="sd">        .. code-block:: python</span>

<span class="sd">            mut_types = contig.mutation_types()</span>
<span class="sd">            for m in muts:</span>
<span class="sd">                dfe_ids = [mut_types[k][&quot;dfe_id&quot;] for md in m.metadata[&quot;muation_list&quot;]]</span>
<span class="sd">                print(f&quot;Mutation {m.id} has mutations from DFE(s) {&#39;,&#39;.join(dfe_ids)&quot;)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">id</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">mut_types</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dfe_list</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">mt</span> <span class="ow">in</span> <span class="n">d</span><span class="o">.</span><span class="n">mutation_types</span><span class="p">:</span>
                <span class="n">mut_types</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;dfe_id&quot;</span><span class="p">:</span> <span class="n">d</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                        <span class="s2">&quot;mutation_type&quot;</span><span class="p">:</span> <span class="n">mt</span><span class="p">,</span>
                        <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="nb">id</span><span class="p">,</span>
                    <span class="p">}</span>
                <span class="p">)</span>
                <span class="nb">id</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">mut_types</span></div>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">gmap</span> <span class="o">=</span> <span class="s2">&quot;None&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genetic_map</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">genetic_map</span><span class="o">.</span><span class="n">id</span>
        <span class="n">s</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;Contig(length=</span><span class="si">{:.2G}</span><span class="s2">, recombination_rate=</span><span class="si">{:.2G}</span><span class="s2">, &quot;</span>
            <span class="s2">&quot;mutation_rate=</span><span class="si">{:.2G}</span><span class="s2">, genetic_map=</span><span class="si">{}</span><span class="s2">, dfe_list=</span><span class="si">{}</span><span class="s2">, &quot;</span>
            <span class="s2">&quot;interval_list=</span><span class="si">{}</span><span class="s2">)&quot;</span>
        <span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">recombination_map</span><span class="o">.</span><span class="n">mean_rate</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mutation_rate</span><span class="p">,</span>
            <span class="n">gmap</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dfe_list</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">interval_list</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">s</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019-2020, PopSim Consortium.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>